<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  
  <title>Jinlong Zhang&#39;s blog</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta property="og:type" content="website">
<meta property="og:title" content="Jinlong Zhang&#39;s blog">
<meta property="og:url" content="https://helixcn.github.io/page/3/index.html">
<meta property="og:site_name" content="Jinlong Zhang&#39;s blog">
<meta property="og:locale" content="en_US">
<meta property="article:author" content="Jinlong Zhang">
<meta name="twitter:card" content="summary">
  
    <link rel="alternate" href="/atom.xml" title="Jinlong Zhang's blog" type="application/atom+xml">
  
  
    <link rel="shortcut icon" href="/favicon.png">
  
  
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-source-code-pro@0.0.71/index.min.css">

  
  
<link rel="stylesheet" href="/css/style.css">

  
    
<link rel="stylesheet" href="/fancybox/jquery.fancybox.min.css">

  
</head>

<body>
  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">Jinlong Zhang&#39;s blog</a>
      </h1>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"></a>
        
          <a class="main-nav-link" href="/">Home</a>
        
          <a class="main-nav-link" href="/archives">Archives</a>
        
          <a class="main-nav-link" href="/about">About</a>
        
      </nav>
      <nav id="sub-nav">
        
          <a id="nav-rss-link" class="nav-icon" href="/atom.xml" title="RSS Feed"></a>
        
        <a id="nav-search-btn" class="nav-icon" title="Search"></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="Search"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="https://helixcn.github.io"></form>
      </div>
    </div>
  </div>
</header>

      <div class="outer">
        <section id="main">
  
    <article id="post-2020-02-20-whatispic" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/02/19/2020-02-20-whatispic/" class="article-date">
  <time class="dt-published" datetime="2020-02-19T08:16:00.000Z" itemprop="datePublished">2020-02-19</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/02/19/2020-02-20-whatispic/">什么是系统发育独立差（Phylogenetic Independent Contrast）？</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>系统发育独立差（Phylogenetic Independent Contrast, PIC）是去除性状分析中物种系统发育关系的一种方法，是美国进化生物学家 J. Felsenstein于1985年提出的。</p>
<p>当时，有动物学家检验了若干哺乳动物脑容量和体重的相关性。然而，所用统计方法都假设各个种的性状是相互独立的，采集自同一个正态分布的总体。由于不同种之间存在系统发育上的联系，因此不能直接套用常规统计模型，所以所得结果也就不可信了。在Felsenstein之前，一些学者已经意识到这个问题，但是都没能给出理想的解决方案。</p>
<p>Felsenstain认真思考了这个问题，并且认为：研究多个不同性状的相关性时，必须考虑系统发育关系。他提出：假设物种性状的进化服从布朗运动，那么用系统发育关系上最近的分类单元性状的差值，再经过枝长加权，就可以得到一组新的数据。这组新的数据去掉了进化信息的干扰，因此可以用常规统计方法检验。上述方法新获得的数据就称为系统发育独立差 （The Phylogenetic Independent Contrast, PIC）。</p>
<p><a target="_blank" rel="noopener" href="http://image.sciencenet.cn/home/202002/20/162337u5u13hjeqj10911u.png"><img src="http://image.sciencenet.cn/home/202002/20/162337u5u13hjeqj10911u.png" alt="img"></a></p>
<p>图1. Felsenstein (1985) 的论文</p>
<p>举例来说，如果要研究哺乳动物脑容量和体重的关系，就要先获得这些种之间的进化关系，也就是经过分子钟校订的进化树。然后再计算脑容量的PIC和体重的PIC，再用脑容量的PIC和体重的PIC建立线性模型，进行相关性检验。</p>
<p>如果假设每个性状的进化服从布朗运动模型，也就是每个子代的性状的值是从祖先节点的性状随机变化而来的，那么进化树上相邻节点的两个种性状偏离于最近共同祖先性状的期望应该是0，且二者之间的差异正比于分化时间，也就是进化树的枝长。</p>
<p>在1985年的论文中，Felsenstein提出了如下算法：</p>
<ol>
<li>假设相邻的分类单元为<em>i</em>和<em>j</em>，其性状分别为<em>X_i</em>和<em>X_j</em>，<em>i</em>和<em>j</em>的共同祖先是节点为<em>k</em>。</li>
<li>先计算<em>X_i</em> - <em>X_j</em>，作为<em>X</em>和<em>Y</em>的contrast。该contrast的期望为0，方差与枝长<em>v_i</em>+<em>v_j</em>成正比</li>
<li>从进化树中去掉节点<em>i</em>和<em>j</em>，保留节点<em>k</em>，将<em>k</em>当做末端节点。具体来说，就是以<em>X_i</em>和<em>X_j</em>的加权平均值作为<em>X_k</em>，公式如下：</li>
</ol>
<p><a target="_blank" rel="noopener" href="http://image.sciencenet.cn/home/202002/20/162259xjn15une5xw55nw5.png"><img src="http://image.sciencenet.cn/home/202002/20/162259xjn15une5xw55nw5.png" alt="img"></a></p>
<ol start="4">
<li>将<em>k</em>节点下的枝长从<em>v_k</em>延长为<em>v_k</em> + <em>v_i**</em>v_j*/(<em>v_i</em> + <em>v_j</em>)。这样做的原因是，第3步中，<em>k</em>节点的<em>X_k</em>并非真实值，而是用<em>X_i</em> 、 <em>X_j</em>计算到的。</li>
</ol>
<p>经过1-4步，进化树上的末端分类单元就减少了1个，同时，也计算出了一个contrast的值。</p>
<p>重复1-4步，直到进化树中只有一个节点。如果进化树中有<em>n</em>个末端节点，那么上述过程将得到*n-*1个独立差(contrast)。图2-3展示了只有五个物种的进化树，PIC的计算公式。</p>
<p><a target="_blank" rel="noopener" href="http://image.sciencenet.cn/home/202002/20/162439xcxju1t4ujw89wno.png"><img src="http://image.sciencenet.cn/home/202002/20/162439xcxju1t4ujw89wno.png" alt="img"></a></p>
<p>图 2. 拥有5个分类单元的进化树，分类单元名称(1-5为末端节点，6-8为内部节点)、枝长（V）</p>
<p><a target="_blank" rel="noopener" href="http://image.sciencenet.cn/home/202002/20/162556bl09vjtnlsoj16j9.png"><img src="http://image.sciencenet.cn/home/202002/20/162556bl09vjtnlsoj16j9.png" alt="img"></a></p>
<p>图3. PIC的计算公式</p>
<p>ape程序包提供了计算PIC的函数pic()，该函数使用的正是Felsenstein （1985）的算法。一般用户并不需要了解函数内部的机制，只需要提供进化树和性状向量即可。</p>
<p>以下为ape程序包中pic函数的示例：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br></pre></td><td class="code"><pre><span class="line">library(ape)</span><br><span class="line"><span class="comment">### The example in Phylip 3.5c (originally from Lynch 1991)</span></span><br><span class="line">cat(<span class="string">&quot;((((Homo:0.21,Pongo:0.21):0.28,&quot;</span>,   </span><br><span class="line"><span class="string">&quot;Macaca:0.49):0.13,Ateles:0.62):0.38,Galago:1.00);&quot;</span>,   </span><br><span class="line">file = <span class="string">&quot;ex.tre&quot;</span>, sep = <span class="string">&quot;\n&quot;</span>)</span><br><span class="line">tree.primates &lt;- read.tree(<span class="string">&quot;ex.tre&quot;</span>)</span><br><span class="line">X &lt;- <span class="built_in">c</span>(<span class="number">4.09434</span>, <span class="number">3.61092</span>, <span class="number">2.37024</span>, <span class="number">2.02815</span>, -<span class="number">1.46968</span>)</span><br><span class="line">Y &lt;- <span class="built_in">c</span>(<span class="number">4.74493</span>, <span class="number">3.33220</span>, <span class="number">3.36730</span>, <span class="number">2.89037</span>, <span class="number">2.30259</span>)</span><br><span class="line"><span class="built_in">names</span>(X) &lt;- <span class="built_in">names</span>(Y) &lt;- <span class="built_in">c</span>(<span class="string">&quot;Homo&quot;</span>, <span class="string">&quot;Pongo&quot;</span>, <span class="string">&quot;Macaca&quot;</span>, <span class="string">&quot;Ateles&quot;</span>, <span class="string">&quot;Galago&quot;</span>)</span><br><span class="line">pic.X &lt;- pic(X, tree.primates)</span><br><span class="line">pic.Y &lt;- pic(Y, tree.primates)</span><br><span class="line">cor.test(X, Y)</span><br><span class="line"><span class="comment">####  Pearson&#x27;s product-moment correlation</span></span><br><span class="line"><span class="comment">#### data:  X and Y</span></span><br><span class="line"><span class="comment">## t = 2.5736, df = 3, p-value = 0.08224</span></span><br><span class="line"><span class="comment">## alternative hypothesis: true correlation is not equal to 0</span></span><br><span class="line"><span class="comment">## 95 percent confidence interval:</span></span><br><span class="line"><span class="comment">##  -0.1964310  0.9884173</span></span><br><span class="line"><span class="comment">## sample estimates:</span></span><br><span class="line"><span class="comment">##       cor</span></span><br><span class="line"><span class="comment">## 0.8296107</span></span><br><span class="line">cor.test(pic.X, pic.Y)</span><br><span class="line"><span class="comment">####  Pearson&#x27;s product-moment correlation</span></span><br><span class="line"><span class="comment">#### data:  pic.X and pic.Y</span></span><br><span class="line"><span class="comment">## t = -0.85623, df = 2, p-value = 0.4821</span></span><br><span class="line"><span class="comment">## alternative hypothesis: true correlation is not equal to 0</span></span><br><span class="line"><span class="comment">## 95 percent confidence interval:</span></span><br><span class="line"><span class="comment">##  -0.9874751  0.8823934</span></span><br><span class="line"><span class="comment">## sample estimates:</span></span><br><span class="line"><span class="comment">##        cor</span></span><br><span class="line"><span class="comment">## -0.5179156</span></span><br><span class="line">lm(pic.Y ~ pic.X - <span class="number">1</span>) <span class="comment"># both regressions</span></span><br><span class="line"><span class="comment">#### Call:</span></span><br><span class="line"><span class="comment">## lm(formula = pic.Y ~ pic.X - 1)</span></span><br><span class="line"><span class="comment">#### Coefficients:</span></span><br><span class="line"><span class="comment">##  pic.X</span></span><br><span class="line"><span class="comment">## 0.4319</span></span><br><span class="line">lm(pic.X ~ pic.Y - <span class="number">1</span>) </span><br><span class="line"><span class="comment"># through the origin</span></span><br><span class="line"><span class="comment">#### Call:</span></span><br><span class="line"><span class="comment">## lm(formula = pic.X ~ pic.Y - 1)</span></span><br><span class="line"><span class="comment">#### Coefficients:</span></span><br><span class="line"><span class="comment">## pic.Y</span></span><br><span class="line"><span class="comment">## 0.998</span></span><br></pre></td></tr></table></figure>





<p>从以上结果可以看出，将性状直接进行相关分析以及用性状的PIC进行相关分析，所得结果是有差异的：去除了系统发育关系之后，性状X和性状Y之间的相关性降低了。去除系统发育关系之前 （Pearson’s <em>r</em> = 0.8296, <em>t</em> = 2.5736, <em>df</em> = 3, <em>p</em>-value = 0.08224），去除系统发育关系之后（Pearson’s <em>r</em> = -0.5179156, <em>t</em> = -0.85623, <em>df</em> = 2, <em>p</em>-value = 0.4821 ）。</p>
<p>这里比较难以理解的是，计算出来的PIC总是比真实性状数据少一个，但是不同性状的PIC都是比真实数据少一个，用不同性状的PIC向量建立线性模型，所得结果，就是我们要计算的性状之间的相关性。</p>
<p>ape的pic函数要求不同种性状出现的顺序与在进化树中出现的顺序完全相同。为此，picante程序包提供了match.phylo.data()函数，只要用户为性状数据（无论是data.frame还是vector）提供了对应的物种名（如果是多个性状，格式应该为dataframe，其行名应该是物种名），该函数就能将性状数据按照进化树中出现的顺序排好，非常方便。</p>
<p>以下代码展示的就是怎样通过植物物种名（拉丁名）获得进化树，再计算植株高度的PIC，源数据来自S. Kembel博士的培训资料。</p>
<p>Felsenstein在1985年的论文中说，他提出的算法可计算多分枝结构的PIC，但是ape的pic函数不能处理含有多分枝结构的进化树，所以下面的代码调用了multi2di()函数，将进化树转换为二叉树。转换的方法是为每一个多分支节点随机添加一个很小的枝长，由于枝长很小，对结果的影响很小，几乎可以忽略不计。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br></pre></td><td class="code"><pre><span class="line">setwd(<span class="string">&quot;...&quot;</span>)</span><br><span class="line">library(plantlist)</span><br><span class="line"><span class="comment">## This is plantlist 0.6.1.</span></span><br><span class="line"><span class="comment"># devtools::install_github(&quot;helixcn/plantlist&quot;)</span></span><br><span class="line">library(V.PhyloMaker)</span><br><span class="line"><span class="comment"># devtools::install_github(&quot;jinyizju/V.PhyloMaker&quot;)</span></span><br><span class="line">library(picante)</span><br><span class="line">library(ape)</span><br><span class="line">dat &lt;- read.csv(<span class="string">&quot;trait_herb_114.csv&quot;</span>, header = <span class="literal">TRUE</span>)</span><br><span class="line">head(dat)</span><br><span class="line"><span class="comment">##                 species height</span></span><br><span class="line"><span class="comment">## 1     Stipa_albasiensis   48.5</span></span><br><span class="line"><span class="comment">## 2 Heteropappus_altaicus   15.6</span></span><br><span class="line"><span class="comment">## 3  Roegneria_alashanica   43.7</span></span><br><span class="line"><span class="comment">## 4 Saussurea_alaschanica   21.5</span></span><br><span class="line"><span class="comment">## 5 Astragalus_alaschanus   18.0</span></span><br><span class="line"><span class="comment">## 6      Panzerina_lanata   12.5</span></span><br><span class="line">height &lt;- dat$height</span><br><span class="line"><span class="built_in">names</span>(height) &lt;- dat$species</span><br><span class="line"></span><br><span class="line"><span class="comment"># Build a phylogeny using phylo.maker</span></span><br><span class="line">species &lt;- gsub(<span class="string">&quot;_&quot;</span>, <span class="string">&quot; &quot;</span>, <span class="built_in">as.character</span>(dat$species))</span><br><span class="line">tab2 &lt;- subset(TPL(gsub(<span class="string">&quot;_&quot;</span>, <span class="string">&quot; &quot;</span>, species)),              </span><br><span class="line">select = <span class="built_in">c</span>(<span class="string">&quot;YOUR_SEARCH&quot;</span>, <span class="string">&quot;POSSIBLE_GENUS&quot;</span>, <span class="string">&quot;FAMILY&quot;</span>))</span><br><span class="line">colnames(tab2) &lt;- <span class="built_in">c</span>(<span class="string">&quot;species&quot;</span>, <span class="string">&quot;genus&quot;</span>, <span class="string">&quot;family&quot;</span>)</span><br><span class="line">result1 &lt;- phylo.maker(tab2, scenarios = <span class="built_in">c</span>(<span class="string">&quot;S1&quot;</span>))</span><br><span class="line"><span class="comment">## [1] &quot;Taxonomic classification not consistent between sp.list and tree.&quot;##           genus family_in_sp.list family_in_tree## 77 Hemerocallis  Xanthorrhoeaceae  Asphodelaceae</span></span><br><span class="line"><span class="comment"># write.tree(result1[[1]], &quot;tree1.newick&quot;) </span></span><br><span class="line"><span class="comment"># If you want to examine the tree</span></span><br><span class="line">tree &lt;- multi2di(result1[[<span class="number">1</span>]])</span><br><span class="line"><span class="comment"># plot(tree)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># &quot;If x has names, its values are matched to the tip labels of phy, otherwise its values are taken to be in the same order than the tip labels of phy.&quot;</span></span><br><span class="line">data_matched &lt;- match.phylo.data(phy = tree, data = height )</span><br><span class="line"><span class="comment"># Phylogenetic Independent Contrastheight_pic &lt;- pic(data_matched$data, data_matched$phy)</span></span><br></pre></td></tr></table></figure>

<h1 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h1><ul>
<li>Felsenstein, J. (1985) Phylogenies and the comparative method. American Naturalist, 125, 1–15.</li>
<li>Jin, Y., &amp; Qian, H. (2019). V. PhyloMaker: an R package that can generate very large phylogenies for vascular plants. Ecography, 42(8), 1353-1359.</li>
<li>Kembel, S. W., Cowan, P. D., Helmus, M. R., Cornwell, W. K., Morlon, H., Ackerly, D. D., … &amp; Webb, C. O. (2010). Picante: R tools for integrating phylogenies and ecology. Bioinformatics, 26(11), 1463-1464.</li>
<li>Kembel S. 编， 张金龙 译， 在R中分析物种和系统发育多样性 <a target="_blank" rel="noopener" href="http://blog.sciencenet.cn/home.php?mod=space&amp;uid=255662&amp;do=blog&amp;id=1116212">http://blog.sciencenet.cn/home.php?mod=space&amp;uid=255662&amp;do=blog&amp;id=1116212</a></li>
<li>Paradis, E., &amp; Schliep, K. (2019). ape 5.0: an environment for modern phylogenetics and evolutionary analyses in R. <em>Bioinformatics</em>, <em>35</em>(3), 526-528.</li>
<li>Zhang, J.L. (2018). plantlist: Looking Up the Status of Plant Scientific Names based on The Plant List Database. R package version 0.6.1. <a target="_blank" rel="noopener" href="https://github.com/helixcn/plantlist/">https://github.com/helixcn/plantlist/</a></li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2020/02/19/2020-02-20-whatispic/" data-id="cl1n7qj9b00e5eo84gcjgf7c0" data-title="什么是系统发育独立差（Phylogenetic Independent Contrast）？" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2020-01-23-tiefeng" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/01/23/2020-01-23-tiefeng/" class="article-date">
  <time class="dt-published" datetime="2020-01-22T23:09:00.000Z" itemprop="datePublished">2020-01-23</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/01/23/2020-01-23-tiefeng/">再忆铁峰</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>再忆铁峰</p>
<p>又到过年了，又想起了铁峰兄。铁峰姓朴，朝鲜族，在黑龙江齐齐哈尔长大，年长我一岁，离开这个世界已经五年了。</p>
<p>十年前，我还在读博的时候，有一次跟研究组的米湘成老师去黑龙江采集树木DNA条形码的样品，就在那次认识的铁峰。</p>
<p>那时候，铁峰还在韩国江原道大学读博士，但主要野外工作是在黑龙江的凉水和丰林自然保护区做的，以东北林业大学金光泽教授的实验室为基地。</p>
<p>到了哈尔滨，我和米老师先拜访了东林的金光泽老师，金老师、铁峰跟我们一起去采样。</p>
<p>我们坐着老式绿皮火车，先去凉水的实习基地。</p>
<p>初夏，东北的景色十分宜人。窗外，小山丘起伏十分平缓，一路上上下下。山上，草木葱葱郁郁地生长着。低地上是一片又一片农田，绿油油的，有时一眼望不到尽头。一座又一座不知道姓名的小村，镶嵌在绿色的地毯上。天空中一朵一朵的白云，懒洋洋地向前移动着。那天阳光特别好，风从车窗飕飕地吹进来，十分温暖。时间仿佛过得又快，又漫长，一个又一个小站，绥化、庆安、神树、带岭……铁峰就坐我身边，我们在火车上聊了一路，到了凉水，天色已晚。</p>
<p>第二天，我们要在这里九公顷的样地采集DNA条形码的样品，大家要一起钻林子了。</p>
<p>正是初夏，凉水的草爬子（全沟蜱 Ixodes persulcatus）很“厚”，虽然我很小心，还是被叮了三处：胸口、肚皮、腋窝。采了植物样品，回到住处，我赶紧脱下上衣，让铁峰帮看看后背上有没有草爬子，然后我帮他看。</p>
<p>确认身上和衣服上没有草爬子，我和铁峰一起在实习基地的楼道处理样品：把叶片装进小塑料袋，加变色硅胶，袋上写标本编号，并采集标本。这个工作有些枯燥，我和他边干活边闲聊。</p>
<p>那时候我没有出过国，很好奇他的留学经历，于是问：“你是朝鲜族，到了韩国，会不会觉得自己回到自己的国家？”</p>
<p>铁峰说：“虽说我是朝鲜族，但是一直说在中国长大，说汉语，只能说一些简单的朝鲜语。朝鲜族本身就是中国的一个民族，我自己就是中国人啊。”</p>
<p>我十分赞许他的话，连忙点头。</p>
<p>凉水样地采样结束后的第二天，我们又乘火车到伊春五营的丰林50公顷大样地采样。</p>
<p>丰林大洋地地势平缓，森林茂密，有很多大树。林下走起来并不困难，但是有时候一些倒木拦在前头，有的地段比较泥泞，这样的地段就有点儿困难，只能手拉手过去。</p>
<p>由于我们事先准备好了调查记录，只需要按照号牌找到相应的树就可以了，采样相对简单。</p>
<p>丰林大样地的记录里，最后要找一棵大青杨（Populus ussuriensis Kom.），但找了许久都没找到。最后，按照调查记录的坐标找到号码相同的一棵树，原来是一棵大黄柳（Salix raddeana Laksch.）。那一刻，大家都笑了起来，原来是鉴定错误。铁峰也笑得和孩子一样纯真。保护区的宋主任有点儿不好意思，也跟着一起说：“哪儿有大青杨啊？根本就是大黄柳！”</p>
<p>铁峰在丰林样地都工作了很久，和保护区的人员很熟，在回去的车上，保护区的工作人员开玩笑称呼他为“铁棒”，这时候他就低下头，笑一笑，也不做声。</p>
<p>晚上，大家一起吃饭。保护区工作人员非常热情，一直不停劝酒。我不胜酒力，被灌多了，感觉浑身发冷，见情况不妙，赶紧招呼铁峰。铁峰见我面色发白，就搀着我回住处。</p>
<p>路上，他跟我说：“幸好你喝多了，我正好才能逃出来，不然也肯定高了。” </p>
<p>夜色朦胧，我只恍惚记着摇晃的街灯，走不动路的双腿。他搀着我，需要的时候轻轻拍打我后背。</p>
<p>离开伊春，再见面好像就是CForBio大样地研究网络数据分析培训班，培训内容好像是植物功能性状分析。那次我给来自全国各地的几十位学员讲了用R怎样基于植物的功能性状建立聚类树，以进行功能多样性分析。中午，我拉着铁峰到香山上吃饭。他特别客气，争着要买单。</p>
<p>后来，听说他从韩国留学回国，去西北农林科技大学工作了。那之后我们经常在QQ上联系，聊各种问题，生活的、教学的、科研的，每次都能聊得很投机。像铁峰这样，能聊到一起，能一起说真话，能懂你的好友，十分难得。也就那期间，他在密度制约方向发了两篇很好的论文，我向他表示祝贺。</p>
<p>2015年过年，我用QQ发消息给他，祝他新年快乐。他的QQ发来消息说：你不知道吗？铁峰已经不在了。</p>
<p>我心头一惊，顿时说不出话来，赶紧打他的手机。接电话的是他姨，告诉我说：他的亲人正在西安处理他的后事……</p>
<p>电话里，我才知道，他是由妈妈一个人抚养长大；才知道他家人到学校后的种种遭遇，特别是工伤认定、补偿金等问题，很长时间才解决好。后来，我又和他家人通了几次电话，只能尽量安慰他的妈妈，说自己愿意提供力所能及的帮助。</p>
<p>铁峰离开这个世界五年了，总让人感觉这一切不是真的。一位优秀的青年生态学家，忽然之间，就走过了短暂的一生，怎能不让人悲伤。</p>
<p>如果有来世，我还愿意认识他，这个简单而纯粹的人。</p>
<p>纯粹的人，总是值得怀念。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2020/01/23/2020-01-23-tiefeng/" data-id="cl1n7qj9a00e3eo84caqcch96" data-title="再忆铁峰" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2020-01-10-mrca" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2020/01/10/2020-01-10-mrca/" class="article-date">
  <time class="dt-published" datetime="2020-01-09T22:27:00.000Z" itemprop="datePublished">2020-01-10</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2020/01/10/2020-01-10-mrca/">计算进化树末端节点出现的时间及两节点共同祖先（MRCA）的年龄</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="目的："><a href="#目的：" class="headerlink" title="目的："></a>目的：</h1><ol>
<li>建立进化树</li>
<li>获得进化树中每个种的延续时间</li>
<li>获得某两个种最近共同祖先（MRCA）距离现在的时间</li>
</ol>
<h1 id="代码"><a href="#代码" class="headerlink" title="代码"></a>代码</h1><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">setwd(<span class="string">&quot;C:/Users/jlzhang/Desktop/age&quot;</span>)</span><br><span class="line">library(plantlist)    </span><br><span class="line"><span class="comment"># devtools::install_github(&quot;helixcn/plantlist&quot;)</span></span><br><span class="line"><span class="comment">## This is plantlist 0.6.1.</span></span><br><span class="line">library(V.PhyloMaker) </span><br><span class="line"><span class="comment"># devtools::install_github(&quot;jinyizju/V.PhyloMaker&quot;)</span></span><br><span class="line">library(phytools)</span><br><span class="line">library(dispRity)</span><br><span class="line">library(openxlsx)</span><br><span class="line">library(phytools)</span><br></pre></td></tr></table></figure>

<h2 id="用中文名建进化树"><a href="#用中文名建进化树" class="headerlink" title="用中文名建进化树"></a>用中文名建进化树</h2><h3 id="查询学名"><a href="#查询学名" class="headerlink" title="查询学名"></a>查询学名</h3><p>注意xlsx中保存的是中文名</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br></pre></td><td class="code"><pre><span class="line">species &lt;- read.xlsx(<span class="string">&quot;checklist.xlsx&quot;</span>)</span><br><span class="line">taxa.table(TPL(CTPL(species[,<span class="number">1</span>])$SPECIES), <span class="string">&quot;taxa_table.txt&quot;</span>)</span><br><span class="line"><span class="comment">## Warning in CTPL(species[, 1]): Taxa: 红枝柴</span></span><br><span class="line"><span class="comment">##  could not be recognized.</span></span><br><span class="line"><span class="comment">## Note: too many rows to show, only the first few rows were printedTAXA_NAME :</span></span><br><span class="line"><span class="comment">## [1] &quot;木荷&quot;     &quot;厚皮香&quot;   &quot;狗骨柴&quot;   &quot;杜鹃&quot;     &quot;薄叶山矾&quot; &quot;赤楠&quot;    </span></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## SPECIES :</span></span><br><span class="line"><span class="comment">## [1] &quot;Schima superba&quot;           &quot;Ternstroemia gymnanthera&quot;</span></span><br><span class="line"><span class="comment">## [3] &quot;Diplospora dubia&quot;         &quot;Rhododendron simsii&quot;     </span></span><br><span class="line"><span class="comment">## [5] &quot;Symplocos anomala&quot;        &quot;Syzygium buxifolium&quot;     </span></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## SPECIES_FULL :</span></span><br><span class="line"><span class="comment">## [1] &quot;Schima superba Gardn. et Champ.&quot;                 </span></span><br><span class="line"><span class="comment">## [2] &quot;Ternstroemia gymnanthera (Wight et Arn.) Beddome&quot;</span></span><br><span class="line"><span class="comment">## [3] &quot;Diplospora dubia (Lindl.) Masam.&quot;                </span></span><br><span class="line"><span class="comment">## [4] &quot;Rhododendron simsii Planch.&quot;                     </span></span><br><span class="line"><span class="comment">## [5] &quot;Symplocos anomala Brand&quot;                         </span></span><br><span class="line"><span class="comment">## [6] &quot;Syzygium buxifolium Hook. et Arn.&quot;               </span></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## GENUS :</span></span><br><span class="line"><span class="comment">## [1] &quot;Schima&quot;       &quot;Ternstroemia&quot; &quot;Diplospora&quot;   &quot;Rhododendron&quot; &quot;Symplocos&quot;   </span></span><br><span class="line"><span class="comment">## [6] &quot;Syzygium&quot;    </span></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## GENUS_AUTHOR :</span></span><br><span class="line"><span class="comment">## [1] &quot;Reinw. ex Blume&quot; &quot;Mutis ex L. f.&quot;  &quot;DC.&quot;             &quot;L.&quot;             </span></span><br><span class="line"><span class="comment">## [5] &quot;Jacq.&quot;           &quot;Gaertn.&quot;        </span></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## GENUS_CN :</span></span><br><span class="line"><span class="comment">## [1] &quot;木荷属&quot;   &quot;厚皮香属&quot; &quot;狗骨柴属&quot; &quot;杜鹃花属&quot; &quot;山矾属&quot;   &quot;蒲桃属&quot;  </span></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## FAMILY :</span></span><br><span class="line"><span class="comment">## [1] &quot;Theaceae&quot;         &quot;Pentaphylacaceae&quot; &quot;Rubiaceae&quot;        &quot;Ericaceae&quot;       </span></span><br><span class="line"><span class="comment">## [5] &quot;Symplocaceae&quot;     &quot;Myrtaceae&quot;       </span></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## FAMILY_CN :</span></span><br><span class="line"><span class="comment">## [1] &quot;山茶科&quot;   &quot;五列木科&quot; &quot;茜草科&quot;   &quot;杜鹃花科&quot; &quot;山矾科&quot;   &quot;桃金娘科&quot;</span></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## FAMILY_NUMBER :</span></span><br><span class="line"><span class="comment">## [1] &quot;APGIII_334&quot; &quot;APGIII_330&quot; &quot;APGIII_350&quot; &quot;APGIII_344&quot; &quot;APGIII_335&quot;</span></span><br><span class="line"><span class="comment">## [6] &quot;APGIII_222&quot;</span></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## ORDER :</span></span><br><span class="line"><span class="comment">## [1] &quot;Ericales&quot;    &quot;Ericales&quot;    &quot;Gentianales&quot; &quot;Ericales&quot;    &quot;Ericales&quot;   </span></span><br><span class="line"><span class="comment">## [6] &quot;Myrtales&quot;   </span></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## GROUP :</span></span><br><span class="line"><span class="comment">## [1] &quot;Angiosperms&quot; &quot;Angiosperms&quot; &quot;Angiosperms&quot; &quot;Angiosperms&quot; &quot;Angiosperms&quot;</span></span><br><span class="line"><span class="comment">## [6] &quot;Angiosperms&quot;</span></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## IUCN_CHINA :</span></span><br><span class="line"><span class="comment">## [1] &quot;无危（LC）&quot; &quot;无危（LC）&quot; &quot;无危（LC）&quot; &quot;无危（LC）&quot; &quot;无危（LC）&quot;</span></span><br><span class="line"><span class="comment">## [6] &quot;无危（LC）&quot;</span></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## ENDEMIC_TO_CHINA :</span></span><br><span class="line"><span class="comment">## [1] &quot;否&quot; &quot;否&quot; &quot;否&quot; &quot;否&quot; &quot;否&quot; &quot;否&quot;</span></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## PROVINTIAL_DISTRIBUTION :</span></span><br><span class="line"><span class="comment">## [1] &quot;浙、赣、湘、黔、闽、台、粤、桂、琼、港&quot;                </span></span><br><span class="line"><span class="comment">## [2] &quot;皖、浙、赣、湘、鄂、川、黔、闽、粤、桂、滇、琼、港&quot;    </span></span><br><span class="line"><span class="comment">## [3] &quot;皖、苏、浙、赣、湘、川、闽、台、粤、桂、滇、琼、港&quot;    </span></span><br><span class="line"><span class="comment">## [4] &quot;皖、苏、浙、赣、湘、鄂、川、黔、闽、台、粤、桂、滇&quot;    </span></span><br><span class="line"><span class="comment">## [5] &quot;皖、苏、浙、湘、鄂、川、黔、闽、台、粤、桂、滇、藏、琼&quot;</span></span><br><span class="line"><span class="comment">## [6] &quot;皖、浙、赣、湘、黔、闽、台、粤、桂&quot;                    </span></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## ALTITUDE :</span></span><br><span class="line"><span class="comment">## [1] &quot;1000&quot;                     &quot;200-1400(2000-2800云南）&quot;</span></span><br><span class="line"><span class="comment">## [3] &quot;40-1500&quot;                  &quot;500-1200（-2500）&quot;       </span></span><br><span class="line"><span class="comment">## [5] &quot;400-3000&quot;                 &quot;&quot;</span></span><br><span class="line"><span class="comment">##  [1] &quot;Theaceae/Schima/Schima_superba&quot;                        </span></span><br><span class="line"><span class="comment">##  [2] &quot;Pentaphylacaceae/Ternstroemia/Ternstroemia_gymnanthera&quot;</span></span><br><span class="line"><span class="comment">##  [3] &quot;Pentaphylacaceae/Cleyera/Cleyera_japonica&quot;             </span></span><br><span class="line"><span class="comment">##  [4] &quot;Rubiaceae/Gardenia/Gardenia_jasminoides&quot;               </span></span><br><span class="line"><span class="comment">##  [5] &quot;Rubiaceae/Diplospora/Diplospora_dubia&quot;                 </span></span><br><span class="line"><span class="comment">##  [6] &quot;Ericaceae/Rhododendron/Rhododendron_simsii&quot;            </span></span><br><span class="line"><span class="comment">##  [7] &quot;Symplocaceae/Symplocos/Symplocos_anomala&quot;              </span></span><br><span class="line"><span class="comment">##  [8] &quot;Symplocaceae/Symplocos/Symplocos_sumuntia&quot;             </span></span><br><span class="line"><span class="comment">##  [9] &quot;Symplocaceae/Symplocos/Symplocos_lucida&quot;               </span></span><br><span class="line"><span class="comment">## [10] &quot;Symplocaceae/Symplocos/Symplocos_stellaris&quot;            </span></span><br><span class="line"><span class="comment">## [11] &quot;Myrtaceae/Syzygium/Syzygium_buxifolium&quot;                </span></span><br><span class="line"><span class="comment">## [12] &quot;Aquifoliaceae/Ilex/Ilex_ficoidea&quot;                      </span></span><br><span class="line"><span class="comment">## [13] &quot;Aquifoliaceae/Ilex/Ilex_rotunda&quot;                       </span></span><br><span class="line"><span class="comment">## [14] &quot;Aquifoliaceae/Ilex/Ilex_micrococca&quot;                    </span></span><br><span class="line"><span class="comment">## [15] &quot;Rutaceae/Evodia/Evodia_glabrifolia&quot;                    </span></span><br><span class="line"><span class="comment">## [16] &quot;Daphniphyllaceae/Daphniphyllum/Daphniphyllum_oldhami&quot;  </span></span><br><span class="line"><span class="comment">## [17] &quot;Lauraceae/Litsea/Litsea_elongata&quot;                      </span></span><br><span class="line"><span class="comment">## [18] &quot;Lauraceae/Litsea/Litsea_cubeba&quot;                        </span></span><br><span class="line"><span class="comment">## [19] &quot;Rosaceae/Photinia/Photinia_beauverdiana&quot;               </span></span><br><span class="line"><span class="comment">## [20] &quot;Moraceae/Ficus/Ficus_erecta&quot;                           </span></span><br><span class="line"><span class="comment">## [21] &quot;Myricaceae/Myrica/Myrica_rubra&quot;                        </span></span><br><span class="line"><span class="comment">## [22] &quot;Styracaceae/Styrax/Styrax_suberifolius&quot;                </span></span><br><span class="line"><span class="comment">## [23] &quot;Fagaceae/Castanopsis/Castanopsis_carlesii&quot;             </span></span><br><span class="line"><span class="comment">## [24] &quot;Fagaceae/Cyclobalanopsis/Cyclobalanopsis_gracilis&quot;     </span></span><br><span class="line"><span class="comment">## [25] &quot;Ebenaceae/Diospyros/Diospyros_morrisiana&quot;              </span></span><br><span class="line"><span class="comment">## [26] &quot;Elaeocarpaceae/Elaeocarpus/Elaeocarpus_japonicus&quot;      </span></span><br><span class="line"><span class="comment">## [27] &quot;Betulaceae/Carpinus/Carpinus_viminea&quot;                  </span></span><br><span class="line"><span class="comment">## [28] &quot;Pittosporaceae/Pittosporum/Pittosporum_illicioides&quot;</span></span><br></pre></td></tr></table></figure>

<h3 id="读取查询到的学名"><a href="#读取查询到的学名" class="headerlink" title="读取查询到的学名"></a>读取查询到的学名</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">tab &lt;- read.table(<span class="string">&quot;taxa_table.txt&quot;</span>, sep = <span class="string">&quot;/&quot;</span>)</span><br><span class="line">colnames(tab) &lt;- <span class="built_in">c</span>(<span class="string">&quot;family&quot;</span>, <span class="string">&quot;genus&quot;</span>, <span class="string">&quot;species&quot;</span>)</span><br><span class="line">tab2 &lt;- tab[, <span class="built_in">c</span>(<span class="string">&quot;species&quot;</span>, <span class="string">&quot;genus&quot;</span>, <span class="string">&quot;family&quot;</span>)]</span><br></pre></td></tr></table></figure>

<h3 id="用V-PhyloMaker建树，速度较慢"><a href="#用V-PhyloMaker建树，速度较慢" class="headerlink" title="用V.PhyloMaker建树，速度较慢"></a>用V.PhyloMaker建树，速度较慢</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">result1 &lt;- phylo.maker(tab2, scenarios = <span class="built_in">c</span>(<span class="string">&quot;S1&quot;</span>))</span><br><span class="line">write.tree(result1[[<span class="number">1</span>]], <span class="string">&quot;tree1.newick&quot;</span>) </span><br><span class="line"><span class="comment"># 保存到硬盘，用于用其他软件（如Figtree）查看</span></span><br><span class="line">tree &lt;- result1[[<span class="number">1</span>]]</span><br></pre></td></tr></table></figure>

<h2 id="获得每个种延续的时间"><a href="#获得每个种延续的时间" class="headerlink" title="获得每个种延续的时间"></a>获得每个种延续的时间</h2><p>注意，这个与进化树收录的分类单元密切相关，一般不宜用于研究。</p>
<p>脚本来源 <a target="_blank" rel="noopener" href="http://blog.phytools.org/2013/10/finding-edge-lengths-of-all-terminal.html">http://blog.phytools.org/2013/10/finding-edge-lengths-of-all-terminal.html</a></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">n &lt;- <span class="built_in">length</span>(tree$tip.label)</span><br><span class="line">ee &lt;- setNames(tree$edge.length[sapply(<span class="number">1</span>:n, <span class="keyword">function</span>(x, y) which(y == x), y = tree$edge[, <span class="number">2</span>])], tree$tip.label)</span><br><span class="line">ee &lt;- as.data.frame(ee)</span><br><span class="line">write.xlsx(ee, <span class="string">&quot;terminal_branch_length.xlsx&quot;</span>, row.names = <span class="literal">TRUE</span>)</span><br></pre></td></tr></table></figure>

<h3 id="查询两个分类单元最近共同祖先的节点名"><a href="#查询两个分类单元最近共同祖先的节点名" class="headerlink" title="查询两个分类单元最近共同祖先的节点名"></a>查询两个分类单元最近共同祖先的节点名</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">tree$node.label &lt;- paste(<span class="number">1</span>:tree$Nnode + Ntip(tree))</span><br><span class="line">name_node &lt;- getMRCA(tree, tip = <span class="built_in">c</span>(<span class="string">&quot;Cleyera_japonica&quot;</span>, <span class="string">&quot;Rhododendron_simsii&quot;</span>))</span><br><span class="line">name_node_ind &lt;- tree$node.label[name_node - Ntip(tree)]</span><br></pre></td></tr></table></figure>

<h3 id="绘制进化树"><a href="#绘制进化树" class="headerlink" title="绘制进化树"></a>绘制进化树</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">plot(tree)</span><br><span class="line">axisPhylo(<span class="number">1</span>, las = <span class="number">1</span>)</span><br><span class="line"><span class="comment">## see  </span></span><br><span class="line">nodelabels(text=tree$node.label,  node=<span class="number">1</span>:tree$Nnode+Ntip(tree))</span><br></pre></td></tr></table></figure>

<p><a target="_blank" rel="noopener" href="http://image.sciencenet.cn/home/202001/10/142643d0fwx9xn79i90zrd.png"><img src="http://image.sciencenet.cn/home/202001/10/142643d0fwx9xn79i90zrd.png" alt="img"></a></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">plot(tree)</span><br><span class="line">axisPhylo(<span class="number">1</span>, las = <span class="number">1</span>)</span><br><span class="line">nodelabels(paste(<span class="string">&quot;MRCA:&quot;</span>, name_node), name_node, frame = <span class="string">&quot;r&quot;</span>, bg = <span class="string">&quot;yellow&quot;</span>, adj = <span class="number">0</span>)</span><br></pre></td></tr></table></figure>

<p><a target="_blank" rel="noopener" href="http://image.sciencenet.cn/home/202001/10/142657vxss8amsrf8am959.png"><img src="http://image.sciencenet.cn/home/202001/10/142657vxss8amsrf8am959.png" alt="img"></a></p>
<h3 id="查询两个种的最近共同祖先MRCA的时间"><a href="#查询两个种的最近共同祖先MRCA的时间" class="headerlink" title="查询两个种的最近共同祖先MRCA的时间"></a>查询两个种的最近共同祖先MRCA的时间</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">### From the dispRity Package</span></span><br><span class="line">age_dat &lt;- tree.age(tree)</span><br><span class="line">age_dat[<span class="built_in">as.character</span>(age_dat$elements) == name_node_ind, ]</span><br><span class="line"><span class="comment">##      ages elements</span></span><br><span class="line"><span class="comment">## 37 99.706       37</span></span><br></pre></td></tr></table></figure>

<h1 id="更多内容"><a href="#更多内容" class="headerlink" title="更多内容"></a>更多内容</h1><ul>
<li>ggtree用于进化树的绘制 <a target="_blank" rel="noopener" href="https://yulab-smu.github.io/treedata-book/chapter5.html#visualizing-and-annotating-tree-using-grammar-of-graphics">https://yulab-smu.github.io/treedata-book/chapter5.html#visualizing-and-annotating-tree-using-grammar-of-graphics</a></li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2020/01/10/2020-01-10-mrca/" data-id="cl1n7qj9800e1eo84epwz2vc8" data-title="计算进化树末端节点出现的时间及两节点共同祖先（MRCA）的年龄" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2019-10-29-phylotoolsandsupermatrix" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/10/29/2019-10-29-phylotoolsandsupermatrix/" class="article-date">
  <time class="dt-published" datetime="2019-10-28T17:41:00.000Z" itemprop="datePublished">2019-10-29</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/10/29/2019-10-29-phylotoolsandsupermatrix/">用phylotools建立supermatrix</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>phylotools是用来建立DNA supermatrix矩阵的R程序包，是为了将比对好的基因文件（Fasta文件或Phylip文件）合并成超级矩阵（supermatrix）而编写的。</p>
<h2 id="Supermatrix与Relaxed-Phylip格式"><a href="#Supermatrix与Relaxed-Phylip格式" class="headerlink" title="Supermatrix与Relaxed Phylip格式"></a>Supermatrix与Relaxed Phylip格式</h2><p>supermatrix，就是将不同基因的片段分别比对后，再按照物种名合并形成的大型数据集，以便建立进化树用。</p>
<p>supermatrix一般为Relaxed Phylip格式。这是因为能够建立大型进化树的软件，如RAxML、Q-Tree等，都能识别这种格式。</p>
<p>Relaxed Phylip格式是Phylip格式的扩展，即第一行的两个数字分别声明序列数和位点数，两个数字中间有一个空格。第二行以序列名开始，序列名称和序列数据之间，有一个空格。传统的Phylip文件，序列名称只允许10个字符，如果超过10个字符，Phylip等软件在建树时会将名称截断，只取前10个字符。Relaxed Phylip格式则没有这种限制。</p>
<h2 id="建立supermatrix的几个常用片段"><a href="#建立supermatrix的几个常用片段" class="headerlink" title="建立supermatrix的几个常用片段"></a>建立supermatrix的几个常用片段</h2><p>为什么建立进化树要用多个DNA片段呢？这主要是因为DNA条形码的单个基因一般都太短，难以全面反映物种之间的关系，所以建立更为可信的进化树，一般需要更长的序列，即所谓多基因策略。</p>
<p>不同的序列，由于编码功能的差别，有些突变的速度快，有些速度慢。有些基因片段就较为稳定，在不同科属之间差异较小，而有些则变异很快，在同一个种不同个体之间都有很大的区别。在DNA条形码研究中，常用比较稳定的rbcLa基因确定植物物种的大致类群，一般可确定到科的水平；用matK等变异稍快的类群鉴定到属甚至到种的水平；用trnH-psbA等变异速率非常快的非编码基因确定更精确的种或种下的水平等。</p>
<p>上述三个DNA条形码序列，都是属于叶绿体基因组。当然，除上述基因外，还有很多别的基因，如ITS等也可用来建树。</p>
<p>DNA条形码序列一般是对PCR产物用传统的桑格测序法获得的，对样品和反应条件的要求较高，有一定的失败率。理想情况下， 一份样品的几个条形码序列都应该是完整的。但是由于各种各样的原因，有些种的某些序列总是不能正常扩增和测序，造成数据缺失。</p>
<h2 id="supermatrix的基本内容"><a href="#supermatrix的基本内容" class="headerlink" title="supermatrix的基本内容"></a>supermatrix的基本内容</h2><p>在建树开始之前，每个基因的序列需要单独比对，然后再拼接在一起。当然，这时候需要保证同一个种的不同序列要接在一起，但是位点的顺序不能错位。与此同时，若某一个种的某一个基因的序列缺失，则其对应的全部位点都需要用<code>?</code>补齐。</p>
<h2 id="类似软件"><a href="#类似软件" class="headerlink" title="类似软件"></a>类似软件</h2><p>用比对好的序列文件建立supermatrix，除了phylotools，还有以下软件：</p>
<ol>
<li>Biopython (<a target="_blank" rel="noopener" href="https://biopython.org/wiki/Concatenate_nexus">https://biopython.org/wiki/Concatenate_nexus</a>)</li>
<li>Utensils (<a target="_blank" rel="noopener" href="https://github.com/ballesterus/Utensils">https://github.com/ballesterus/Utensils</a>)</li>
<li>Geneious (<a target="_blank" rel="noopener" href="https://assets.geneious.com/manual/11.1/GeneiousManualsu53.html">https://assets.geneious.com/manual/11.1/GeneiousManualsu53.html</a>)</li>
<li>catfasta2phyml (<a target="_blank" rel="noopener" href="https://github.com/nylander/catfasta2phyml">https://github.com/nylander/catfasta2phyml</a>)</li>
<li>SEDA (<a target="_blank" rel="noopener" href="http://www.sing-group.org/seda/">http://www.sing-group.org/seda/</a> , <a target="_blank" rel="noopener" href="https://www.biostars.org/p/332853/">https://www.biostars.org/p/332853/</a>)</li>
<li>SuperMatrix: (<a target="_blank" rel="noopener" href="http://coleoguy.blogspot.com/2015/04/r-concatenate-alignments-into.html">http://coleoguy.blogspot.com/2015/04/r-concatenate-alignments-into.html</a>)</li>
<li>concat (<a target="_blank" rel="noopener" href="https://figshare.com/articles/concatenation_R/3839538">https://figshare.com/articles/concatenation_R/3839538</a>)</li>
<li>AMAS (<a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/pmc/articl::es/PMC4734057/">https://www.ncbi.nlm.nih.gov/pmc/articl::es/PMC4734057/</a>)</li>
</ol>
<p>其中Genious可以通过图形界面实现。</p>
<h2 id="Phylotools的特点"><a href="#Phylotools的特点" class="headerlink" title="Phylotools的特点"></a>Phylotools的特点</h2><ul>
<li>phylotools全部用R写成，操作简单。</li>
<li>phylotools假设同一个种，在不同的fasta或者phylip文件中有相同的名称。</li>
<li>phylotools的supermatrix可以识别比对好的fasta文件或者phylip文件，生成的supermatrix以Relaxed Phylip格式保存。</li>
</ul>
<h2 id="phylotools的安装"><a href="#phylotools的安装" class="headerlink" title="phylotools的安装"></a>phylotools的安装</h2><p>phylotools的最新版保存在github上，而并未上传到CRAN。要安装phylotools，可使用如下命令：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">devtools::install_github(&quot;helixcn&#x2F;phylotools&quot;)</span><br></pre></td></tr></table></figure>

<p>若未安装devtools，可使用 <code>install.packages(&quot;devtools&quot;)</code>安装</p>
<h2 id="phylotools程序包函数列表"><a href="#phylotools程序包函数列表" class="headerlink" title="phylotools程序包函数列表"></a>phylotools程序包函数列表</h2><ul>
<li><code>clean.fasta.name</code> 用于清除fasta序列名中的特殊字符，以便建树的时候能够正常输入和显示</li>
<li><code>dat2fasta</code> 将读取为data.frame的DNA序列转换为fasta文件</li>
<li><code>dat2phylip</code> 将读取为data.frame的DNA序列转换为phylip格式</li>
<li><code>get.fasta.name</code> 获取fasta文件中所有序列的名称</li>
<li><code>get.phylip.name</code> 获取phylip文件中所有序列的名称</li>
<li><code>read.fasta</code> 读取fasta文件</li>
<li><code>read.phylip</code> 读取phylip文件</li>
<li><code>rename.fasta</code> 重命名fasta文件</li>
<li><code>rm.sequence.fasta</code> 从fasta文件中删除某一条特定的序列</li>
<li><code>split_dat</code> 根据序列所在的分组，将数据分成独立的fasta文件，例如，可用于trnH-psbA序列按照目(orders)分组</li>
<li><code>sub.taxa.label</code> 替换进化树中的物种名</li>
<li><code>supermat</code> 用已经比对好的fasta文件或者phylip文件建立supermatrix超级矩阵以及RAxML分隔文件。</li>
</ul>
<h2 id="用phylip文件生成Relaxed-Phylip格式的supermatrix"><a href="#用phylip文件生成Relaxed-Phylip格式的supermatrix" class="headerlink" title="用phylip文件生成Relaxed Phylip格式的supermatrix"></a>用phylip文件生成Relaxed Phylip格式的supermatrix</h2><p>假设现有三个Phylip文件，分别编码不同的基因，建立者三个基因的supermatrix，并保存为Relaxed Phylip格式，同时生成RAxML Partition File，记录每个基因的起讫位点。</p>
<p>相应代码如下：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line">cat(<span class="string">&quot;6 22&quot;</span>,</span><br><span class="line"><span class="string">&quot;seq_1 --TTACAAATTGACTTATTATA&quot;</span>,</span><br><span class="line"><span class="string">&quot;seq_2 GATTACAAATTGACTTATTATA&quot;</span>,</span><br><span class="line"><span class="string">&quot;seq_3 GATTACAAATTGACTTATTATA&quot;</span>,</span><br><span class="line"><span class="string">&quot;seq_5 GATTACAAATTGACTTATTATA&quot;</span>,</span><br><span class="line"><span class="string">&quot;seq_8 GATTACAAATTGACTTATTATA&quot;</span>,</span><br><span class="line"><span class="string">&quot;seq_10---TACAAATTGAATTATTATA&quot;</span>,file = <span class="string">&quot;matk.phy&quot;</span>, sep = <span class="string">&quot;\n&quot;</span>)</span><br><span class="line"></span><br><span class="line">cat(<span class="string">&quot;5 15&quot;</span>,</span><br><span class="line"><span class="string">&quot;seq_1  GATTACAAATTGACT&quot;</span>,</span><br><span class="line"><span class="string">&quot;seq_3  GATTACAAATTGACT&quot;</span>,</span><br><span class="line"><span class="string">&quot;seq_4  GATTACAAATTGACT&quot;</span>,</span><br><span class="line"><span class="string">&quot;seq_5  GATTACAAATTGACT&quot;</span>,</span><br><span class="line"><span class="string">&quot;seq_8  GATTACAAATTGACT&quot;</span>,</span><br><span class="line">file = <span class="string">&quot;rbcla.phy&quot;</span>, sep = <span class="string">&quot;\n&quot;</span>)</span><br><span class="line"></span><br><span class="line">cat(<span class="string">&quot;5 50&quot;</span>,</span><br><span class="line"><span class="string">&quot;seq_2 GTCTTATAAGAAAGAATAAGAAAG--AAATACAAA-------AAAAAAGA&quot;</span>,</span><br><span class="line"><span class="string">&quot;seq_3 GTCTTATAAGAAAGAAATAGAAAAGTAAAAAAAAA-------AAAAAAAG&quot;</span>,</span><br><span class="line"><span class="string">&quot;seq_5 GACATAAGACATAAAATAGAATACTCAATCAGAAACCAACCCATAAAAAC&quot;</span>,</span><br><span class="line"><span class="string">&quot;seq_8 ATTCCAAAATAAAATACAAAAAGAAAAAACTAGAAAGTTTTTTTTCTTTG&quot;</span>,</span><br><span class="line"><span class="string">&quot;seq_9 ATTCTTTGTTCTTTTTTTTCTTTAATCTTTAAATAAACCTTTTTTTTTTA&quot;</span>,</span><br><span class="line">file = <span class="string">&quot;trn1.phy&quot;</span>, sep = <span class="string">&quot;\n&quot;</span>)</span><br><span class="line"></span><br><span class="line">supermat(infiles = <span class="built_in">c</span>(<span class="string">&quot;matk.phy&quot;</span>, <span class="string">&quot;rbcla.phy&quot;</span>, <span class="string">&quot;trn1.phy&quot;</span>))</span><br></pre></td></tr></table></figure>

<h2 id="用fasta文件生成supermatrix"><a href="#用fasta文件生成supermatrix" class="headerlink" title="用fasta文件生成supermatrix"></a>用fasta文件生成supermatrix</h2><p>要使用比对好的fasta文件生成supermatrix， 则命令改为对应的fasta文件名即可：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line">cat(<span class="string">&quot;&gt;seq_1&quot;</span>,  <span class="string">&quot;--TTACAAATTGACTTATTATA&quot;</span>,</span><br><span class="line"><span class="string">&quot;&gt;seq_2&quot;</span>,  <span class="string">&quot;GATTACAAATTGACTTATTATA&quot;</span>,</span><br><span class="line"><span class="string">&quot;&gt;seq_3&quot;</span>,  <span class="string">&quot;GATTACAAATTGACTTATTATA&quot;</span>,</span><br><span class="line"><span class="string">&quot;&gt;seq_5&quot;</span>,  <span class="string">&quot;GATTACAAATTGACTTATTATA&quot;</span>,</span><br><span class="line"><span class="string">&quot;&gt;seq_8&quot;</span>,  <span class="string">&quot;GATTACAAATTGACTTATTATA&quot;</span>,</span><br><span class="line"><span class="string">&quot;&gt;seq_10&quot;</span>, <span class="string">&quot;---TACAAATTGAATTATTATA&quot;</span>,</span><br><span class="line">file = <span class="string">&quot;matk.fasta&quot;</span>, sep = <span class="string">&quot;\n&quot;</span>)</span><br><span class="line"></span><br><span class="line">cat(</span><br><span class="line"><span class="string">&quot;&gt;seq_1&quot;</span>, <span class="string">&quot;GATTACAAATTGACT&quot;</span>,</span><br><span class="line"><span class="string">&quot;&gt;seq_3&quot;</span>, <span class="string">&quot;GATTACAAATTGACT&quot;</span>,</span><br><span class="line"><span class="string">&quot;&gt;seq_4&quot;</span>, <span class="string">&quot;GATTACAAATTGACT&quot;</span>,</span><br><span class="line"><span class="string">&quot;&gt;seq_5&quot;</span>, <span class="string">&quot;GATTACAAATTGACT&quot;</span>,</span><br><span class="line"><span class="string">&quot;&gt;seq_8&quot;</span>, <span class="string">&quot;GATTACAAATTGACT&quot;</span>,</span><br><span class="line">file = <span class="string">&quot;rbcla.fasta&quot;</span>, sep = <span class="string">&quot;\n&quot;</span>)</span><br><span class="line"></span><br><span class="line">cat(</span><br><span class="line"><span class="string">&quot;&gt;seq_2&quot;</span>, <span class="string">&quot;GTCTTATAAGAAAGAATAAGAAAG--AAATACAAA-------AAAAAAGA&quot;</span>,</span><br><span class="line"><span class="string">&quot;&gt;seq_3&quot;</span>, <span class="string">&quot;GTCTTATAAGAAAGAAATAGAAAAGTAAAAAAAAA-------AAAAAAAG&quot;</span>,</span><br><span class="line"><span class="string">&quot;&gt;seq_5&quot;</span>, <span class="string">&quot;GACATAAGACATAAAATAGAATACTCAATCAGAAACCAACCCATAAAAAC&quot;</span>,</span><br><span class="line"><span class="string">&quot;&gt;seq_8&quot;</span>, <span class="string">&quot;ATTCCAAAATAAAATACAAAAAGAAAAAACTAGAAAGTTTTTTTTCTTTG&quot;</span>,</span><br><span class="line"><span class="string">&quot;&gt;seq_9&quot;</span>, <span class="string">&quot;ATTCTTTGTTCTTTTTTTTCTTTAATCTTTAAATAAACCTTTTTTTTTTA&quot;</span>,</span><br><span class="line">file = <span class="string">&quot;trn1.fasta&quot;</span>, sep = <span class="string">&quot;\n&quot;</span>)</span><br><span class="line"></span><br><span class="line">supermat(infiles = <span class="built_in">c</span>(<span class="string">&quot;matk.fasta&quot;</span>, <span class="string">&quot;rbcla.fasta&quot;</span>, <span class="string">&quot;trn1.fasta&quot;</span>))</span><br></pre></td></tr></table></figure>

<h2 id="参考文献"><a href="#参考文献" class="headerlink" title="参考文献"></a>参考文献</h2><ul>
<li>Kress, W. J., Erickson, D. L., Jones, F. A., Swenson, N. G., Perez, R., Sanjur, O., &amp; Bermingham, E. (2009). Plant DNA barcodes and a community phylogeny of a tropical forest dynamics plot in Panama. Proceedings of the National Academy of Sciences, 106(44), 18621-18626.</li>
<li>裴男才, 张金龙, 米湘成, 葛学军 (2011) 植物DNA条形码促进系统发育群落生态学发展, 生物多样性, 19(3):284–294</li>
<li>Roquet, C., Thuiller, W., &amp; Lavergne, S. (2013). Building megaphylogenies for macroecology: taking up the challenge. Ecography, 36(1), 13-26.</li>
<li><a target="_blank" rel="noopener" href="https://cme.h-its.org/exelixis/web/software/raxml/index.html">https://cme.h-its.org/exelixis/web/software/raxml/index.html</a></li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2019/10/29/2019-10-29-phylotoolsandsupermatrix/" data-id="cl1n7qj9100dpeo84340ya45t" data-title="用phylotools建立supermatrix" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2019-10-21-samplingefficiency" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/10/21/2019-10-21-samplingefficiency/" class="article-date">
  <time class="dt-published" datetime="2019-10-20T16:32:00.000Z" itemprop="datePublished">2019-10-21</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/10/21/2019-10-21-samplingefficiency/">怎样确定一个地区标本采集的完整程度</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="问题："><a href="#问题：" class="headerlink" title="问题："></a>问题：</h1><p>现有若干县的植物标本记录，即每个种在该县境内采集的标本数量，求这些县植物标本采集的完整程度。</p>
<h1 id="解答："><a href="#解答：" class="headerlink" title="解答："></a>解答：</h1><p>标本采集的完整程度可以用随标本数量增加，物种累计数的快慢表示。</p>
<p>假设A县采集了5000份标本，而这5000份标本总共记录了500个种；B县采集了5000份标本，而这5000份标本包括了2000个种，很显然，A县的标本采集是比较充分的；相比之下，B县就很不充足，因为B县境内，物种数仍然随着标本数快速增加。</p>
<p>那么如何度量物种累计的快慢呢？中科院植物所的阳文静博士（Yang et al. 2013）在度量中国各县植物标本采集完整程度时，采用了物种累计数达到90%时的斜率。也就是计算标本数达到90%-100%时物种累计曲线的平均斜率。该思路简化下来，就是先计算标本数达到90%时对应的物种数，然后取平均值，然后将所有标本数对应的总物种数作为另外一个坐标点，两点确定一条直线，确定这条直线的斜率（原文信息不够详尽，可能本人理解有误）。</p>
<p>另一种思路是计算物种数达到90%时的切线斜率。由于是随机抽样，物种数从&lt;90%到跨越90%的一瞬间，物种数会增加，随机抽样时，物种数会围绕某个值上下波动，因此这里可以计算多次，取平均值。</p>
<p>这里只给出第二种思路的R代码，即求切线的斜率。</p>
<p><img src="https://mmbiz.qpic.cn/mmbiz_jpg/TmqE3hm8mvYhRDJjEiaq9QeayvNYz3ibkFEia6gpCqZSdkqVzw3tnqSpSpEPw4rqb5DibpKMRTkYzt2RmX1qP5lUug/640?wx_fmt=jpeg&tp=webp&wxfrom=5&wx_lazy=1&wx_co=1" alt="图片"></p>
<p> 图1. 要读取的标本记录格式</p>
<p><img src="https://mmbiz.qpic.cn/mmbiz_jpg/TmqE3hm8mvYhRDJjEiaq9QeayvNYz3ibkF95p5q8ebkIAYldFbNibRk3RWs1p0TOps9nNk8bS1YqiaiaUs11WZ36Tsw/640?wx_fmt=jpeg&tp=webp&wxfrom=5&wx_lazy=1&wx_co=1" alt="图片"></p>
<p>\2. 阳文静（Yang et al., 2013）中国各县植物标本采集完整程度</p>
<p><img src="https://mmbiz.qpic.cn/mmbiz_jpg/TmqE3hm8mvYhRDJjEiaq9QeayvNYz3ibkF98X5r9X5Cuspz3dFw2bnOibz4oCavSowEsOWfTvvRDYOK3l9jHG03gw/640?wx_fmt=jpeg&tp=webp&wxfrom=5&wx_lazy=1&wx_co=1" alt="图片"></p>
<p><img src="https://mmbiz.qpic.cn/mmbiz_jpg/TmqE3hm8mvYhRDJjEiaq9QeayvNYz3ibkFhSGQqgDNM0zvmaCcu3CAdAmUs3pJuiaooorDIZAr3TYGW539Y37lYmA/640?wx_fmt=jpeg&tp=webp&wxfrom=5&wx_lazy=1&wx_co=1" alt="图片"></p>
<p>图3 物种数累计曲线：标本采集完整的县斜率小，标本采集不完整的县斜率大</p>
<p>R脚本</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br></pre></td><td class="code"><pre><span class="line">setwd(<span class="string">&quot;C:/Users/helixcn/Desktop/completeness&quot;</span>)</span><br><span class="line">library(openxlsx)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 获得某个文件夹的所有xlsx文件字符串向量</span></span><br><span class="line">all_files &lt;- list.files()</span><br><span class="line">excel_files &lt;- all_files[grepl(<span class="string">&quot;.xlsx&quot;</span>, all_files)]</span><br><span class="line"></span><br><span class="line"><span class="comment"># 针对每个标本数量，例如从30000个标本中，随机抽26000个标本，这里要随机抽取100次， nperm就是控制随机抽取的数量</span></span><br><span class="line">nperm = <span class="number">100</span></span><br><span class="line"></span><br><span class="line">slope_90_100 &lt;- <span class="literal">NA</span> <span class="comment"># 初始化, slope_90_100 保存每个文件（也就是每个地点的slope值）</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># k为excel文件的数量</span></span><br><span class="line"><span class="keyword">for</span>(k <span class="keyword">in</span> <span class="number">1</span>:<span class="built_in">length</span>(excel_files))&#123;</span><br><span class="line">    </span><br><span class="line">    <span class="comment"># 读取一个excel文件</span></span><br><span class="line">    specimen_names &lt;- read.xlsx(excel_files[k])[,<span class="number">1</span>]</span><br><span class="line">    </span><br><span class="line">    <span class="comment"># 求物种数</span></span><br><span class="line">    nspecies &lt;- <span class="built_in">length</span>(unique(specimen_names))</span><br><span class="line">    </span><br><span class="line">    <span class="comment"># 物种数不能少于10</span></span><br><span class="line">    <span class="keyword">if</span>(nspecies &lt; <span class="number">10</span>)&#123;</span><br><span class="line">        stop(<span class="string">&quot;Number of species should be more than 10&quot;</span>)</span><br><span class="line">    &#125;</span><br><span class="line">    </span><br><span class="line">    <span class="comment"># 标本数不能少于50</span></span><br><span class="line">    <span class="keyword">if</span>(<span class="built_in">length</span>(specimen_names) &lt; <span class="number">50</span>)&#123;</span><br><span class="line">        stop(<span class="string">&quot;Number of specimens should be more than 50&quot;</span>)</span><br><span class="line">    &#125;</span><br><span class="line">    </span><br><span class="line">    <span class="comment"># 90%的物种数</span></span><br><span class="line">    nspecies_90 &lt;- nspecies * <span class="number">0.9</span></span><br><span class="line">    </span><br><span class="line">    <span class="comment">## 为了减少计算量，这里只计算物种数达到90%以后的物种累计曲线</span></span><br><span class="line">    <span class="comment">## 因此，先要计算在标本达到总标本数的多少时（求一个比例ratio），物种数不超过90%</span></span><br><span class="line"></span><br><span class="line">    ratio = <span class="number">1.0</span> <span class="comment"># 设定ratio的初始值，然后猜100次，一直到标本数刚刚少于90%的物种数为止。每一步递减0.01，猜100次（nguess），足以覆盖所有的情况</span></span><br><span class="line">    nguess = <span class="number">100</span></span><br><span class="line">    <span class="keyword">for</span>(m <span class="keyword">in</span> <span class="number">1</span>:nguess)&#123;</span><br><span class="line">        <span class="keyword">if</span>(<span class="built_in">length</span>(unique(sample(specimen_names, <span class="built_in">ceiling</span>(<span class="built_in">length</span>(specimen_names) * ratio)))) &gt; nspecies_90) &#123;</span><br><span class="line">             ratio &lt;- ratio - <span class="number">0.01</span></span><br><span class="line">         &#125; <span class="keyword">else</span> &#123;</span><br><span class="line">             ratio &lt;- ratio</span><br><span class="line">         &#125;</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    n_species_90_to_100 &lt;- <span class="literal">NA</span></span><br><span class="line">    <span class="comment"># 从能达到90%物种数的标本比例开始计算随标本数增加的物种累计曲线</span></span><br><span class="line">    <span class="keyword">for</span> (j <span class="keyword">in</span> <span class="built_in">floor</span>(<span class="built_in">length</span>(specimen_names) * ratio):<span class="built_in">length</span>(specimen_names))&#123;</span><br><span class="line">        n_species_90_to_100_i &lt;- <span class="literal">NA</span> <span class="comment"># 初始化随机抽取100次的平均物种数</span></span><br><span class="line">        <span class="keyword">for</span> (i <span class="keyword">in</span> <span class="number">1</span>:nperm)&#123;</span><br><span class="line">            n_species_90_to_100_i &lt;- <span class="built_in">c</span>(n_species_90_to_100_i, <span class="built_in">length</span>(unique(sample(specimen_names, size = j))))</span><br><span class="line">        &#125;</span><br><span class="line">        n_species_90_to_100 &lt;- <span class="built_in">c</span>(n_species_90_to_100, mean(na.omit(n_species_90_to_100_i)))  <span class="comment">#</span></span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="comment"># 初始化的时候，NA要去掉</span></span><br><span class="line">    n_species_90_to_100_clean &lt;- na.omit(n_species_90_to_100)</span><br><span class="line"></span><br><span class="line">    <span class="comment"># 只取物种数大于百分之九十物种数时的情况</span></span><br><span class="line">    n_species_90_to_100_clean &lt;- n_species_90_to_100_clean[n_species_90_to_100_clean &gt; nspecies_90]</span><br><span class="line"></span><br><span class="line">    <span class="comment"># 为了求得每增加一份标本，平均增加多少种，要将所得结果错位，然后相减</span></span><br><span class="line">    par1 &lt;- <span class="built_in">c</span>(<span class="literal">NA</span>, n_species_90_to_100_clean)</span><br><span class="line">    par2 &lt;- <span class="built_in">c</span>(n_species_90_to_100_clean, <span class="literal">NA</span>)</span><br><span class="line"></span><br><span class="line">    <span class="comment"># 相减之前，要先去掉NA，否则平均值也是NA</span></span><br><span class="line">    <span class="comment"># slope_90_100为一个向量，从NA开始，长度不断增加，注意slope_90_100在循环外面已经定义了</span></span><br><span class="line">    <span class="comment"># mean(na.omit(c(par2 - par1))) 就是从物种数达到90%之后，每增加一份标本，平均增加的物种数</span></span><br><span class="line"></span><br><span class="line">    slope_90_100 &lt;- <span class="built_in">c</span>(slope_90_100, mean(na.omit(<span class="built_in">c</span>(par2 - par1)))) <span class="comment"># 求斜率，以总标本数为基准</span></span><br><span class="line"></span><br><span class="line">    <span class="comment"># 显示每个文件，slope的值</span></span><br><span class="line">    print(paste(<span class="string">&quot;The average slope (90-100%) for file&quot;</span>, excel_files[k] ,<span class="string">&quot;is:&quot;</span>, mean(na.omit(<span class="built_in">c</span>(par2 - par1)))))</span><br><span class="line">&#125;</span><br><span class="line"><span class="comment"># 将所有文件的slope，保存到一个CSV中，同时提供文件名</span></span><br><span class="line">write.csv(data.frame(site = excel_files, slope_90_100 = na.omit(slope_90_100)), <span class="string">&quot;survey_completeness.csv&quot;</span>)</span><br></pre></td></tr></table></figure>

<h1 id="致谢"><a href="#致谢" class="headerlink" title="致谢"></a>致谢</h1><ul>
<li>感谢中科院植物所刘慧圆老师一起讨论问题</li>
</ul>
<h1 id="参考文献"><a href="#参考文献" class="headerlink" title="参考文献"></a>参考文献</h1><ul>
<li>Yang, W., Ma, K., &amp; Kreft, H. (2013). Geographical sampling bias in a large distributional database and its effects on species richness-environment models. Journal of Biogeography, 40(8), 1415–1426. <a target="_blank" rel="noopener" href="https://doi.org/10.1111/jbi.12108">https://doi.org/10.1111/jbi.12108</a></li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2019/10/21/2019-10-21-samplingefficiency/" data-id="cl1n7qj9000dneo84cucx6b3m" data-title="怎样确定一个地区标本采集的完整程度" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2019-10-10-DNA" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/10/10/2019-10-10-DNA/" class="article-date">
  <time class="dt-published" datetime="2019-10-09T18:26:02.000Z" itemprop="datePublished">2019-10-10</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/10/10/2019-10-10-DNA/">从DNA序列到系统发育多样性-在环保部环科院的讲座</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>讲座内容包括：</p>
<p>1 进化树及相关基本概念<br>2 DNA 序列的获取和比对<br>3 进化模型及其筛选<br>4 建立进化树的方法<br>5 启发式搜索和 bootstrap 可靠性评估<br>6 贝叶斯法建树<br>7 分化时间的校对：分子钟<br>8 通过植物学名建立进化树<br>9 系统发育多样性分析： ape、 vegan 和 picante</p>
<p>本讲座是2019年9月16日应环保部环境科学研究院的刘勇波副研究员（<a target="_blank" rel="noopener" href="http://www.craes.cn/zjhky/craes_kydw/sssds/201808/t20180819_454216.shtml%EF%BC%89%E7%9A%84%E9%82%80%E8%AF%B7%E8%80%8C%E5%87%86%E5%A4%87%E7%9A%84%EF%BC%8C%E5%9C%A8%E6%AD%A4%E7%89%B9%E5%88%AB%E6%84%9F%E8%B0%A2%E5%88%98%E5%B8%88%E5%85%84%E7%9A%84%E9%82%80%E8%AF%B7%E5%92%8C%E7%83%AD%E6%83%85%E6%AC%BE%E5%BE%85%E3%80%82">http://www.craes.cn/zjhky/craes_kydw/sssds/201808/t20180819_454216.shtml）的邀请而准备的，在此特别感谢刘师兄的邀请和热情款待。</a></p>
<p>特别感谢沈泽昊、张英、杨拓、高芳銮、向小果、刘勇波、俞方圆、曾思金等老师提出宝贵意见！</p>
<p>更多内容敬请关注 <a target="_blank" rel="noopener" href="http://blog.sciencenet.cn/u/zjlcas">http://blog.sciencenet.cn/u/zjlcas</a> 和 <a target="_blank" rel="noopener" href="https://github.com/helixcn">https://github.com/helixcn</a> 。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2019/10/10/2019-10-10-DNA/" data-id="cl1n7qj8y00dleo843br79x4w" data-title="从DNA序列到系统发育多样性-在环保部环科院的讲座" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2019-09-21-greatcircle" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/09/21/2019-09-21-greatcircle/" class="article-date">
  <time class="dt-published" datetime="2019-09-20T22:52:54.000Z" itemprop="datePublished">2019-09-21</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/09/21/2019-09-21-greatcircle/">已知经纬度计算两点间地理距离</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="1-问题"><a href="#1-问题" class="headerlink" title="1 问题"></a>1 问题</h1><p>已知巴黎的地理坐标为东经2°20’14’’，北纬48°50’11’’，华盛顿的地理坐标为西经 77°03’56’’，北纬 38°55’17’’，求两城间的距离（单位千米）。</p>
<h1 id="2-分析："><a href="#2-分析：" class="headerlink" title="2 分析："></a>2 分析：</h1><p>已知经纬度，求两点间的地理距离，不能直接用经纬度求平面距离，而是要计算球面距离（Great Circle Distance）。</p>
<p>如果对精度要求不高，可以用低精度公式，计算过程相对简单。如果对经度要求较高， 则要用高精度公式。高精度公式一般用于天文计算、大地测量等。在距离较远的城市之间， 两种方法所得结果可相差几十千米。</p>
<p>当然，现在全部通过计算机程序计算，都是直接使用高精度方法了。本文参考比利时天文学家Jean Meeus (1991)的《天文算法》81-86页的公式，给出相应的R代码。</p>
<h2 id="2-1-将经纬度的度、分、秒转换为十进制"><a href="#2-1-将经纬度的度、分、秒转换为十进制" class="headerlink" title="2.1 将经纬度的度、分、秒转换为十进制"></a>2.1 将经纬度的度、分、秒转换为十进制</h2><p>定义函数deg2dec完成这一转换:</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">deg2dec &lt;- <span class="keyword">function</span>(h,m,s)&#123;   </span><br><span class="line">   <span class="keyword">if</span>(h &lt; <span class="number">0</span>)&#123;</span><br><span class="line">       m = - m</span><br><span class="line">       s = -s</span><br><span class="line">   &#125;</span><br><span class="line">   res = h + m/<span class="number">60</span> + s/<span class="number">3600</span></span><br><span class="line">   <span class="built_in">return</span>(res)</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="comment">## 巴黎</span></span><br><span class="line">L1 = deg2dec(-<span class="number">2</span>,<span class="number">20</span>,<span class="number">14</span>)</span><br><span class="line">phi1 = deg2dec(<span class="number">48</span>, <span class="number">50</span>, <span class="number">11</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">## 华盛顿</span></span><br><span class="line">L2 = deg2dec(<span class="number">77</span>,<span class="number">03</span>,<span class="number">56</span>)</span><br><span class="line">phi2 = deg2dec(<span class="number">38</span>,<span class="number">55</span>,<span class="number">17</span>)</span><br></pre></td></tr></table></figure>

<h2 id="2-2-低精度公式"><a href="#2-2-低精度公式" class="headerlink" title="2.2 低精度公式"></a>2.2 低精度公式</h2><p>由于地球的扁率较小，可近似看作球体，利用球面三角公式即可求出距离，R代码如下:</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">geodist &lt;- <span class="keyword">function</span>(L1, phi1, L2, phi2)&#123;</span><br><span class="line">    Ri = <span class="number">6371</span></span><br><span class="line">    d = (Ri * <span class="built_in">acos</span>(<span class="built_in">sin</span>(phi1 * (<span class="built_in">pi</span> / <span class="number">180</span>)) *</span><br><span class="line">                      <span class="built_in">sin</span>(phi2 * (<span class="built_in">pi</span> / <span class="number">180</span>)) +</span><br><span class="line">                      <span class="built_in">cos</span>(phi1 * (<span class="built_in">pi</span> / <span class="number">180</span>)) *</span><br><span class="line">                      <span class="built_in">cos</span>(phi2 * (<span class="built_in">pi</span> / <span class="number">180</span>)) *</span><br><span class="line">                      <span class="built_in">cos</span>((L1 - L2) * (<span class="built_in">pi</span> / <span class="number">180</span>))))</span><br><span class="line">    res &lt;- <span class="built_in">round</span>(d, <span class="number">3</span>)    </span><br><span class="line">    <span class="built_in">return</span>(res)</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">geodist(L1, phi1, L2, phi2)</span><br><span class="line"><span class="comment">## [1] 6165.597</span></span><br></pre></td></tr></table></figure>

<p>因此，如果假设地球是球体，华盛顿和巴黎之间的距离是6165.597km</p>
<h2 id="2-3-高精度公式"><a href="#2-3-高精度公式" class="headerlink" title="2.3 高精度公式"></a>2.3 高精度公式</h2><p>假设地球是椭球，a是长轴，f是扁率，公式如下：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line">hageodist &lt;- <span class="keyword">function</span>(L1, phi1, L2, phi2)&#123;</span><br><span class="line">   a = <span class="number">6378.14</span></span><br><span class="line">   f = <span class="number">1</span>/<span class="number">298.257</span></span><br><span class="line">   <span class="built_in">F</span> = (phi1 + phi2)/<span class="number">2</span></span><br><span class="line">   G = (phi1 - phi2)/<span class="number">2</span></span><br><span class="line">   ramda &lt;- (L1 - L2)/<span class="number">2</span></span><br><span class="line">   S = (<span class="built_in">sin</span>(G * <span class="built_in">pi</span> / <span class="number">180</span>)^<span class="number">2</span>)*(<span class="built_in">cos</span>(ramda * <span class="built_in">pi</span> / <span class="number">180</span>)^<span class="number">2</span>) + </span><br><span class="line">       (<span class="built_in">cos</span>(<span class="built_in">F</span> * <span class="built_in">pi</span> / <span class="number">180</span>)^<span class="number">2</span>)*(<span class="built_in">sin</span>(ramda * <span class="built_in">pi</span> / <span class="number">180</span>)^<span class="number">2</span>) </span><br><span class="line">   C = (<span class="built_in">cos</span>(G * <span class="built_in">pi</span> / <span class="number">180</span>)^<span class="number">2</span>)*(<span class="built_in">cos</span>(ramda * <span class="built_in">pi</span> / <span class="number">180</span>)^<span class="number">2</span>) + </span><br><span class="line">       (<span class="built_in">sin</span>(<span class="built_in">F</span> * <span class="built_in">pi</span> / <span class="number">180</span>)^<span class="number">2</span>)*(<span class="built_in">sin</span>(ramda * <span class="built_in">pi</span> / <span class="number">180</span>)^<span class="number">2</span>) </span><br><span class="line">   omega = <span class="built_in">atan</span>(<span class="built_in">sqrt</span>(S / C))</span><br><span class="line">   R = <span class="built_in">sqrt</span>(S * C)/omega</span><br><span class="line">   D = <span class="number">2</span> * omega * a</span><br><span class="line">   H1 = (<span class="number">3</span> * R - <span class="number">1</span>)/(<span class="number">2</span> * C)</span><br><span class="line">   H2 = (<span class="number">3</span> * R + <span class="number">1</span>)/(<span class="number">2</span> * S)</span><br><span class="line">   res = D * (<span class="number">1</span> + f * H1 * (<span class="built_in">sin</span>(<span class="built_in">F</span> * <span class="built_in">pi</span> / <span class="number">180</span>)^<span class="number">2</span>)*</span><br><span class="line">               (<span class="built_in">cos</span>(G * <span class="built_in">pi</span> / <span class="number">180</span>)^<span class="number">2</span>) -</span><br><span class="line">               f * H2 * (<span class="built_in">cos</span>(<span class="built_in">F</span> * <span class="built_in">pi</span>/<span class="number">180</span>)^<span class="number">2</span>)*</span><br><span class="line">               (<span class="built_in">sin</span>(G * <span class="built_in">pi</span> / <span class="number">180</span>)^<span class="number">2</span>))   </span><br><span class="line">   <span class="built_in">return</span>(<span class="built_in">round</span>(res, <span class="number">3</span>))</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">hageodist(L1, phi1, L2, phi2)</span><br><span class="line"><span class="comment">## [1] 6181.628</span></span><br></pre></td></tr></table></figure>

<p>因此，华盛顿和巴黎之间的距离是6181.628km。高精度和低精度公式的结果相差16km。</p>
<h2 id="2-4-注意事项"><a href="#2-4-注意事项" class="headerlink" title="2.4 注意事项"></a>2.4 注意事项</h2><p>Meeus这本书中，东经为负，西经为正，公式采用国际天文学联合会IAU1976规定的相关参数，其坐标系未指定。</p>
<p>注意，Meeus的算法中，经度与常用的WGS84坐标系正好相反。在WGS84坐标系中，东经为正，西经为负。现在通常所说的经纬度是用GPS获得的WGS84经纬度，所以最保险的方法是以下R脚本计算：</p>
<h1 id="3-推荐的R程序包和脚本"><a href="#3-推荐的R程序包和脚本" class="headerlink" title="3 推荐的R程序包和脚本"></a>3 推荐的R程序包和脚本</h1><h2 id="3-1-sp包"><a href="#3-1-sp包" class="headerlink" title="3.1 sp包"></a>3.1 sp包</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">library(sp)</span><br><span class="line">longitude &lt;- <span class="built_in">as.numeric</span>(<span class="built_in">c</span>(L1, L2))</span><br><span class="line">latitude  &lt;- <span class="built_in">as.numeric</span>(<span class="built_in">c</span>(phi1, phi2))</span><br><span class="line">cities &lt;- <span class="built_in">c</span>(<span class="string">&quot;Paris&quot;</span>, <span class="string">&quot;Washington&quot;</span>)</span><br><span class="line">location &lt;- cbind(longitude = -longitude, latitude)</span><br><span class="line">row.names(location) &lt;- cities</span><br><span class="line">location &lt;- data.frame(location)</span><br><span class="line">coordinates(location) &lt;- ~longitude+latitude</span><br><span class="line">proj4string(location) &lt;- CRS(<span class="string">&quot;+proj=longlat +datum=WGS84&quot;</span>) <span class="comment"># 明确使用WGS84坐标系</span></span><br><span class="line">spDists(location)</span><br><span class="line"><span class="comment">##          [,1]     [,2]</span></span><br><span class="line"><span class="comment">## [1,]    0.000 6181.626</span></span><br><span class="line"><span class="comment">## [2,] 6181.626    0.000</span></span><br></pre></td></tr></table></figure>

<h2 id="3-2-geosphere包"><a href="#3-2-geosphere包" class="headerlink" title="3.2 geosphere包"></a>3.2 geosphere包</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">xy &lt;- rbind(paris =<span class="built_in">c</span>(-L1, phi1), washington =<span class="built_in">c</span>(-L2, phi2))</span><br><span class="line">library(geosphere)</span><br><span class="line">distm(xy)</span><br><span class="line"><span class="comment">##         [,1]    [,2]</span></span><br><span class="line"><span class="comment">## [1,]       0 6181632</span></span><br><span class="line"><span class="comment">## [2,] 6181632       0</span></span><br></pre></td></tr></table></figure>

<h1 id="4-在地图上绘制Great-Circle-Distance"><a href="#4-在地图上绘制Great-Circle-Distance" class="headerlink" title="4 在地图上绘制Great Circle Distance"></a>4 在地图上绘制Great Circle Distance</h1><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">library(maps)</span><br><span class="line">library(geosphere)</span><br><span class="line">map(<span class="string">&#x27;world&#x27;</span>,col=<span class="string">&quot;grey&quot;</span>, fill=<span class="literal">FALSE</span>, bg=<span class="string">&quot;white&quot;</span>, lwd=<span class="number">0.05</span>,interior = <span class="literal">FALSE</span>)</span><br><span class="line">points(x = -L1, y = phi1, pch = <span class="number">19</span>)</span><br><span class="line">points(x = -L2, y = phi2, pch = <span class="number">19</span>)</span><br><span class="line">text(x = -L1, y = phi1 + <span class="number">5</span>, labels = <span class="string">&quot;Paris&quot;</span>, col = <span class="string">&quot;blue&quot;</span>)</span><br><span class="line">text(x = -L2, y = phi2 + <span class="number">5</span>, labels = <span class="string">&quot;Washington&quot;</span>, col = <span class="string">&quot;blue&quot;</span>)</span><br><span class="line"></span><br><span class="line">connection &lt;- gcIntermediate(<span class="built_in">c</span>(-L1, phi1), <span class="built_in">c</span>(-L2, phi2), </span><br><span class="line">    n=<span class="number">100</span>, addStartEnd = <span class="literal">TRUE</span>, breakAtDateLine=<span class="literal">FALSE</span>)             </span><br><span class="line">lines(connection, col=<span class="string">&quot;red&quot;</span>, lwd=<span class="number">2</span>)</span><br></pre></td></tr></table></figure>

<h1 id="5-参考"><a href="#5-参考" class="headerlink" title="5 参考"></a>5 参考</h1><ul>
<li><a target="_blank" rel="noopener" href="https://www.r-graph-gallery.com/how-to-draw-connecting-routes-on-map-with-r-and-great-circles/">https://www.r-graph-gallery.com/how-to-draw-connecting-routes-on-map-with-r-and-great-circles/</a></li>
<li><a target="_blank" rel="noopener" href="https://www.jessesadler.com/post/great-circles-sp-sf/">https://www.jessesadler.com/post/great-circles-sp-sf/</a></li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2019/09/21/2019-09-21-greatcircle/" data-id="cl1n7qj8x00djeo84ejg9gtc2" data-title="已知经纬度计算两点间地理距离" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E5%9C%B0%E7%90%86/" rel="tag">地理</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2019-7-29-icecream" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/07/29/2019-7-29-icecream/" class="article-date">
  <time class="dt-published" datetime="2019-07-28T16:45:02.000Z" itemprop="datePublished">2019-07-29</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/07/29/2019-7-29-icecream/">冰糕、奶糕</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>华北，6月中旬，深蓝的天空万里无云。盛夏的太阳散发着威力，炙烤着金黄色的麦田。干热风吹过，麦穗沙沙作响。此时，麦穗早已又干又脆，轻轻一碰，麦粒就会掉在地上。</p>
<p>晌午，割麦子的人挥汗如雨。镰刀轻轻往回一收，麦杆就清脆地断开了。每割一下，地上就升起一丝尘土，仿佛一缕淡淡的青烟。每呼吸一下，干燥的空气就带走嗓子里的一点水分，仿佛要把肺抽干。塑料水壶里的水早已被烤得发烫，喝上几大口，也不觉得多解渴。</p>
<p>远远的，有一个小白点来了。借着麦浪，声音传得更远：“冰糕、奶糕、刨冰……”：一个不到二十岁的姑娘，推着一辆飞鸽牌二十八加重自行车，车后座挂着两个冰棍儿箱子。这是二十多年前，麦收时候的一天中午。</p>
<p>卖冰棍儿姑娘的出现，比凉爽的风更让人激动。</p>
<p>掀开冰棍儿箱的盖子，一股凉气让拿冰棍儿的手别提多舒服了。冰棍儿上面盖着一个白色的小棉被。掀开被子，冰棍儿、雪糕整齐码着：这边是奶油，那边是小豆，还有巧克力的，“俩色俩味”的……另一个箱子里是“刨冰”，坚韧的塑料袋里冻着糖水，这个“刨冰”也很受欢迎。</p>
<p>干渴至极的时候能吃到冰棍儿，让人难掩内心的激动。割麦子的人赶紧掏出两元钱，冰棍儿两毛，刨冰五毛，不仅仅是给一起下地干活的孩子，自己也要赶紧吃上两块消暑。</p>
<p>如果天气不热，大部分人是不舍得买冰棍儿吃的。只有在烈日下，冰棍儿的销路才好。今天生意确实不错。</p>
<p>姑娘扶了扶草帽。汗水从额头流下来，沾湿了头发，她赶忙擦了擦汗。口渴难耐，她赶紧打开水壶，灌了几口早上在家里罐好的井水。“的确良”衬衫也早已湿透。还好，还差几块钱，本钱就上来了，这样一来，剩下的二十几只冰棍儿就全是今天赚的，也许能挣上十几块。想到此处，姑娘心里宽慰了许多。</p>
<p>冰棍儿都是一早上从二十多里路以外的县城趸回来的。天蒙蒙亮就要出发，那还是早上四点多。</p>
<p>姑娘从小到大一直在学校读书：家里人都觉得读书是改变命运的唯一希望，所以一直全力支持。只是姑娘高考发挥失常，没有考上大学，落榜在家。</p>
<p>没有什么选择，只有安心务农，早早找户人家嫁了。出嫁之前要努力挣点儿钱，给自己攒点儿嫁妆。</p>
<p>之前，她没受过多少日晒雨淋，都一直是在教室里学习，听老师讲语、数、外、史、地、生，做各种练习题，参加各种测试，对未来充满憧憬，从未想过自己要去卖冰棍儿。</p>
<p>住校的几年，她一直都是吃家里带的咸菜，营养不良，身体瘦弱。自行车上挂上两个空冰棍儿箱子，她觉得很不好控制，特别是上下车，右腿都要跨过车的大梁，很不习惯。</p>
<p>从家里出发一个小时，终于骑到了冰棍儿厂了。天早已大亮，太阳升起来，照着稚嫩的脸。前面还有很多人在往箱子里装冰棍，趸冰棍儿要挨个儿。好不容易交上钱，装好箱，两个箱子忽然仿佛装满了两大桶水，沉重异常，车子就更难控制。</p>
<p>来到田间小路，她不敢再骑，生怕撞到树上或者拐进沟里，瘦小的她只能慢慢向前推着车子，边走，边吆喝，“冰糕、奶糕、刨冰……”。她知道，只有最暴晒的麦田里，冰棍儿才有人买。就这样，她在烈日下转了几个钟头，早已有点儿头晕目眩，中午也只是啃了两口凉馒头就咸菜。</p>
<p>太阳慢慢转到西边了，没那么热了，现在还有十几根冰棍儿没有卖出去。虽然天气不太热了，但是箱子里开始暖和起来，冰棍儿变得很软，马上就要化了。</p>
<p>她也赶紧回到村里，将软到快要拿不起来的贱卖处理，原来五毛钱一块的，现在一块钱给三块、四块。还好，最终只是还剩几块在箱子里融化了，箱子里，糖水上漂着几根木棍儿……辛苦了一整天，挣了几块钱，姑娘心里说不出的滋味。</p>
<p>天快黑了，她回到家。</p>
<p>父母没有说她。父亲沉默不语，在一旁抽着烟。母亲帮她把箱子擦干净。</p>
<p>晚饭很简单，炒土豆，煮凉汤（过水面）。她吃了几口，眼泪落下来，然后独自回房间了。一个高考落榜姑娘的一天就这样结束了。</p>
<p>其实，这里的姑娘，跟我高考落榜的姑姑所经历的，并没有什么区别。我两个姑姑在高中毕业后，在九十年代都卖过冰棍儿。二十多年后的今天，我还能想起姑姑冰棍儿没有卖完，回来后一家人一起吃冰棍儿的情景。</p>
<p>行走在烈日下，我经常想起麦田边上那个由远及近的小白点，一个瘦弱的姑娘，推着自行车，后面挂着冰棍儿箱子，不停吆喝着：“冰糕、奶糕、刨冰……”</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2019/07/29/2019-7-29-icecream/" data-id="cl1n7qj9700dzeo840fjk5vu8" data-title="冰糕、奶糕" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2019-7-18-scientist" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/07/18/2019-7-18-scientist/" class="article-date">
  <time class="dt-published" datetime="2019-07-17T17:01:02.000Z" itemprop="datePublished">2019-07-18</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/07/18/2019-7-18-scientist/">描写一流科学家需要怎样的写作技巧？《张益唐：天才的野心》读后感</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>写文章无外乎要传递信息，本文要传播什么信息呢？那就是张益唐这样一位天才，与常人有何不同，他过着怎样的生活。作者文笔一流，用对比、烘托等笔法，写出了张益唐如神仙般的存在：一位孤独、谦逊、深邃的谦谦君子。</p>
<p>本文的核心思想正如标题所讲，要突出的是张益唐诗意一样的野心：专注于大问题，专注到不管庭前花开花落，达到宠辱不惊的程度——除了纯粹数学，其他都完全不在乎。</p>
<p>如果成功解决这些“经典问题”，数学家本人将名垂青史，但如果没有成功解决，就什么都不存在。这是一种非常危险的研究方式，绝大多数学者都不敢选择，但是张益唐这么做了，而且做到了。作者抓住了这一点，将张益唐的经历写得十分精彩。</p>
<p>在作者笔下，张不是刻意忽略名利与社会地位，而是完全脱离现实中的名利，只有无时无刻对自己感兴趣数学问题的不停思考。换句话说，是为了学术而学术，完全过着一种理想主义的生活。张教授以一种现代学术界无法接受的方式存在着，也就是“不按套路出牌”。</p>
<p>一般来说，无论你从事那一领域的研究，最少在几年时间，要有若干成果，也就是有足够数量的论文发表，否则是无法获得项目资助的，业界也认为你从此沉沦。如果几年不发表一篇论文，一般业界会认为你的职业生涯已经结束。</p>
<p>当代学术界的大多数论文，读者只有从事最相近研究领域的同行。而外界看一个学者的能力和重要性，多多少少都只看对方有多少第一作者的文章，看论文发表在什么刊物上，看谷歌学术的引用次数，甚至是ResearchGate分数等。在这种情况下，学术贡献本身是用一个数字衡量的，至于工作是否实现了某些方面从零到一的突破，很多情况下直接用上述“加减法”去衡量，当然是十分荒谬的，但这是目前学界广泛认同的评价体系，绝大部分人不得不接受这种评价方式。</p>
<p>张益唐当然是另类中的另类，假如真如文章所写，几十年只发表三篇论文，是绝对不可能在学术机构里面获得广泛认可的，因为从论文数量上看，根本不可能达到评教授的要求。在国内目前的学术环境下，也更不可能容忍，恐怕也早已被扫地出门。</p>
<p>本文的作者大量运用了对比的手法，突出了张益唐过往经历的跌宕起伏与与众不同。“文似看山不喜平”，将这种反差写得淋漓尽致，因此读者才愿意看下去。</p>
<p>为了营造氛围，同时将人物写得生动，作者同时提供了大量细节，并引用张益唐及其亲友的话，不断强化上述反差（衣食住行等等方面，都写得很详细，让人觉得可信）。最后，作者再通过学生和朋友的第三方描述，让人物的形象更丰满。</p>
<p>所以说，本文作者用很纯熟的笔法，把张益唐刻画得很生动、性格很鲜明，也更深入人心，从而让一位儒雅的大学者形象跃然纸上。</p>
<p>这也是需要很长时间训练才能达到的写作技巧。</p>
<h1 id="进一步阅读"><a href="#进一步阅读" class="headerlink" title="进一步阅读"></a>进一步阅读</h1><p>《张益唐：天才的野心》原文链接： <a target="_blank" rel="noopener" href="https://mp.weixin.qq.com/s/7cCpCAwqjIYUsARWSQZrlw">https://mp.weixin.qq.com/s/7cCpCAwqjIYUsARWSQZrlw</a></p>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2019/07/18/2019-7-18-scientist/" data-id="cl1n7qj9500dveo84a6ez6qjt" data-title="描写一流科学家需要怎样的写作技巧？《张益唐：天才的野心》读后感" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2019-7-10-beta" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2019/07/10/2019-7-10-beta/" class="article-date">
  <time class="dt-published" datetime="2019-07-09T17:13:02.000Z" itemprop="datePublished">2019-07-10</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2019/07/10/2019-7-10-beta/">一个群落beta多样性分析的完整R脚本</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>本文给出Abbas S. et al. (2019)的完整R脚本，内容包括群落稀疏化曲线、beta多样性和指示种分析、mantel检验、NMDS非参数排序、空间关系的PCNM转换和前向筛选、方差分解等各种分析以及作图等，供感兴趣的同行参考。</p>
<p>本R脚本全部为本人编写，因群落数据本人无权公开，这里只公布R脚本。篇幅所限，部分数据分析段落有删节。如您有任何问题或者建议，请跟本人联系（微信 <code>jinlongzhang01</code>）。</p>
<p>因水平有限，脚本中难免出现错误，欢迎各位读者指正。</p>
<p>R脚本各部分内容如下：</p>
<ul>
<li>1 稀疏化曲线，用于比较个体数不同样方的物种丰富度</li>
<li>2 Mantel检验两个距离矩阵的相关性是否显著</li>
<li>3 指示种分析、PCNM、环境因子的前向筛选、物种组成的方差分解以及显著性检验</li>
<li>4 betadisper检验beta多样性各组的离散度是否有统计差异，Mantel Correlogram展示距离矩阵相关的精细结构</li>
<li>5 分析不同演替阶段生活型组成的变化</li>
<li>6 提取丰富度最高的种</li>
<li>7 NMDS分析，用以展示不同演替阶段样方物种组成相似性的关系</li>
</ul>
<p>下载完整PDF文档 <a target="_blank" rel="noopener" href="http://blog.sciencenet.cn/home.php?mod=attachment&id=398724">res.pdf</a>。PDF文件中部分代码语句没有换行，造成显示不完整，关注本人微信公众号ecoinformatics（二维码见本文最后），直接查看html版本的代码。</p>
<p>参考文献：</p>
<ul>
<li>Abbas, S., Nichol, J. E., Zhang, J., &amp; Fischer, G. A. (2019). The accumulation of species and recovery of species composition along a 70-year succession in a tropical secondary forest. Ecological Indicators, 106, 105524. DOI:10.1016/j.ecolind.2019.105524</li>
</ul>
<p><a target="_blank" rel="noopener" href="http://image.sciencenet.cn/home/201907/10/090330os7n7uskrvkx713q.png"><img src="http://image.sciencenet.cn/home/201907/10/090330os7n7uskrvkx713q.png" alt="img"></a></p>
<p><a target="_blank" rel="noopener" href="http://image.sciencenet.cn/home/201907/10/090332rxw5bb3xms3wfmas.png"><img src="http://image.sciencenet.cn/home/201907/10/090332rxw5bb3xms3wfmas.png" alt="img"></a></p>
<p><a target="_blank" rel="noopener" href="http://image.sciencenet.cn/home/201907/10/090340h7wqpw3yz39psxz9.png"><img src="http://image.sciencenet.cn/home/201907/10/090340h7wqpw3yz39psxz9.png" alt="img"></a></p>
<p><a target="_blank" rel="noopener" href="http://image.sciencenet.cn/home/201907/10/090349czeaigzna0e04tg4.png"><img src="http://image.sciencenet.cn/home/201907/10/090349czeaigzna0e04tg4.png" alt="img"></a></p>
<p><a target="_blank" rel="noopener" href="http://image.sciencenet.cn/home/201907/10/090353qxlm5xwl91w5hic1.png"><img src="http://image.sciencenet.cn/home/201907/10/090353qxlm5xwl91w5hic1.png" alt="img"></a></p>
<p><a target="_blank" rel="noopener" href="http://image.sciencenet.cn/home/201907/10/090357uud8lehd43tg9b43.png"><img src="http://image.sciencenet.cn/home/201907/10/090357uud8lehd43tg9b43.png" alt="img"></a></p>
<p><a target="_blank" rel="noopener" href="http://image.sciencenet.cn/home/201907/10/090400hr1vr3wxwo3f4q33.png"><img src="http://image.sciencenet.cn/home/201907/10/090400hr1vr3wxwo3f4q33.png" alt="img"></a></p>
<p><a target="_blank" rel="noopener" href="http://image.sciencenet.cn/home/201907/10/090405wtpj7ozoj34vz7v1.png"><img src="http://image.sciencenet.cn/home/201907/10/090405wtpj7ozoj34vz7v1.png" alt="img"></a></p>
<p><a target="_blank" rel="noopener" href="http://image.sciencenet.cn/home/201907/10/090408hcuvmvm7sl5ommm5.png"><img src="http://image.sciencenet.cn/home/201907/10/090408hcuvmvm7sl5ommm5.png" alt="img"></a></p>
<p><a target="_blank" rel="noopener" href="http://image.sciencenet.cn/home/201907/10/090412uw6vlfvwqrwlaw7l.png"><img src="http://image.sciencenet.cn/home/201907/10/090412uw6vlfvwqrwlaw7l.png" alt="img"></a></p>
<p><a target="_blank" rel="noopener" href="http://image.sciencenet.cn/home/201907/10/090414by1yguy1uu64uxhl.png"><img src="http://image.sciencenet.cn/home/201907/10/090414by1yguy1uu64uxhl.png" alt="img"></a></p>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2019/07/10/2019-7-10-beta/" data-id="cl1n7qj9600dxeo849614c7ts" data-title="一个群落beta多样性分析的完整R脚本" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>

    </footer>
  </div>
  
</article>



  


  <nav id="page-nav">
    
    <a class="extend prev" rel="prev" href="/page/2/">&laquo; Prev</a><a class="page-number" href="/">1</a><a class="page-number" href="/page/2/">2</a><span class="page-number current">3</span><a class="page-number" href="/page/4/">4</a><a class="page-number" href="/page/5/">5</a><span class="space">&hellip;</span><a class="page-number" href="/page/26/">26</a><a class="extend next" rel="next" href="/page/4/">Next &raquo;</a>
  </nav>

</section>
        
          <aside id="sidebar">
  
    

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">Tags</h3>
    <div class="widget">
      <ul class="tag-list" itemprop="keywords"><li class="tag-list-item"><a class="tag-list-link" href="/tags/Python/" rel="tag">Python</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/R%E7%BC%96%E7%A8%8B/" rel="tag">R编程</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E4%BA%BA%E7%89%A9/" rel="tag">人物</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%86%99%E4%BD%9C/" rel="tag">写作</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" rel="tag">可视化</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9B%BE%E4%B9%A6/" rel="tag">图书</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9B%BE%E5%83%8F%E5%A4%84%E7%90%86/" rel="tag">图像处理</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9C%B0%E7%90%86/" rel="tag">地理</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%A3%B0%E6%98%8E/" rel="tag">声明</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%A4%A9%E6%96%87/" rel="tag">天文</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%AD%A6%E4%B9%A0%E8%B5%84%E6%BA%90/" rel="tag">学习资源</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%AF%B9%E8%AF%9D%E5%BD%95/" rel="tag">对话录</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%BF%83%E5%BE%97/" rel="tag">心得</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%8A%A5%E5%91%8A/" rel="tag">报告</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%8E%92%E7%89%88/" rel="tag">排版</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%85%E4%B9%A1/" rel="tag">故乡</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E5%BA%93/" rel="tag">数据库</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%A0%87%E6%9C%AC%E9%A6%86/" rel="tag">标本馆</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%A4%8D%E7%89%A9%E5%88%86%E7%B1%BB/" rel="tag">植物分类</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%A4%8D%E7%89%A9%E5%AD%A6%E5%90%8D/" rel="tag">植物学名</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B1%82%E8%81%8C/" rel="tag">求职</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%89%88%E6%9C%AC%E6%8E%A7%E5%88%B6/" rel="tag">版本控制</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%94%9F%E6%80%81/" rel="tag">生态</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%94%9F%E6%80%81%E8%B5%84%E8%AE%AF/" rel="tag">生态资讯</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%94%9F%E6%B4%BB%E7%82%B9%E6%BB%B4/" rel="tag">生活点滴</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%A0%94%E7%A9%B6%E8%B5%84%E8%AE%AF/" rel="tag">研究资讯</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" rel="tag">系统发育</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%9F%E8%AE%A1/" rel="tag">统计</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BC%96%E7%A8%8B/" rel="tag">编程</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%A1%8C%E6%91%84%E5%8C%86%E5%8C%86/" rel="tag">行摄匆匆</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%A7%81%E9%97%BB%E5%BD%95/" rel="tag">见闻录</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%AF%97%E6%AD%8C/" rel="tag">诗歌</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6/" rel="tag">软件</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BF%9B%E5%8C%96/" rel="tag">进化</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%9A%8F%E6%83%B3/" rel="tag">随想</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%A9%B3%E8%AE%BA/" rel="tag">驳论</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">Tag Cloud</h3>
    <div class="widget tagcloud">
      <a href="/tags/Python/" style="font-size: 10px;">Python</a> <a href="/tags/R%E7%BC%96%E7%A8%8B/" style="font-size: 18.75px;">R编程</a> <a href="/tags/%E4%BA%BA%E7%89%A9/" style="font-size: 11.88px;">人物</a> <a href="/tags/%E5%86%99%E4%BD%9C/" style="font-size: 13.13px;">写作</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" style="font-size: 12.5px;">可视化</a> <a href="/tags/%E5%9B%BE%E4%B9%A6/" style="font-size: 10px;">图书</a> <a href="/tags/%E5%9B%BE%E5%83%8F%E5%A4%84%E7%90%86/" style="font-size: 11.88px;">图像处理</a> <a href="/tags/%E5%9C%B0%E7%90%86/" style="font-size: 17.5px;">地理</a> <a href="/tags/%E5%A3%B0%E6%98%8E/" style="font-size: 10.63px;">声明</a> <a href="/tags/%E5%A4%A9%E6%96%87/" style="font-size: 16.25px;">天文</a> <a href="/tags/%E5%AD%A6%E4%B9%A0%E8%B5%84%E6%BA%90/" style="font-size: 10px;">学习资源</a> <a href="/tags/%E5%AF%B9%E8%AF%9D%E5%BD%95/" style="font-size: 10px;">对话录</a> <a href="/tags/%E5%BF%83%E5%BE%97/" style="font-size: 10px;">心得</a> <a href="/tags/%E6%8A%A5%E5%91%8A/" style="font-size: 12.5px;">报告</a> <a href="/tags/%E6%8E%92%E7%89%88/" style="font-size: 13.75px;">排版</a> <a href="/tags/%E6%95%85%E4%B9%A1/" style="font-size: 10.63px;">故乡</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E5%BA%93/" style="font-size: 10.63px;">数据库</a> <a href="/tags/%E6%A0%87%E6%9C%AC%E9%A6%86/" style="font-size: 14.38px;">标本馆</a> <a href="/tags/%E6%A4%8D%E7%89%A9%E5%88%86%E7%B1%BB/" style="font-size: 10px;">植物分类</a> <a href="/tags/%E6%A4%8D%E7%89%A9%E5%AD%A6%E5%90%8D/" style="font-size: 10px;">植物学名</a> <a href="/tags/%E6%B1%82%E8%81%8C/" style="font-size: 10px;">求职</a> <a href="/tags/%E7%89%88%E6%9C%AC%E6%8E%A7%E5%88%B6/" style="font-size: 11.25px;">版本控制</a> <a href="/tags/%E7%94%9F%E6%80%81/" style="font-size: 16.88px;">生态</a> <a href="/tags/%E7%94%9F%E6%80%81%E8%B5%84%E8%AE%AF/" style="font-size: 10px;">生态资讯</a> <a href="/tags/%E7%94%9F%E6%B4%BB%E7%82%B9%E6%BB%B4/" style="font-size: 10px;">生活点滴</a> <a href="/tags/%E7%A0%94%E7%A9%B6%E8%B5%84%E8%AE%AF/" style="font-size: 10px;">研究资讯</a> <a href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" style="font-size: 10px;">系统发育</a> <a href="/tags/%E7%BB%9F%E8%AE%A1/" style="font-size: 18.13px;">统计</a> <a href="/tags/%E7%BC%96%E7%A8%8B/" style="font-size: 10.63px;">编程</a> <a href="/tags/%E8%A1%8C%E6%91%84%E5%8C%86%E5%8C%86/" style="font-size: 15px;">行摄匆匆</a> <a href="/tags/%E8%A7%81%E9%97%BB%E5%BD%95/" style="font-size: 10px;">见闻录</a> <a href="/tags/%E8%AF%97%E6%AD%8C/" style="font-size: 15.63px;">诗歌</a> <a href="/tags/%E8%BD%AF%E4%BB%B6/" style="font-size: 19.38px;">软件</a> <a href="/tags/%E8%BF%9B%E5%8C%96/" style="font-size: 10.63px;">进化</a> <a href="/tags/%E9%9A%8F%E6%83%B3/" style="font-size: 11.25px;">随想</a> <a href="/tags/%E9%9A%8F%E7%AC%94/" style="font-size: 20px;">随笔</a> <a href="/tags/%E9%A9%B3%E8%AE%BA/" style="font-size: 10px;">驳论</a>
    </div>
  </div>

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">Archives</h3>
    <div class="widget">
      <ul class="archive-list"><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/04/">April 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/03/">March 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/01/">January 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/11/">November 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/09/">September 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/07/">July 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/05/">May 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/03/">March 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/02/">February 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/12/">December 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/11/">November 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/06/">June 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/05/">May 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/02/">February 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/01/">January 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/10/">October 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/09/">September 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/07/">July 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/06/">June 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/05/">May 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/04/">April 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/03/">March 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/02/">February 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/01/">January 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/12/">December 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/11/">November 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/10/">October 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/09/">September 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/08/">August 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/07/">July 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/06/">June 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/05/">May 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/04/">April 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/03/">March 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/02/">February 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/01/">January 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/11/">November 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/10/">October 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/09/">September 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/08/">August 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/06/">June 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/05/">May 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/04/">April 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/02/">February 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/12/">December 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/11/">November 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/09/">September 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/06/">June 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/05/">May 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/04/">April 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/03/">March 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/01/">January 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/11/">November 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/10/">October 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/09/">September 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/07/">July 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/05/">May 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/04/">April 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/03/">March 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/02/">February 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/01/">January 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/12/">December 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/11/">November 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/10/">October 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/09/">September 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/07/">July 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/05/">May 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/03/">March 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/02/">February 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/01/">January 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/12/">December 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/10/">October 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/05/">May 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/04/">April 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/03/">March 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/02/">February 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/01/">January 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/12/">December 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/10/">October 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/08/">August 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/07/">July 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/06/">June 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/05/">May 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/04/">April 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/02/">February 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/01/">January 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/12/">December 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/11/">November 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/10/">October 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/08/">August 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/07/">July 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/05/">May 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/04/">April 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/03/">March 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/02/">February 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/12/">December 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/11/">November 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/09/">September 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/08/">August 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/07/">July 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/06/">June 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/05/">May 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/04/">April 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/03/">March 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/01/">January 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2009/08/">August 2009</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2009/07/">July 2009</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2009/06/">June 2009</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2009/05/">May 2009</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">Recent Posts</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/2022/04/01/2022-03-31run_phylomatic/">(no title)</a>
          </li>
        
          <li>
            <a href="/2022/03/31/2022-03-31run_phylomatic/">在本地运行awk版本的phylomatic-awk利用植物名录生成进化树</a>
          </li>
        
          <li>
            <a href="/2022/01/15/2022-01-15dangxia/">当下</a>
          </li>
        
          <li>
            <a href="/2021/11/05/2021-11-05lamtsuenriver/">林村河</a>
          </li>
        
          <li>
            <a href="/2021/09/13/2021-10-19school/">最初的小学时光</a>
          </li>
        
      </ul>
    </div>
  </div>

  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      
      &copy; 2022 Jinlong Zhang<br>
      Powered by <a href="https://hexo.io/" target="_blank">Hexo</a>
    </div>
  </div>
</footer>

    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">Home</a>
  
    <a href="/archives" class="mobile-nav-link">Archives</a>
  
    <a href="/about" class="mobile-nav-link">About</a>
  
</nav>
    


<script src="/js/jquery-3.4.1.min.js"></script>



  
<script src="/fancybox/jquery.fancybox.min.js"></script>




<script src="/js/script.js"></script>





  </div>
</body>
</html>