<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  
  <title>Jinlong Zhang&#39;s blog</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta property="og:type" content="website">
<meta property="og:title" content="Jinlong Zhang&#39;s blog">
<meta property="og:url" content="https://helixcn.github.io/page/10/index.html">
<meta property="og:site_name" content="Jinlong Zhang&#39;s blog">
<meta property="og:locale" content="en_US">
<meta property="article:author" content="Jinlong Zhang">
<meta name="twitter:card" content="summary">
  
    <link rel="alternate" href="/atom.xml" title="Jinlong Zhang's blog" type="application/atom+xml">
  
  
    <link rel="shortcut icon" href="/favicon.png">
  
  
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-source-code-pro@0.0.71/index.min.css">

  
  
<link rel="stylesheet" href="/css/style.css">

  
    
<link rel="stylesheet" href="/fancybox/jquery.fancybox.min.css">

  
<meta name="generator" content="Hexo 5.4.0"></head>

<body>
  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">Jinlong Zhang&#39;s blog</a>
      </h1>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"></a>
        
          <a class="main-nav-link" href="/">Home</a>
        
          <a class="main-nav-link" href="/archives">Archives</a>
        
          <a class="main-nav-link" href="/about">About</a>
        
      </nav>
      <nav id="sub-nav">
        
          <a id="nav-rss-link" class="nav-icon" href="/atom.xml" title="RSS Feed"></a>
        
        <a id="nav-search-btn" class="nav-icon" title="Search"></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="Search"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="https://helixcn.github.io"></form>
      </div>
    </div>
  </div>
</header>

      <div class="outer">
        <section id="main">
  
    <article id="post-2017-10-27-janzen" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2017/10/27/2017-10-27-janzen/" class="article-date">
  <time class="dt-published" datetime="2017-10-26T16:54:54.000Z" itemprop="datePublished">2017-10-27</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2017/10/27/2017-10-27-janzen/">幼苗生存状态、负密度制约与Janzen Connell假说的检验</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>检验同种和不同种在群落内的竞争效应是群落生态学的重要研究内容。 负密度制约一般表现在同种个体或者异种个体在空间距离越近，对资源的竞争也就越激烈。与此同时，相同种在空间上的聚集，容易导致疾病传播，容易被捕食或者取食，对该种的幼苗生长不利。这种情况也发生在幼苗与母树之间，生态学家提出距离母树更近的幼苗死亡率更高，这就是Janzen Connell假说，并认为这是群落内物种多样性维持的重要机制之一。近年来，对于Janzen-Connell的研究也不断增加。</p>
<p>检验同一物种以及不同物种之间的密度制约现象， 一般是用到幼苗的两次调查数据， 用广义线性混合模型去探讨各因子的相对贡献。 响应变量一般为幼苗在两次调查时的生存状态。 解释变量一般为同种和异种幼苗在任何一棵幼苗内的胸径截面积，幼苗胸径、高度等。并将样方号、物种等并不太关心的因子作为随机效应处理。 根据研究问题和研究尺度的不同，也可能也会考虑环境因子的贡献等， 不过由于很难直接获得每棵幼苗对应的环境因子，所以整合环境因子的密度制约分析还不多。</p>
<p>在本文先随机生成一套样地数据， 然后给出一个函数compute_neighbour_area，该函数能计算指定半径下，每个个体相同种和不同树种的胸径截面积，生成的数据可以直接用来拟合广义线性混合模型，以探讨同种和不同种的竞争效应，探讨密度制约机制。</p>
<p>模拟生成的练习数据要求如下，面积为5公顷，长250m， 宽200m， 有90个物种， 12700个个体， 详情参见下面的R代码和注释 。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br><span class="line">133</span><br><span class="line">134</span><br><span class="line">135</span><br><span class="line">136</span><br><span class="line">137</span><br><span class="line">138</span><br><span class="line">139</span><br><span class="line">140</span><br><span class="line">141</span><br><span class="line">142</span><br><span class="line">143</span><br><span class="line">144</span><br><span class="line">145</span><br><span class="line">146</span><br><span class="line">147</span><br><span class="line">148</span><br><span class="line">149</span><br><span class="line">150</span><br><span class="line">151</span><br><span class="line">152</span><br><span class="line">153</span><br><span class="line">154</span><br><span class="line">155</span><br><span class="line">156</span><br><span class="line">157</span><br><span class="line">158</span><br><span class="line">159</span><br><span class="line">160</span><br><span class="line">161</span><br><span class="line">162</span><br><span class="line">163</span><br><span class="line">164</span><br><span class="line">165</span><br><span class="line">166</span><br><span class="line">167</span><br><span class="line">168</span><br><span class="line">169</span><br><span class="line">170</span><br><span class="line">171</span><br><span class="line">172</span><br><span class="line">173</span><br><span class="line">174</span><br><span class="line">175</span><br><span class="line">176</span><br><span class="line">177</span><br><span class="line">178</span><br><span class="line">179</span><br><span class="line">180</span><br><span class="line">181</span><br><span class="line">182</span><br><span class="line">183</span><br><span class="line">184</span><br><span class="line">185</span><br><span class="line">186</span><br><span class="line">187</span><br><span class="line">188</span><br><span class="line">189</span><br><span class="line">190</span><br><span class="line">191</span><br><span class="line">192</span><br><span class="line">193</span><br><span class="line">194</span><br><span class="line">195</span><br><span class="line">196</span><br><span class="line">197</span><br><span class="line">198</span><br><span class="line">199</span><br><span class="line">200</span><br><span class="line">201</span><br><span class="line">202</span><br><span class="line">203</span><br><span class="line">204</span><br><span class="line">205</span><br><span class="line">206</span><br><span class="line">207</span><br><span class="line">208</span><br><span class="line">209</span><br><span class="line">210</span><br><span class="line">211</span><br><span class="line">212</span><br><span class="line">213</span><br><span class="line">214</span><br><span class="line">215</span><br><span class="line">216</span><br><span class="line">217</span><br><span class="line">218</span><br><span class="line">219</span><br><span class="line">220</span><br><span class="line">221</span><br><span class="line">222</span><br><span class="line">223</span><br><span class="line">224</span><br><span class="line">225</span><br><span class="line">226</span><br><span class="line">227</span><br><span class="line">228</span><br><span class="line">229</span><br><span class="line">230</span><br><span class="line">231</span><br><span class="line">232</span><br><span class="line">233</span><br><span class="line">234</span><br><span class="line">235</span><br><span class="line">236</span><br><span class="line">237</span><br><span class="line">238</span><br><span class="line">239</span><br><span class="line">240</span><br><span class="line">241</span><br><span class="line">242</span><br><span class="line">243</span><br><span class="line">244</span><br><span class="line">245</span><br><span class="line">246</span><br><span class="line">247</span><br><span class="line">248</span><br><span class="line">249</span><br><span class="line">250</span><br><span class="line">251</span><br><span class="line">252</span><br><span class="line">253</span><br><span class="line">254</span><br><span class="line">255</span><br><span class="line">256</span><br><span class="line">257</span><br><span class="line">258</span><br><span class="line">259</span><br><span class="line">260</span><br><span class="line">261</span><br><span class="line">262</span><br><span class="line">263</span><br><span class="line">264</span><br><span class="line">265</span><br><span class="line">266</span><br><span class="line">267</span><br><span class="line">268</span><br><span class="line">269</span><br><span class="line">270</span><br><span class="line">271</span><br><span class="line">272</span><br><span class="line">273</span><br><span class="line">274</span><br><span class="line">275</span><br><span class="line">276</span><br><span class="line">277</span><br><span class="line">278</span><br><span class="line">279</span><br><span class="line">280</span><br><span class="line">281</span><br><span class="line">282</span><br><span class="line">283</span><br><span class="line">284</span><br><span class="line">285</span><br><span class="line">286</span><br><span class="line">287</span><br><span class="line">288</span><br><span class="line">289</span><br><span class="line">290</span><br><span class="line">291</span><br><span class="line">292</span><br><span class="line">293</span><br><span class="line">294</span><br><span class="line">295</span><br><span class="line">296</span><br><span class="line">297</span><br><span class="line">298</span><br><span class="line">299</span><br><span class="line">300</span><br><span class="line">301</span><br><span class="line">302</span><br><span class="line">303</span><br><span class="line">304</span><br><span class="line">305</span><br><span class="line">306</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#### 假设样地有如下90个物种 （参考 陈云等 2017）</span></span><br><span class="line"></span><br><span class="line">sp &lt;- <span class="built_in">c</span>(<span class="string">&quot;Abelia biflora&quot;</span>, </span><br><span class="line"><span class="string">&quot;Abelia uniflora&quot;</span>,<span class="string">&quot;Acer davidii&quot;</span>,<span class="string">&quot;Acer davidii subsp. grosseri&quot;</span>,<span class="string">&quot;Acer griseum&quot;</span>,</span><br><span class="line"><span class="string">&quot;Acer pictum subsp. mono&quot;</span>,<span class="string">&quot;Acer truncatum&quot;</span>,<span class="string">&quot;Ailanthus altissima&quot;</span>,<span class="string">&quot;Aralia elata&quot;</span>,</span><br><span class="line"><span class="string">&quot;Berberis circumserrata&quot;</span>,<span class="string">&quot;Berberis mitifolia&quot;</span>,<span class="string">&quot;Betula chinensis&quot;</span>,<span class="string">&quot;Betula luminifera&quot;</span>,</span><br><span class="line"><span class="string">&quot;Betula platyphylla&quot;</span>,<span class="string">&quot;Buckleya graebneriana&quot;</span>,<span class="string">&quot;Carpinus cordata&quot;</span>,<span class="string">&quot;Carpinus turczaninowii&quot;</span>,</span><br><span class="line"><span class="string">&quot;Cerasus clarofolia&quot;</span>,<span class="string">&quot;Cerasus polytricha&quot;</span>,<span class="string">&quot;Cerasus serrulata&quot;</span>,<span class="string">&quot;Cercidiphyllum japonicum&quot;</span>,</span><br><span class="line"><span class="string">&quot;Cornus controversa&quot;</span>,<span class="string">&quot;Cornus kousa subsp. chinensis&quot;</span>,<span class="string">&quot;Cornus macrophylla&quot;</span>,</span><br><span class="line"><span class="string">&quot;Cornus schindleri subsp. poliophylla&quot;</span>,<span class="string">&quot;Cornus walteri&quot;</span>,<span class="string">&quot;Corylus chinensis&quot;</span>,</span><br><span class="line"><span class="string">&quot;Corylus heterophylla&quot;</span>,<span class="string">&quot;Cotoneaster acutifolius&quot;</span>,<span class="string">&quot;Cotoneaster multiflorus&quot;</span>,</span><br><span class="line"><span class="string">&quot;Crataegus hupehensis&quot;</span>,<span class="string">&quot;Crataegus pinnatifida&quot;</span>,<span class="string">&quot;Deutzia parviflora&quot;</span>,</span><br><span class="line"><span class="string">&quot;Diospyros lotus&quot;</span>,<span class="string">&quot;Elaeagnus umbellata&quot;</span>,<span class="string">&quot;Euonymus maackii&quot;</span>,</span><br><span class="line"><span class="string">&quot;Euonymus phellomanus&quot;</span>,<span class="string">&quot;Euonymus schensianus&quot;</span>,<span class="string">&quot;Euptelea pleiosperma&quot;</span>,</span><br><span class="line"><span class="string">&quot;Evodia daniellii&quot;</span>,<span class="string">&quot;Evodia fargesii&quot;</span>,<span class="string">&quot;Forsythia suspensa&quot;</span>,<span class="string">&quot;Fraxinus chinensis&quot;</span>,</span><br><span class="line"><span class="string">&quot;Fraxinus mandshurica&quot;</span>,<span class="string">&quot;Fraxinus paxiana&quot;</span>,<span class="string">&quot;Juglans cathayensis&quot;</span>,</span><br><span class="line"><span class="string">&quot;Koelreuteria paniculata&quot;</span>,<span class="string">&quot;Lespedeza bicolor&quot;</span>,<span class="string">&quot;Lespedeza buergeri&quot;</span>,</span><br><span class="line"><span class="string">&quot;Lindera obtusiloba&quot;</span>,<span class="string">&quot;Litsea tsinlingensis&quot;</span>,<span class="string">&quot;Lonicera hispida&quot;</span>,</span><br><span class="line"><span class="string">&quot;Lonicera microphylla&quot;</span>,<span class="string">&quot;Lonicera tatarinowii&quot;</span>,<span class="string">&quot;Maddenia hypoleuca&quot;</span>,</span><br><span class="line"><span class="string">&quot;Malus baccata&quot;</span>,<span class="string">&quot;Malus honanensis&quot;</span>,<span class="string">&quot;Malus hupehensis&quot;</span>,<span class="string">&quot;Malus kansuensis&quot;</span>,</span><br><span class="line"><span class="string">&quot;Meliosma flexuosa&quot;</span>,<span class="string">&quot;Meliosma veitchiorum&quot;</span>,<span class="string">&quot;Ostrya japonica&quot;</span>,<span class="string">&quot;Padus buergeriana&quot;</span>,</span><br><span class="line"><span class="string">&quot;Philadelphus incanus&quot;</span>,<span class="string">&quot;Pinus armandii&quot;</span>,<span class="string">&quot;Pinus tabuliformis&quot;</span>,<span class="string">&quot;Populus davidiana&quot;</span>,</span><br><span class="line"><span class="string">&quot;Quercus aliena var. acutiserrata&quot;</span>,<span class="string">&quot;Quercus serrata var. brevipetiolata&quot;</span>,</span><br><span class="line"><span class="string">&quot;Quercus variabilis&quot;</span>,<span class="string">&quot;Rhododendron micranthum&quot;</span>,<span class="string">&quot;Ribes pachysandroides&quot;</span>,</span><br><span class="line"><span class="string">&quot;Salix floderusii&quot;</span>,<span class="string">&quot;Sambucus williamsii&quot;</span>,<span class="string">&quot;Sinowilsonia henryi&quot;</span>,</span><br><span class="line"><span class="string">&quot;Sorbaria sorbifolia&quot;</span>,<span class="string">&quot;Sorbus alnifolia&quot;</span>,<span class="string">&quot;Sorbus hupehensis&quot;</span>,<span class="string">&quot;Spiraea hirsuta&quot;</span>,</span><br><span class="line"><span class="string">&quot;Stachyurus chinensis&quot;</span>,<span class="string">&quot;Styrax hemsleyanus&quot;</span>,<span class="string">&quot;Styrax obassis&quot;</span>,<span class="string">&quot;Symplocos paniculata&quot;</span>,</span><br><span class="line"><span class="string">&quot;Tilia japonica&quot;</span>,<span class="string">&quot;Tilia paucicostata&quot;</span>,<span class="string">&quot;Toxicodendron vernicifluum&quot;</span>,<span class="string">&quot;Ulmus davidiana&quot;</span>,</span><br><span class="line"><span class="string">&quot;Viburnum betulifolium&quot;</span>,<span class="string">&quot;Viburnum opulus var. sargentii&quot;</span>,<span class="string">&quot;Yulania denudata&quot;</span>)</span><br><span class="line"></span><br><span class="line">set.seed(<span class="number">12345</span>)  <span class="comment">#### 设定随机数种子</span></span><br><span class="line">library(sads)    <span class="comment">#### 导入物种多度分布程序包</span></span><br><span class="line"><span class="comment">## Loading required package: bbmle</span></span><br><span class="line"><span class="comment">## Loading required package: stats4</span></span><br><span class="line">library(stringr) <span class="comment">#### 导入字符串处理程序包</span></span><br><span class="line">library(vegan)   <span class="comment">#### 导入物种名处理程序包</span></span><br><span class="line"><span class="comment">## Loading required package: permute</span></span><br><span class="line"><span class="comment">## Loading required package: lattice</span></span><br><span class="line"><span class="comment">## This is vegan 2.5-3</span></span><br><span class="line">library(coefplot)<span class="comment">#### 绘制参数图</span></span><br><span class="line"><span class="comment">## Loading required package: ggplot2</span></span><br><span class="line">library(knitr)</span><br><span class="line">library(lme4)</span><br><span class="line"><span class="comment">## Loading required package: Matrix</span></span><br><span class="line">library(broom)</span><br><span class="line"><span class="comment">#### 生成物种编码， 并全部转换为大写</span></span><br><span class="line">spcode &lt;- toupper(make.cepnames(sp))</span><br><span class="line"><span class="comment">### 物种数90</span></span><br><span class="line">nsp &lt;- <span class="built_in">length</span>(spcode)<span class="comment">### 个体数</span></span><br><span class="line">nind &lt;- 12700</span><br><span class="line"><span class="comment">### 假设多度服从以下分布</span></span><br><span class="line">temp &lt;- rls(n = nsp, N = <span class="number">10000</span>, alpha = <span class="number">6</span>)</span><br><span class="line"><span class="comment">#### 生成名称字符串向量</span></span><br><span class="line">sppool &lt;- <span class="built_in">rep</span>(spcode, temp)</span><br><span class="line"><span class="comment">#### 随机筛选出物种， 长度与样地的个体数相同</span></span><br><span class="line">species &lt;- sample(sppool, nind, replace = <span class="literal">TRUE</span>)</span><br><span class="line"><span class="comment">#### 查看每个种的个体数#### </span></span><br><span class="line">table(species)</span><br><span class="line"><span class="comment">#### 计算总的个体数</span></span><br><span class="line"><span class="built_in">sum</span>(table(species))</span><br><span class="line"><span class="comment">## [1] 12700</span></span><br><span class="line"><span class="comment">#### 样地长250m， 对应nind个x坐标，坐标的取值范围从0-250</span></span><br><span class="line">gx &lt;- <span class="built_in">round</span>(runif(nind)*<span class="number">250</span>, digits = <span class="number">2</span>)</span><br><span class="line"><span class="comment">#### 样地宽200m， 对应nind个y坐标，坐标的取值范围从0-250</span></span><br><span class="line">gy &lt;- <span class="built_in">round</span>(runif(nind)*<span class="number">200</span>, digits = <span class="number">2</span>)</span><br><span class="line"></span><br><span class="line">xlab20 &lt;- cut(gx, breaks = seq(<span class="number">0</span>, <span class="number">260</span>, by = <span class="number">20</span>), labels = <span class="number">1</span>:<span class="number">13</span>)</span><br><span class="line">ylab20 &lt;- cut(gy, breaks = seq(<span class="number">0</span>, <span class="number">220</span>, by = <span class="number">20</span>), labels = <span class="number">1</span>:<span class="number">11</span>)</span><br><span class="line"><span class="comment">#### check the quardrat name</span></span><br><span class="line">sort(unique(ylab20))</span><br><span class="line"><span class="comment">##  [1] 1  2  3  4  5  6  7  8  9  10</span></span><br><span class="line"><span class="comment">## Levels: 1 2 3 4 5 6 7 8 9 10 11</span></span><br><span class="line">sort(unique(xlab20))</span><br><span class="line"><span class="comment">##  [1] 1  2  3  4  5  6  7  8  9  10 11 12 13</span></span><br><span class="line"><span class="comment">## Levels: 1 2 3 4 5 6 7 8 9 10 11 12 13</span></span><br><span class="line"><span class="comment">#### generate quadrat name</span></span><br><span class="line">quadrat20 &lt;- paste(str_pad(xlab20, <span class="number">2</span>, pad =<span class="string">&quot;0&quot;</span>), </span><br><span class="line">                       str_pad(ylab20, <span class="number">2</span>, pad =<span class="string">&quot;0&quot;</span>), sep = <span class="string">&quot;&quot;</span>)</span><br><span class="line"><span class="comment">##### table(quadrat20)</span></span><br><span class="line"><span class="comment">#### 假设dbh服从指数分布</span></span><br><span class="line">dbh = dbh &lt;-<span class="built_in">round</span>(rexp(nind,rate =<span class="number">2</span>)*<span class="number">50</span>+<span class="number">1</span>, <span class="number">1</span>)</span><br><span class="line"><span class="comment">#### 假设样地名为白云山BYS （白云山物种名录参考 陈云等 2017）, 物种标牌</span></span><br><span class="line">tag &lt;- paste(<span class="string">&quot;BYS&quot;</span>, str_pad(<span class="number">1</span>:nind, <span class="number">6</span>, pad =<span class="string">&quot;0&quot;</span>), sep =<span class="string">&quot;&quot;</span>)</span><br><span class="line"><span class="comment">#### 物种的生存状态</span></span><br><span class="line">status &lt;- rbinom(nind, size = <span class="number">1</span>, prob = <span class="number">0.01</span>)</span><br><span class="line"><span class="comment">#### 合并到一个data.frame中</span></span><br><span class="line">bys &lt;- data.frame(tag, species, gx, gy, quadrat20, dbh, status)</span><br><span class="line"><span class="comment">#### 以上样地模拟数据生成。 请注意以上数据都是随机生成， 数据只供练习用，不宜用来做零模型或其他分析。 </span></span><br><span class="line"><span class="comment">#### 数据准备</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#### 首先，去除个体数少于30个的个体</span></span><br><span class="line"><span class="comment">####  每个种的个体数</span></span><br><span class="line">nindividual &lt;- table(bys$species)</span><br><span class="line"><span class="comment">#### 大于等于30个个体的物种名</span></span><br><span class="line">sp_to_select &lt;- <span class="built_in">names</span>(nindividual[nindividual &gt;= <span class="number">30</span>])</span><br><span class="line"><span class="comment">#### 仅保留名单中出现的物种</span></span><br><span class="line">sub_bys &lt;- subset(bys, subset = species %in% sp_to_select)</span><br><span class="line"><span class="comment">#### 提取一部分数据的前2000条， 用于检验下面的函数</span></span><br><span class="line">sub_bys_test &lt;- sub_bys[<span class="number">1</span>:<span class="number">2000</span>,]</span><br><span class="line"></span><br><span class="line"><span class="comment">#### 计算每一个个体， 在某一半径范围内同种和异种的胸径截面积</span></span><br><span class="line"><span class="comment">#### dat 为样地数据， 必须包括以下各列</span></span><br><span class="line"><span class="comment">#### gx 个体在样地内的x坐标， 单位为m</span></span><br><span class="line"><span class="comment">#### gy 个体在样地内的y坐标， 单位为m</span></span><br><span class="line"><span class="comment">#### species 个体的名称缩写</span></span><br><span class="line"><span class="comment">#### dbh 每个个体的胸径，单位为cm</span></span><br><span class="line"><span class="comment">#### r 从任何一个个体出发所画圆盘的半径</span></span><br><span class="line"></span><br><span class="line">compute_neighbour_area &lt;- <span class="keyword">function</span> (dat, r = <span class="literal">NULL</span>)</span><br><span class="line">&#123;   </span><br><span class="line">    <span class="keyword">if</span>(<span class="built_in">any</span>(<span class="built_in">is.na</span>(dat)))&#123;  </span><br><span class="line">    <span class="comment">### 如果出现NA缺失数据， 则需要先去除， 以免产生未知后果</span></span><br><span class="line">        dat &lt;- na.omit(dat)        </span><br><span class="line">        warning(<span class="string">&quot;NAs detected in the data and the entries </span></span><br><span class="line"><span class="string">                  containing NAs have been omitted.&quot;</span>)</span><br><span class="line">    &#125;</span><br><span class="line">    dat$ID &lt;- 1:nrow(dat)  </span><br><span class="line">    <span class="comment">### 为每一条物种记录增加一个ID</span></span><br><span class="line">    </span><br><span class="line">    <span class="keyword">if</span>(<span class="built_in">is.null</span>(dat$gx))&#123;   </span><br><span class="line">      <span class="comment">### 检查gx</span></span><br><span class="line">      stop(<span class="string">&quot;Column gx is missing&quot;</span>)</span><br><span class="line">    &#125; </span><br><span class="line">       </span><br><span class="line">    <span class="keyword">if</span>(<span class="built_in">is.null</span>(dat$gy))&#123;   </span><br><span class="line">      <span class="comment">### 检查gy</span></span><br><span class="line">      stop(<span class="string">&quot;Column gy is missing&quot;</span>)</span><br><span class="line">    &#125; </span><br><span class="line">       </span><br><span class="line">    <span class="keyword">if</span>(<span class="built_in">is.null</span>(dat$species))&#123;  </span><br><span class="line">      <span class="comment">### 检查物种名</span></span><br><span class="line">      stop(<span class="string">&quot;Column species is missing&quot;</span>)</span><br><span class="line">    &#125;    </span><br><span class="line">    </span><br><span class="line">    <span class="keyword">if</span>(<span class="built_in">is.null</span>(dat$dbh))&#123;   </span><br><span class="line">      <span class="comment">### 检查dbh</span></span><br><span class="line">      stop(<span class="string">&quot;Column species is missing&quot;</span>)</span><br><span class="line">    &#125;    </span><br><span class="line">    </span><br><span class="line">    <span class="keyword">if</span>(<span class="built_in">is.null</span>(r))&#123;  </span><br><span class="line">      <span class="comment">### 检查是否有提供r</span></span><br><span class="line">      stop(<span class="string">&quot;r is not specified&quot;</span>)</span><br><span class="line">    &#125;</span><br><span class="line">    gx &lt;- dat$gx</span><br><span class="line">    gy &lt;- dat$gy</span><br><span class="line">    dat$species &lt;- <span class="built_in">as.character</span>(dat$species)  </span><br><span class="line">    <span class="comment">### 转换为字符串主要是为了方便后面进行物种筛选</span></span><br><span class="line">    species_each_id &lt;- dat$species   </span><br><span class="line">    <span class="comment">### 由于进行的循环数较多， 所以最好不要在循环中不断提取物种， 而是先生成一个字符串向量</span></span><br><span class="line">    area_of_neighbours_same_species &lt;- <span class="built_in">rep</span>(<span class="literal">NA</span>, nrow(dat))  </span><br><span class="line">    <span class="comment">### 生成空白数据</span></span><br><span class="line">    area_of_neighbours_diff_species &lt;- <span class="built_in">rep</span>(<span class="literal">NA</span>, nrow(dat))  </span><br><span class="line">    <span class="comment">### 生成空白数据</span></span><br><span class="line">    print(paste(nrow(dat), <span class="string">&quot;rows to process.&quot;</span>))   </span><br><span class="line">    <span class="comment">### 提示数据量</span></span><br><span class="line">    <span class="keyword">for</span> (i <span class="keyword">in</span> <span class="number">1</span>:nrow(dat))&#123;        </span><br><span class="line">       <span class="keyword">if</span>(i %% <span class="number">100</span> == <span class="number">0</span> | i == <span class="number">1</span>)&#123;   </span><br><span class="line">    <span class="comment">### 每计算100个，打印正在计算哪条</span></span><br><span class="line">            print(paste(<span class="string">&quot;Computing neighbourhood area for row:&quot;</span>, i))</span><br><span class="line">        &#125;</span><br><span class="line">        x &lt;- gx[i]   <span class="comment">### 提取第i个</span></span><br><span class="line">        y &lt;- gy[i]        </span><br><span class="line">        <span class="comment">#### 先把数据量缩小到个体加减半径范围内的方块内， 这样能减少很多计算量</span></span><br><span class="line">        dat_sub &lt;- dat[(gx &lt; x + r) &amp; (gx &gt; x - r) &amp; (gy &lt; y + r) &amp; (gy &gt; y - r), ] </span><br><span class="line">        <span class="comment">### Select the area, reduce the computation of each loop</span></span><br><span class="line">        </span><br><span class="line">        <span class="comment">#### 计算相邻树木的胸径截面积需要将中心树种本身去除</span></span><br><span class="line">        <span class="comment">#### Exclude the focusing tree during the computation</span></span><br><span class="line">        dat_sub &lt;- dat_sub[dat_sub$ID != i, ]        </span><br><span class="line">        <span class="comment">#### 如果dat_sub 的行数大于0， 表明在缩小的方块内是有其他树种的</span></span><br><span class="line">        <span class="keyword">if</span>(nrow(dat_sub) &gt; <span class="number">0</span>)&#123;          </span><br><span class="line">            <span class="comment">#### 先生成一个空白的向量， 用来保存哪些记录是处在圆盘内的， 这里的下表是针对方块内的小数据的</span></span><br><span class="line">            sp &lt;- <span class="built_in">rep</span>(<span class="literal">NA</span>, nrow(dat_sub))  </span><br><span class="line">            <span class="comment">### empty vector, the index of the entries within the distance </span></span><br><span class="line">            <span class="comment">### r will be saved to sp.</span></span><br><span class="line">            <span class="keyword">for</span>(j <span class="keyword">in</span> <span class="number">1</span>:nrow(dat_sub))&#123; </span><br><span class="line">                <span class="comment">### 对方块内所有的物种计算欧氏距离，并做比较， 做j次 </span></span><br><span class="line">                mat &lt;- data.frame(<span class="built_in">c</span>(x, dat_sub$gx[j]), <span class="built_in">c</span>(y, dat_sub$gy[j]))</span><br><span class="line">                sp[j] &lt;- (dist(mat) &lt;= r)</span><br><span class="line">            &#125;            </span><br><span class="line">            <span class="comment">#### 所有个体的胸径截面积</span></span><br><span class="line">            area_of_neighbours &lt;- <span class="built_in">pi</span>*(dat_sub$dbh[sp]/<span class="number">2</span>)^<span class="number">2</span>  </span><br><span class="line">            <span class="comment">### Trunk Area at the breast height for all the neighbouring trees </span></span><br><span class="line">            <span class="comment">### in the subset dataset, within distance R</span></span><br><span class="line">            </span><br><span class="line">            <span class="comment">#### 如果同种的在圆盘内出现了，那么胸径截面积向量的长度肯定不是0，这时候就保存结果到第i个位置</span></span><br><span class="line">            <span class="keyword">if</span>(<span class="built_in">length</span>(area_of_neighbours[dat_sub$species[sp] %in% </span><br><span class="line">                  species_each_id[i]]) != <span class="number">0</span> )&#123;</span><br><span class="line">                area_of_neighbours_same_species[i] &lt;- </span><br><span class="line">                    <span class="built_in">sum</span>(area_of_neighbours[dat_sub$species[sp] %in% species_each_id[i]])  </span><br><span class="line">                    <span class="comment">### The trunk area at breast height of the same species</span></span><br><span class="line">            &#125;            </span><br><span class="line">            <span class="comment">#### 如果不同种在圆盘内出现了，那么胸径截面积向量的长度肯定不是0，这时候就保存结果到第i个位置</span></span><br><span class="line">            <span class="keyword">if</span>(<span class="built_in">length</span>(area_of_neighbours[!dat_sub$species[sp] %in% </span><br><span class="line">                species_each_id[i]]) != <span class="number">0</span>)&#123;</span><br><span class="line">                area_of_neighbours_diff_species[i] &lt;- </span><br><span class="line">                     <span class="built_in">sum</span>(area_of_neighbours[!dat_sub$species[sp] %in% </span><br><span class="line">                species_each_id[i]]) </span><br><span class="line">                <span class="comment">### The trunk area of breast height of different species</span></span><br><span class="line">            &#125;</span><br><span class="line">        &#125; <span class="keyword">else</span> &#123; </span><br><span class="line">          <span class="comment">### If there is no tree within the distance r, return NA</span></span><br><span class="line">          <span class="comment">#### 如果没有出现任何树木，第i个胸径截面积都是NA</span></span><br><span class="line">            area_of_neighbours_same_species[i] &lt;- <span class="literal">NA</span></span><br><span class="line">            area_of_neighbours_diff_species[i] &lt;- <span class="literal">NA</span></span><br><span class="line">        &#125;</span><br><span class="line">    &#125;    <span class="comment">#### 合并成data.frame</span></span><br><span class="line">    res &lt;- data.frame(area_of_neighbours_same_species, </span><br><span class="line">                      area_of_neighbours_diff_species)    </span><br><span class="line">    <span class="comment">#### 为了数据处理方便，在返回值的列名上增加半径 r的取值</span></span><br><span class="line">    colnames(res) &lt;- paste(<span class="built_in">c</span>(<span class="string">&quot;area_of_neighbours_same_species&quot;</span>, </span><br><span class="line">                           <span class="string">&quot;area_of_neighbours_diff_species&quot;</span>), </span><br><span class="line">                           <span class="string">&quot;_r_&quot;</span>, r, sep = <span class="string">&quot;&quot;</span>)    </span><br><span class="line">    </span><br><span class="line">    <span class="built_in">return</span>(res)</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="comment">#### debug(compute_neighbour_area)</span></span><br><span class="line"></span><br><span class="line">neighbour_area_res10 &lt;- compute_neighbour_area(sub_bys_test, r = <span class="number">10</span>) </span><br><span class="line"><span class="comment">### 半径10m</span></span><br><span class="line"><span class="comment">## [1] &quot;2000 rows to process.&quot;</span></span><br><span class="line"><span class="comment">## [1] &quot;Computing neighbourhood area for row: 1&quot;</span></span><br><span class="line"><span class="comment">## [1] &quot;Computing neighbourhood area for row: 100&quot;</span></span><br><span class="line"><span class="comment">## [1] &quot;Computing neighbourhood area for row: 200&quot;</span></span><br><span class="line"><span class="comment">## [1] &quot;Computing neighbourhood area for row: 300&quot;</span></span><br><span class="line"><span class="comment">## [1] &quot;Computing neighbourhood area for row: 400&quot;</span></span><br><span class="line"><span class="comment">## [1] &quot;Computing neighbourhood area for row: 500&quot;</span></span><br><span class="line"><span class="comment">## [1] &quot;Computing neighbourhood area for row: 600&quot;</span></span><br><span class="line"><span class="comment">## [1] &quot;Computing neighbourhood area for row: 700&quot;</span></span><br><span class="line"><span class="comment">## [1] &quot;Computing neighbourhood area for row: 800&quot;</span></span><br><span class="line"><span class="comment">## [1] &quot;Computing neighbourhood area for row: 900&quot;</span></span><br><span class="line"><span class="comment">## [1] &quot;Computing neighbourhood area for row: 1000&quot;</span></span><br><span class="line"><span class="comment">## [1] &quot;Computing neighbourhood area for row: 1100&quot;</span></span><br><span class="line"><span class="comment">## [1] &quot;Computing neighbourhood area for row: 1200&quot;</span></span><br><span class="line"><span class="comment">## [1] &quot;Computing neighbourhood area for row: 1300&quot;</span></span><br><span class="line"><span class="comment">## [1] &quot;Computing neighbourhood area for row: 1400&quot;</span></span><br><span class="line"><span class="comment">## [1] &quot;Computing neighbourhood area for row: 1500&quot;</span></span><br><span class="line"><span class="comment">## [1] &quot;Computing neighbourhood area for row: 1600&quot;</span></span><br><span class="line"><span class="comment">## [1] &quot;Computing neighbourhood area for row: 1700&quot;</span></span><br><span class="line"><span class="comment">## [1] &quot;Computing neighbourhood area for row: 1800&quot;</span></span><br><span class="line"><span class="comment">## [1] &quot;Computing neighbourhood area for row: 1900&quot;</span></span><br><span class="line"><span class="comment">## [1] &quot;Computing neighbourhood area for row: 2000&quot;</span></span><br><span class="line"><span class="comment">#### 合并数据</span></span><br><span class="line">sub_bys_test &lt;- cbind(sub_bys_test, neighbour_area_res10)</span><br><span class="line"><span class="comment">#### </span></span><br><span class="line">head(sub_bys_test)</span><br><span class="line"><span class="comment">##         tag  species     gx    gy quadrat20   dbh status</span></span><br><span class="line"><span class="comment">## 1 BYS000001 SORBALNI  53.21 78.17      0304  54.7      0</span></span><br><span class="line"><span class="comment">## 2 BYS000002 POPUDAVI 115.82 79.36      0604  40.0      0</span></span><br><span class="line"><span class="comment">## 3 BYS000003 BERBCIRC  45.61  2.98      0301   6.4      0</span></span><br><span class="line"><span class="comment">## 4 BYS000004 PADUBUER  12.58  3.22      0101   6.5      0</span></span><br><span class="line"><span class="comment">## 5 BYS000005 PINUARMA 244.71 80.23      1305  44.2      0</span></span><br><span class="line"><span class="comment">## 6 BYS000006 LITSTSIN  20.92 68.87      0204 140.7      0</span></span><br><span class="line"><span class="comment">##   area_of_neighbours_same_species_r_10</span></span><br><span class="line"><span class="comment">## 1                           1899.35979</span></span><br><span class="line"><span class="comment">## 2                                   NA</span></span><br><span class="line"><span class="comment">## 3                            183.85386</span></span><br><span class="line"><span class="comment">## 4                             11.34115</span></span><br><span class="line"><span class="comment">## 5                           5242.44635</span></span><br><span class="line"><span class="comment">## 6                                   NA</span></span><br><span class="line"><span class="comment">##   area_of_neighbours_diff_species_r_10</span></span><br><span class="line"><span class="comment">## 1                            18553.107</span></span><br><span class="line"><span class="comment">## 2                            10102.828</span></span><br><span class="line"><span class="comment">## 3                            10170.050</span></span><br><span class="line"><span class="comment">## 4                             6330.403</span></span><br><span class="line"><span class="comment">## 5                             9277.956</span></span><br><span class="line"><span class="comment">## 6                             4614.992</span></span><br><span class="line"><span class="comment">#### 拟合广义线性混合模型</span></span><br><span class="line"><span class="comment">#### 各列数据标准化， 以便在参数图中进行比较相对重要性</span></span><br><span class="line">sub_bys_test$dbh  &lt;- scale(sub_bys_test$dbh)</span><br><span class="line">sub_bys_test$area_of_neighbours_same_species_r_10 &lt;- </span><br><span class="line">     scale( sub_bys_test$area_of_neighbours_same_species_r_10)</span><br><span class="line"></span><br><span class="line">sub_bys_test$area_of_neighbours_diff_species_r_10 &lt;- </span><br><span class="line">    scale (sub_bys_test$area_of_neighbours_diff_species_r_10)</span><br><span class="line">    <span class="comment">#### 设定广义线性混合模型</span></span><br><span class="line">    glmm_model_10m = glmer(status ~ dbh + </span><br><span class="line">                                    area_of_neighbours_same_species_r_10 + </span><br><span class="line">                                    area_of_neighbours_diff_species_r_10 + </span><br><span class="line">                                    (<span class="number">1</span>+dbh|species)+(<span class="number">1</span>|quadrat20),</span><br><span class="line">                                    data=sub_bys_test, </span><br><span class="line">                                    family = binomial(link=<span class="string">&quot;cloglog&quot;</span>))</span><br><span class="line"><span class="comment">## singular fit</span></span><br><span class="line">tidy(glmm_model_10m)</span><br><span class="line"><span class="comment">## Warning in bind_rows_(x, .id): binding factor and character vector,</span></span><br><span class="line"><span class="comment">## coercing into character vector</span></span><br><span class="line"><span class="comment">## Warning in bind_rows_(x, .id): binding character and factor vector,</span></span><br><span class="line"><span class="comment">## coercing into character vector</span></span><br><span class="line"><span class="comment">## # A tibble: 8 x 6</span></span><br><span class="line"><span class="comment">##   term                      estimate std.error statistic   p.value group  </span></span><br><span class="line"><span class="comment">##   &lt;chr&gt;                        &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt; &lt;chr&gt;  </span></span><br><span class="line"><span class="comment">## 1 (Intercept)               -4.62e+0     0.394   -11.7    9.18e-32 fixed  </span></span><br><span class="line"><span class="comment">## 2 dbh                       -1.23e-1     0.347    -0.354  7.24e- 1 fixed  </span></span><br><span class="line"><span class="comment">## 3 area_of_neighbours_sam…   -8.53e-1     0.777    -1.10   2.73e- 1 fixed  </span></span><br><span class="line"><span class="comment">## 4 area_of_neighbours_dif…    1.99e-1     0.280     0.712  4.76e- 1 fixed  </span></span><br><span class="line"><span class="comment">## 5 sd_(Intercept).quadrat…    2.29e-7    NA        NA     NA        quadra…</span></span><br><span class="line"><span class="comment">## 6 sd_(Intercept).species     0.         NA        NA     NA        species</span></span><br><span class="line"><span class="comment">## 7 sd_dbh.species             3.36e-8    NA        NA     NA        species</span></span><br><span class="line"><span class="comment">## 8 cor_(Intercept).dbh.sp…  NaN          NA        NA     NA        species</span></span><br><span class="line">glance(glmm_model_10m)</span><br><span class="line"><span class="comment">## # A tibble: 1 x 6</span></span><br><span class="line"><span class="comment">##   sigma logLik   AIC   BIC deviance df.residual</span></span><br><span class="line"><span class="comment">##   &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt;       &lt;int&gt;</span></span><br><span class="line"><span class="comment">## 1     1  -58.3  133.  171.     117.         918</span></span><br><span class="line">coefplot(glmm_model_10m)</span><br></pre></td></tr></table></figure>

<p>当然， 实际研究中， 最好能考虑每个个体调查时间的间隔 offset， 或者考虑更复杂的随机效应，特比是针对不同种， 在数据量允许的情况下， 计算各个种生存状态的影响因素等。</p>
<h1 id="参考文献"><a href="#参考文献" class="headerlink" title="参考文献"></a>参考文献</h1><ol>
<li>陈云,郭凌,姚成亮,韦博良,袁志良,叶永忠.暖温带⁃北亚热带过渡区落叶阔叶林群落特征.生态学报,2017,37(17):5602-5611.</li>
<li>Zhu, Yan, et al. “Conspecific and phylogenetic density‐dependent survival differs across life stages in a tropical forest.” Journal of Ecology 103.4 (2015): 957-966.</li>
</ol>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2017/10/27/2017-10-27-janzen/" data-id="ckrsvkzav008590w00rle6t0d" data-title="幼苗生存状态、负密度制约与Janzen Connell假说的检验" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E7%94%9F%E6%80%81/" rel="tag">生态</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2017-10-21-trees" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2017/10/21/2017-10-21-trees/" class="article-date">
  <time class="dt-published" datetime="2017-10-20T23:03:54.000Z" itemprop="datePublished">2017-10-21</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2017/10/21/2017-10-21-trees/">繁花满树 (十六首)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="1-小村里正午的阳光"><a href="#1-小村里正午的阳光" class="headerlink" title="1. 小村里正午的阳光"></a>1. 小村里正午的阳光</h1><p>  &nbsp;</p>
<p>  买好了回家的机票，</p>
<p>  但愿这次一切平安顺利，</p>
<p>  不要遇到大风雪。</p>
<p>  2000公里3.5小时的飞行</p>
<p>  比大雁更快</p>
<p>  &nbsp;</p>
<p>  不在乎朋友圈有多少赞，</p>
<p>  只愿家里一切安好。</p>
<p>  没有人在乎你的孤独，</p>
<p>  除了你自己，</p>
<p>  因为孤独只有自己能理解</p>
<p>  &nbsp;</p>
<p>  离乡十六年，</p>
<p>  电脑、手机、ipad，</p>
<p>  数字电视、数码相机、互联网，</p>
<p>  QQ、微信、电子邮件、高铁， </p>
<p>  还有家家户户的小汽车….</p>
<p>  还有DNA提取、基因组、PCR、Illumina，</p>
<p>  R语言、大样地、红色名录、Maxent….</p>
<p>  &nbsp;</p>
<p>  该来的就一定要来</p>
<p>  该走的就一定要走</p>
<p>  录音机、VCD、DVD、1.44M软盘、随身听、mp3….</p>
<p>  仿佛正午的阳光</p>
<p>  只维持一顿饭的功夫</p>
<p>  烟一样消散</p>
<p>  &nbsp;</p>
<p>  小村里只有捕鱼、耪地，</p>
<p>  不需要动植物标本，</p>
<p>  不需要多元回归树、最大熵和蒙特卡洛马尔科夫链….</p>
<p>  小村能自给自足，</p>
<p>  额外的都不需要，</p>
<p>  小村里说话时第二声还是要压低一半。</p>
<p>  &nbsp;</p>
<p>  在远方，</p>
<p>  总会想起小村。</p>
<p>  十六个春秋，</p>
<p>  多少不认识的孩子出生了，</p>
<p>  多少人曾在花季和雨季，</p>
<p>  多少人去了远方</p>
<p>  &nbsp;</p>
<p>  时间改变了你的容颜</p>
<p>  不变的只有乡音</p>
<p>  某一天正午的阳光</p>
<p>  和低沉的西北风</p>
<p>  &nbsp;</p>
<p>  2017年10月21日于大埔</p>
<p>  &nbsp;</p>
<h1 id="2-初冬的柿子"><a href="#2-初冬的柿子" class="headerlink" title="2. 初冬的柿子"></a>2. 初冬的柿子</h1><p>  &nbsp;</p>
<p>  初冬的清晨</p>
<p>  柿子挂满枝头</p>
<p>  散发着温暖的光</p>
<p>  静静守候</p>
<p>  &nbsp;</p>
<p>  37度的温暖</p>
<p>  照亮整洁的院子</p>
<p>  也照亮那一片荒芜</p>
<p>  还有深蓝的天</p>
<p>  &nbsp;</p>
<p>  叶子落了</p>
<p>  柿子还在</p>
<p>  南国的清晨五点</p>
<p>  北京也是五点</p>
<p>  &nbsp;</p>
<p>  寒夜走了</p>
<p>  天蒙蒙亮时</p>
<p>  柿树下的灯光还是那么温暖</p>
<p>  让你一下子回到了春天</p>
<p>  &nbsp;</p>
<h1 id="3-山楂"><a href="#3-山楂" class="headerlink" title="3. 山楂"></a>3. 山楂</h1><p>  &nbsp;</p>
<p>  那时候</p>
<p>  一小盆山楂</p>
<p>  就是外面的世界</p>
<p>  那天下午</p>
<p>  送货的卡车洒了一些山楂在村里</p>
<p>  这是村罐头厂对村民们的全部恩惠</p>
<p>  人人有份</p>
<p>  &nbsp;</p>
<p>  清晨的霜</p>
<p>  还未散去</p>
<p>  山楂就来了</p>
<p>  山楂也就一直没有走：</p>
<p>  在蓟县北山的树上</p>
<p>  在东城路</p>
<p>  在二道贩子的麻袋里</p>
<p>  住满了山楂</p>
<p>  &nbsp;</p>
<p>  山楂和山楂树</p>
<p>  一年又一年</p>
<p>  开花、结果</p>
<p>  从不知疲劳</p>
<p>  如同种山楂的人</p>
<p>  要养活自己和正在读书的儿女</p>
<p>  &nbsp;</p>
<p>  穿行在大街小巷的糖葫芦杆子</p>
<p>  哪里有婚丧嫁娶</p>
<p>  哪里下课、放学</p>
<p>  就聚集在哪里</p>
<p>  那是孩子们最喜欢的味道</p>
<p>  &nbsp;</p>
<p>  倾斜着的铝锅里</p>
<p>  绵白糖已经浓稠变黄</p>
<p>  串好的山楂在里面打个滚</p>
<p>  然后清脆地摔在湿木板上</p>
<p>  就生出了晶莹剔透的翅膀</p>
<p>  有了爬上竿子的资格</p>
<p>  就在那一瞬间</p>
<p>  &nbsp;</p>
<p>  三十年前的山楂籽从未发芽</p>
<p>  山楂还是一样红</p>
<p>  一样酸</p>
<p>  蘸糖葫芦的人</p>
<p>  现在开电瓶车</p>
<p>  不再蘸糖葫芦</p>
<p>  &nbsp;</p>
<p>  2017年10月18日</p>
<p>  &nbsp;</p>
<h1 id="4-清晨随想"><a href="#4-清晨随想" class="headerlink" title="4. 清晨随想"></a>4. 清晨随想</h1><p>  &nbsp;</p>
<p>  哈佛、耶鲁、斯坦福、佛罗里达、奥胡斯、根特…..，</p>
<p>  朋友圈里一长串大学的名字，</p>
<p>  熟悉而陌生，</p>
<p>  智慧在远方开枝散叶</p>
<p>  如果没有互联网，</p>
<p>  我只埋头于眼前的柴米油盐</p>
<p>  &nbsp;</p>
<p>  梦见联合国专门派人教我怎样养鸡</p>
<p>  怎样用水帘给鸡舍降温</p>
<p>  怎样无损伤收鸡蛋</p>
<p>  梦见派人教我怎样种桃子</p>
<p>  怎样避免收获时伤到桃子</p>
<p>  以免影响农产品的价格</p>
<p>  ……</p>
<p>  凡事都要小心</p>
<p>  &nbsp;</p>
<p>  四十岁的老屋里</p>
<p>  假装停着那三十五岁的自行车</p>
<p>  学校八点钟上课</p>
<p>  现在已经9点05，</p>
<p>  这堂课，</p>
<p>  要么是数学</p>
<p>  要么是物理</p>
<p>  我深信，都是因为那课程</p>
<p>  进入了一种让人听不懂的状态</p>
<p>  我才会迟到</p>
<p>  &nbsp;</p>
<p>  不知道为什么</p>
<p>  人要在火柴盒中生存</p>
<p>  也许因为这座城太小</p>
<p>  &nbsp;</p>
<p>  楼道里人来人往</p>
<p>  街上人来人往</p>
<p>  街上的人各自行路</p>
<p>  都不在乎</p>
<p>  远方的大学和那边的风景</p>
<p>  &nbsp;</p>
<p>  2017年10月9日清晨</p>
<p>  &nbsp;</p>
<h1 id="5-大埔的黄昏"><a href="#5-大埔的黄昏" class="headerlink" title="5. 大埔的黄昏"></a>5. 大埔的黄昏</h1><p>  &nbsp;</p>
<p>  大埔，</p>
<p>  华灯初上，</p>
<p>  深蓝的天空，</p>
<p>  多了一抹昏黄</p>
<p>  &nbsp;</p>
<p>  香港，</p>
<p>  一个人的傍晚，</p>
<p>  时间，</p>
<p>  在公路旁流淌</p>
<p>  &nbsp;</p>
<p>  来自南海的风，</p>
<p>  带来了一弯月色；</p>
<p>  林村河旁，</p>
<p>  是否遍洒流光?</p>
<p>  &nbsp;</p>
<p>  车走了，</p>
<p>  明天还会不会来？</p>
<p>  人生易老，</p>
<p>  作伴好还乡</p>
<p>  &nbsp;</p>
<p>  2017年9月28日于香港大埔</p>
<p>  &nbsp;</p>
<h1 id="6-倦"><a href="#6-倦" class="headerlink" title="6. 倦"></a>6. 倦</h1><p>  &nbsp;</p>
<p>  黑云压城：</p>
<p>  暴风雨到来的一瞬间，</p>
<p>  终于上了巴士，</p>
<p>  外面又是倾盆大雨……</p>
<p>  &nbsp;</p>
<p>  一年600次乘车，车上居然没有一张熟悉的面孔，</p>
<p>  这些人从哪里来？</p>
<p>  这些人会不会想家？</p>
<p>  也许不会想</p>
<p>  &nbsp;</p>
<p>  他们从没有背井离乡。</p>
<p>  &nbsp;</p>
<p>  但是我为什么要离开故乡？</p>
<p>  然后为什么要从零开始？</p>
<p>  还不都是因为自己太无能，</p>
<p>  太渺小……</p>
<p>  &nbsp;</p>
<p>  朋友圈分享自己的感受：</p>
<p>  其实都是为了慰藉自己的心，</p>
<p>  让自己少些孤独和伤感，</p>
<p>  请原谅我的自私</p>
<p>  &nbsp;</p>
<p>  倦了：</p>
<p>  你不停地索取，</p>
<p>  我不停地给予，</p>
<p>  从不懂怜悯或拒绝</p>
<p>  &nbsp;</p>
<p>  孤独的雨夜，一个人听雨</p>
<p>  仿佛一个人饮酒</p>
<p>  窗外灯光昏黄……</p>
<p>  我只想着回不去的故乡</p>
<p>  &nbsp;</p>
<h1 id="7-台风之夜"><a href="#7-台风之夜" class="headerlink" title="7. 台风之夜"></a>7. 台风之夜</h1><p>  &nbsp;</p>
<p>  台风如过客，</p>
<p>  来也匆匆，</p>
<p>  去也匆匆。</p>
<p>  &nbsp;</p>
<p>  回到多年前那个夏天的傍晚，</p>
<p>  也许……</p>
<p>  一切能重新开始</p>
<p>  没有那一晚，</p>
<p>  就不会有现在….</p>
<p>  &nbsp;</p>
<p>  奋力追赶，</p>
<p>  却依然落后，</p>
<p>  也许根本不在正确的轨道上</p>
<p>  &nbsp;</p>
<p>  你说我意气用事，</p>
<p>  我比谁都更清楚自己；</p>
<p>  追寻台风的脚步，</p>
<p>  风里来，雨里去….</p>
<p>  &nbsp;</p>
<p>  我愿沉醉在这雨中，</p>
<p>  走走停停，</p>
<p>  聆听，聆听….</p>
<p>  &nbsp;</p>
<h1 id="8-秋天快来"><a href="#8-秋天快来" class="headerlink" title="8. 秋天快来"></a>8. 秋天快来</h1><p>  台风在辽阔的太平洋上转出一个大写的M，</p>
<p>  也许它的征途是星辰大海，</p>
<p>  我并不理会，</p>
<p>  但是，</p>
<p>  这是多好的广告</p>
<p>  &nbsp;</p>
<p>  世界需要降温、降雨，</p>
<p>  发展速度需要降，</p>
<p>  房价需要降，</p>
<p>  只是网速和工资都不要降….</p>
<p>  &nbsp;</p>
<p>  改完稿子已经凌晨一点，</p>
<p>  不知道世界上为什么要有那么多论文，</p>
<p>  为什么要有那么多人做研究，</p>
<p>  其实很多人只是在科研中寻求温饱</p>
<p>  &nbsp;</p>
<p>  音乐还在响着，</p>
<p>  下一首曲子是贝多芬的《命运》，</p>
<p>  这星期还有一篇稿子要审….</p>
<p>  …..</p>
<p>  &nbsp;</p>
<p>  不想这么焦虑，</p>
<p>  不想这样疲惫，</p>
<p>  不想这样打拼，</p>
<p>  背着厚厚的蜗牛壳，</p>
<p>  我并不知自己被什么驱动，</p>
<p>  在外国人的语言与文字间艰难地爬行</p>
<p>  &nbsp;</p>
<p>  只希望秋天快来：</p>
<p>  到那时候，</p>
<p>  去收获满车的高粱和玉米，</p>
<p>  一定要堆得老高，</p>
<p>  同时收获满院子的金色阳光，</p>
<p>  收获黎明前的露水，</p>
<p>  还有院子里那棵玉簪花的阵阵香气</p>
<p>  &nbsp;</p>
<h1 id="9-明天就回去"><a href="#9-明天就回去" class="headerlink" title="9. 明天就回去"></a>9. 明天就回去</h1><p>  &nbsp;</p>
<p>  月光照亮你</p>
<p>  也照亮了安静的小村</p>
<p>  &nbsp;</p>
<p>  你留住了宁静，</p>
<p>  也留住了一样宁静的天空</p>
<p>  没有路灯的小村</p>
<p>  我们就只需要月光；</p>
<p>  没有灯红酒绿的小村，</p>
<p>  我们就只需要夜的沉寂</p>
<p>  &nbsp;</p>
<p>  在那里长大，</p>
<p>  一生一世就留在了那里：</p>
<p>  十八年的日日夜夜，</p>
<p>  乡音，</p>
<p>  还有母亲的双眼，</p>
<p>  都盼你回去</p>
<p>  &nbsp;</p>
<p>  明天就回去：</p>
<p>  回那里耕田、放牛、种菜，</p>
<p>  回那里划船、逮鱼、做小本生意</p>
<p>  再回到那平平常常的日子，</p>
<p>  然后去五里以外赶集</p>
<p>  那个只有半天的小集市….</p>
<p>  &nbsp;</p>
<p>  明天就回去，</p>
<p>  再不要四处流浪；</p>
<p>  明天就回去，</p>
<p>  回到皎洁的月光里</p>
<p>  &nbsp;</p>
<h1 id="10-天牛"><a href="#10-天牛" class="headerlink" title="10. 天牛"></a>10. 天牛</h1><p>  &nbsp;</p>
<p>  这一刻，</p>
<p>  你行走在一片风景里。</p>
<p>  &nbsp;</p>
<p>  这一刻，</p>
<p>  你忘记了曾经的孤独，</p>
<p>  也许也忘记了自己。</p>
<p>  &nbsp;</p>
<p>  这一刻，</p>
<p>  你行走在一间大房子外面，</p>
<p>  想看看谁在办公室里偷懒….</p>
<p>  &nbsp;</p>
<p>  你行走在空中，</p>
<p>  就像行走在树上；</p>
<p>  你行走在树上，</p>
<p>  阳光就填充了你的思想</p>
<p>  &nbsp;</p>
<h1 id="11-地球日植树"><a href="#11-地球日植树" class="headerlink" title="11. 地球日植树"></a>11. 地球日植树</h1><p>  &nbsp;</p>
<p>  当你种下一棵树，</p>
<p>  铺上防草垫，</p>
<p>  就又为大地打了一块补丁</p>
<p>  &nbsp;</p>
<p>  从枯黄到翠绿，</p>
<p>  到山溪再次流淌，</p>
<p>  你的双手成就了大自然的欢愉</p>
<p>  &nbsp;</p>
<p>  一米高的芒草，</p>
<p>  伴随着太阳，清风，白云，</p>
<p>  掠过植树区</p>
<p>  – 你需要漫长的等待</p>
<p>  &nbsp;</p>
<p>  当你种下一棵树，</p>
<p>  锤击钢铁的声音，</p>
<p>  刺痛了耳膜，</p>
<p>  也刺痛了大地</p>
<p>  &nbsp;</p>
<p>  6月4日</p>
<p>  于香港大埔</p>
<p>  &nbsp;</p>
<h1 id="12-夏天里的一朵花"><a href="#12-夏天里的一朵花" class="headerlink" title="12. 夏天里的一朵花"></a>12. 夏天里的一朵花</h1><p>  &nbsp;</p>
<p>  没人记得一朵花，</p>
<p>  在大地上</p>
<p>  怎样绽放、枯萎，</p>
<p>  怎样随风而逝，</p>
<p>  怎样埋入黄土</p>
<p>  &nbsp;</p>
<p>  没人记得一朵丝瓜花</p>
<p>  在某一天傍晚</p>
<p>  瓜熟蒂落</p>
<p>  那时候火烧云正红彻天际</p>
<p>  &nbsp;</p>
<p>  夏天里重复着夏天</p>
<p>  草芥依然是草芥，</p>
<p>  夏天里的一朵花，</p>
<p>  随着火烧云消失在深蓝色的夜空里</p>
<p>  没有人再记得</p>
<p>  &nbsp;</p>
<h1 id="13-时间里的花朵"><a href="#13-时间里的花朵" class="headerlink" title="13. 时间里的花朵"></a>13. 时间里的花朵</h1><p>  &nbsp;</p>
<p>  —— <strong>写给采标本的人</strong></p>
<p>  &nbsp;</p>
<p>  5月26日同金刚、慧玲、阿芳一起参观香港渔农自然护理署植物标本室，见到若干超过百年的植物标本，其中不乏Champion、Ford、Dunn、Levine、Maclure等名家的标本有感</p>
<p>  &nbsp;</p>
<p>  你的工作默默无闻，</p>
<p>  然而时间会记得，</p>
<p>  一百年后的人们也会记得</p>
<p>  &nbsp;</p>
<p>  清晰的记录，</p>
<p>  一百年后，</p>
<p>  依然映出你在那一刻是怎样屏息凝神：</p>
<p>  在那一刻，</p>
<p>  你想在时间的长河里留下一朵花，</p>
<p>  让它永不凋谢；</p>
<p>  &nbsp;</p>
<p>  仿佛一壶老酒，</p>
<p>  芳香四溢</p>
<p>  岁月留住了你想留住的花朵，</p>
<p>  也留住了你的心</p>
<p>  &nbsp;</p>
<h1 id="14-霞光里的少年"><a href="#14-霞光里的少年" class="headerlink" title="14. 霞光里的少年"></a>14. <strong>霞光里的少年</strong></h1><p>  &nbsp;</p>
<p>  遥想1997年初春里的一天，</p>
<p>  一个十三岁的少年，</p>
<p>  在寒夜里爬起身，</p>
<p>  一个人到一里路以外的河边看海尔-波普彗星</p>
<p>  &nbsp;</p>
<p>  寒气袭人，</p>
<p>  东方泛白。</p>
<p>  彗星高悬在东方的天空，</p>
<p>  就像快要降落天际的羽毛球</p>
<p>  &nbsp;</p>
<p>  满地白霜，</p>
<p>  潮白河边，</p>
<p>  脚下那一层薄冰，</p>
<p>  咔咔作响</p>
<p>  &nbsp;</p>
<p>  二十岁的河流沐浴在朝霞里，</p>
<p>  还有河面上的一层浮冰，</p>
<p>  以及二三里以外河堤上的树木</p>
<p>  ……</p>
<p>  &nbsp;</p>
<p>  大地还在沉睡，</p>
<p>  村庄就像一个婴儿，</p>
<p>  做着自己甜美的梦</p>
<p>  &nbsp;</p>
<p>  我只愿做一个渔夫，</p>
<p>  撑船，</p>
<p>  撒网，</p>
<p>  收获这样的清晨和喜悦</p>
<p>  &nbsp;</p>
<p>  二十年后，</p>
<p>  在五千里之外，</p>
<p>  大地还是大地</p>
<p>  回忆起那一天，</p>
<p>  少年又回到了那个清晨</p>
<p>  &nbsp;</p>
<p>  2017年5月24日</p>
<p>  于香港大埔</p>
<p>  &nbsp;</p>
<h1 id="15-大帽山以北"><a href="#15-大帽山以北" class="headerlink" title="15. 大帽山以北"></a>15. <strong>大帽山以北</strong></h1><p>  &nbsp;</p>
<p>  春天里，</p>
<p>  &nbsp;</p>
<p>  大雾弥漫。</p>
<p>  春天的天气完全不顾老板的理想：</p>
<p>  我们一安排上山，</p>
<p>  就要下雨。</p>
<p>  &nbsp;</p>
<p>  春天的植树区，</p>
<p>  海拔还是600米。</p>
<p>  &nbsp;</p>
<p>  春天的大帽山北坡，</p>
<p>  不管我们多着急，</p>
<p>  还是只有两三个人在测量树木。</p>
<p>  &nbsp;</p>
<h1 id="16-繁花满树—-写给钟花樱桃"><a href="#16-繁花满树—-写给钟花樱桃" class="headerlink" title="16. 繁花满树—-写给钟花樱桃"></a>16. <strong>繁花满树—-写给钟花樱桃</strong></h1><p>  &nbsp;</p>
<p>  盼望你繁花满树，</p>
<p>  盼望你果实挂满枝头，</p>
<p>  盼望你不再孤独，</p>
<p>  盼望你平安如意。</p>
<p>  花非花，</p>
<p>  雾非雾…</p>
<p>  &nbsp;</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2017/10/21/2017-10-21-trees/" data-id="ckrsvkzav008390w04y4lh33u" data-title="繁花满树 (十六首)" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E8%AF%97%E6%AD%8C/" rel="tag">诗歌</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2017-10-19-glm" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2017/10/19/2017-10-19-glm/" class="article-date">
  <time class="dt-published" datetime="2017-10-19T05:06:54.000Z" itemprop="datePublished">2017-10-19</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2017/10/19/2017-10-19-glm/">广义线性模型中的逻辑斯蒂回归</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>其实题目不太准确，这篇短文介绍的是二项分布线性模型， 也就是当响应变量为01时的线性模型，例如物种是否出现，病人是否需要住院，学生是否成功考入大学，英语四六级成绩是否及格等。逻辑斯蒂回归模型，是二项分布线性模型中的一种，也是最广泛的一种。</p>
<p>事情还要从1935年说起。那一年，英国伦敦大学学院的C.I. Bliss 在《Annals of Applied Biology》上发表论文The Calculation of the Dosage**-**Mortality Curve。这篇论文的表4，就是现在广泛用于逻辑斯蒂回归教学的beetle数据。该数据是C.I. Bliss从Strand在1930年发表的论文中摘出来的。</p>
<p>Bliss的这篇论文主要以杂拟谷盗**(<strong>Tribolium confusum</strong>)**为例，系统介绍拟合存活比例曲线的方法。杂拟谷盗是一种粮食害虫。当时生物学家将481只杂拟谷盗分成8组，每一组分别在暴露在不同浓度的二硫化碳5小时，之后记录每个组有多少只甲虫死了。由于甲虫对二硫化碳的反应只有两种可能，要么存活，要么死亡，所以这个数据集是典型的二项分布数据。</p>
<p>在这篇论文中，Bliss提出了一种连接函数，也就是事件发生与否，用转换以后的正态分布累计曲线去估算，称为Probit。实际上，二项分布线性模型的连接函数有三种，分别为logit, probit以及cloglog。 logit、probit以及cloglog都是统计变换，目的是将二项分布数据转换为一种形式，以便能用线性模型估计直接估计模型的参数。其中最常见的为logit，使用logit连接函数的二项分布回归称为逻辑斯蒂回归。logit是事件发生与是否发生的比例 （odds）的对数值，形为log**(<strong>y</strong>/(<strong>1</strong>-<strong>y</strong>))<strong>，称为log odds。cloglog是complementary log log的缩写，是将响应变量进行两次log变换， 公式为log</strong>(-<strong>log</strong>(<strong>1</strong>-<strong>y</strong>))**。 cloglog常用来拟合数据出现大量极端数值的数据， 例如，大样地的幼苗的负密度制约研究中，部分胸径等级的树苗只有很少一部分死亡，用cloglog能较准确捕捉到与树苗死亡相关性较强的环境因子， 如同种密度制约、光环境等。</p>
<p>下面用R代码演示如何进行二项分布回归，分别用logit, probit以及cloglog三种连接函数。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br></pre></td><td class="code"><pre><span class="line">library(broom)</span><br><span class="line"><span class="comment">####表格形式的数据</span></span><br><span class="line"><span class="comment">####tabular form</span></span><br><span class="line"><span class="comment">####数据来源www.stat.ualberta.ca/~kcarrier/STAT562/comp_log_log</span></span><br><span class="line"><span class="comment">####beetle</span></span><br><span class="line"><span class="comment">####   LOGDOSE     NUMBER_OF_BEETLES  NUMBER_KILLED  COMP_LOG_LOGPROBITLOGIT</span></span><br><span class="line"><span class="comment">####1   1.691                59             6          5.7   3.4 35.0</span></span><br><span class="line"><span class="comment">####2   1.724                60            13         11.3  10.7  9.8</span></span><br><span class="line"><span class="comment">####3   1.755                62           18         20.9   23.422.4</span></span><br><span class="line"><span class="comment">####4   1.784                56            28         30.3  33.8 33.9</span></span><br><span class="line"><span class="comment">####5   1.811                63            52         47.7  49.6 50.0</span></span><br><span class="line"><span class="comment">####6   1.837                59            53         54.2  53.4 53.3</span></span><br><span class="line"><span class="comment">####7   1.861                62            61         61.1  59.7 59.2</span></span><br><span class="line"><span class="comment">####8   1.884                60            60         59.9  59.2 58.8</span></span><br><span class="line"><span class="comment">####&gt;</span></span><br><span class="line"><span class="comment">####二硫化碳的浓度</span></span><br><span class="line">LOGDOSE          &lt;-<span class="built_in">c</span>(<span class="number">1.691</span>, <span class="number">1.724</span>, <span class="number">1.755</span>, <span class="number">1.784</span>, <span class="number">1.811</span>, <span class="number">1.837</span>, <span class="number">1.861</span>, <span class="number">1.884</span> )</span><br><span class="line"><span class="comment">####各小组杂拟谷盗的头数</span></span><br><span class="line">NUMBER_OF_BEETLES&lt;-<span class="built_in">c</span>(<span class="number">59</span>,    <span class="number">60</span>,    <span class="number">62</span>,    <span class="number">56</span>,    <span class="number">63</span>,    <span class="number">59</span>,    <span class="number">62</span>,    <span class="number">60</span>    )    </span><br><span class="line"><span class="comment">####杂拟谷盗经过二硫化碳熏蒸5小时后，死亡的个体数</span></span><br><span class="line">NUMBER_KILLED    &lt;-<span class="built_in">c</span>(<span class="number">6</span>,     <span class="number">13</span>,    <span class="number">18</span>,    <span class="number">28</span>,    <span class="number">52</span>,    <span class="number">53</span>,    <span class="number">61</span>,    <span class="number">60</span>    )</span><br><span class="line"><span class="comment">####合并成一个数据框</span></span><br><span class="line">beetle&lt;-data.frame(NUMBER_OF_BEETLES,NUMBER_KILLED, LOGDOSE)</span><br><span class="line"><span class="comment">####第一种方法，数据表格式</span></span><br><span class="line"><span class="comment">####调用glm函数，注意连接函数的选择</span></span><br><span class="line">mod_logit  &lt;-glm(cbind(NUMBER_KILLED, NUMBER_OF_BEETLES -NUMBER_KILLED)~LOGDOSE,data=beetle,  family=binomial(link=<span class="string">&quot;logit&quot;</span>) )</span><br><span class="line">mod_probit &lt;-glm(cbind(NUMBER_KILLED, NUMBER_OF_BEETLES -NUMBER_KILLED)~LOGDOSE,data=beetle,  family=binomial(link=<span class="string">&quot;probit&quot;</span>))</span><br><span class="line">mod_cloglog&lt;-glm(cbind(NUMBER_KILLED, NUMBER_OF_BEETLES -NUMBER_KILLED)~LOGDOSE,data=beetle,  family=binomial(link=<span class="string">&quot;cloglog&quot;</span>))</span><br><span class="line"><span class="comment">####Showing significance of the variables</span></span><br><span class="line">tidy(mod_logit )</span><br><span class="line"></span><br><span class="line">\<span class="comment">##         term  estimate std.error statistic      p.value</span></span><br><span class="line">\<span class="comment">##1 (Intercept) -60.74013  5.181870 -11.72166 9.871561e-32</span></span><br><span class="line">\<span class="comment">##2     LOGDOSE  34.28593  2.913213  11.76911 5.631446e-32</span></span><br><span class="line"></span><br><span class="line">tidy(mod_probit)</span><br><span class="line"></span><br><span class="line">\<span class="comment">##         term  estimate std.error statistic      p.value</span></span><br><span class="line">\<span class="comment">##1 (Intercept) -34.95614  2.648984 -13.19605 9.245343e-40</span></span><br><span class="line">\<span class="comment">##2     LOGDOSE  19.74103  1.488046  13.26640 3.625628e-40</span></span><br><span class="line"></span><br><span class="line">tidy(mod_cloglog)</span><br><span class="line"></span><br><span class="line">\<span class="comment">##         term  estimate std.error statistic      p.value</span></span><br><span class="line">\<span class="comment">##1 (Intercept) -39.52224  3.235649 -12.21463 2.596958e-34</span></span><br><span class="line">\<span class="comment">##2     LOGDOSE  22.01474  1.796996  12.25085 1.662370e-34</span></span><br><span class="line"></span><br><span class="line">\<span class="comment">####Showing AIC et al.</span></span><br><span class="line">glance(mod_logit )</span><br><span class="line"></span><br><span class="line">\<span class="comment">##  null.deviance df.null    logLik      AIC     BIC deviancedf.residual</span></span><br><span class="line">\<span class="comment">##1      284.2024       7 -18.65681 41.31361 41.4725 11.11558          6</span></span><br><span class="line"></span><br><span class="line">glance(mod_probit)</span><br><span class="line"></span><br><span class="line">\<span class="comment">##  null.deviance df.null   logLik      AIC      BIC deviancedf.residual</span></span><br><span class="line">\<span class="comment">##1      284.2024       7 -18.0925 40.18499 40.34388 9.986957          6</span></span><br><span class="line"></span><br><span class="line">glance(mod_cloglog)</span><br><span class="line"></span><br><span class="line">\<span class="comment">##  null.deviance df.null    logLik      AIC      BIC deviancedf.residual</span></span><br><span class="line">\<span class="comment">##1      284.2024       7 -14.85619 33.71237 33.87126 3.514334          6</span></span><br><span class="line"></span><br><span class="line">\<span class="comment">####绘图Visualisingthe results</span></span><br><span class="line">mortality_rate&lt;-beetle$NUMBER_KILLED/beetle$NUMBER_OF_BEETLES</span><br><span class="line">dose&lt;-beetle$LOGDOSE</span><br><span class="line">plot(mortality_rate~dose,xlab=<span class="string">&quot;Dose of Carbon Disulfide&quot;</span>,ylab=<span class="string">&quot;MortalityRate of Beetles&quot;</span>,main=<span class="string">&quot;BinomialRegression&quot;</span>)</span><br><span class="line">LOGDOSE_new=seq(<span class="built_in">min</span>(dose),<span class="built_in">max</span>(dose),length.out=<span class="number">200</span>)</span><br><span class="line">mod_logit_line   &lt;-predict(mod_logit,newdata=data.frame(LOGDOSE=LOGDOSE_new), type=<span class="string">&quot;response&quot;</span>)</span><br><span class="line">lines(mod_logit_line~LOGDOSE_new,col=<span class="string">&quot;blue&quot;</span>)</span><br><span class="line">mod_probit_line  &lt;-predict(mod_probit,newdata=data.frame(LOGDOSE=LOGDOSE_new), type=<span class="string">&quot;response&quot;</span>)</span><br><span class="line">lines(mod_probit_line~LOGDOSE_new,col=<span class="string">&quot;red&quot;</span>)</span><br><span class="line">mod_cloglog_line &lt;-predict(mod_cloglog,newdata=data.frame(LOGDOSE=LOGDOSE_new), type=<span class="string">&quot;response&quot;</span>)</span><br><span class="line">lines(mod_cloglog_line~LOGDOSE_new,col=<span class="string">&quot;green&quot;</span>)</span><br><span class="line">legend(<span class="string">&quot;bottomright&quot;</span>,title=<span class="string">&quot;Linkfunctions&quot;</span>,legend=<span class="built_in">c</span>(<span class="string">&quot;logit&quot;</span>,<span class="string">&quot;probit&quot;</span>,<span class="string">&quot;cloglog&quot;</span>),lty=<span class="built_in">c</span>(<span class="number">1</span>,<span class="number">1</span>,<span class="number">1</span>),col=<span class="built_in">c</span>(<span class="string">&quot;blue&quot;</span>,<span class="string">&quot;red&quot;</span>,<span class="string">&quot;green&quot;</span>))</span><br></pre></td></tr></table></figure>


<p><a target="_blank" rel="noopener" href="http://image.sciencenet.cn/home/201710/19/235751xnqz1qtzcz2annqp.jpeg"><img src="http://image.sciencenet.cn/home/201710/19/235751xnqz1qtzcz2annqp.jpeg" alt="img"></a></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br></pre></td><td class="code"><pre><span class="line">\<span class="comment">#####第二种方法，每个个体生存状态的长数据格式。glm是可以接受这种格式的。</span></span><br><span class="line">\<span class="comment">#####按照个体来计</span></span><br><span class="line">dose_surived&lt;-</span><br><span class="line"><span class="built_in">c</span>(<span class="built_in">rep</span>(<span class="number">1.691</span>,<span class="number">59</span>-<span class="number">6</span>),         </span><br><span class="line"><span class="built_in">rep</span>(<span class="number">1.724</span>,<span class="number">60</span>-<span class="number">13</span>),         </span><br><span class="line"><span class="built_in">rep</span>(<span class="number">1.755</span>,<span class="number">62</span>-<span class="number">18</span>),         </span><br><span class="line"><span class="built_in">rep</span>(<span class="number">1.784</span>,<span class="number">56</span>-<span class="number">28</span>),         </span><br><span class="line"><span class="built_in">rep</span>(<span class="number">1.811</span>,<span class="number">63</span>-<span class="number">52</span>),         </span><br><span class="line"><span class="built_in">rep</span>(<span class="number">1.837</span>,<span class="number">59</span>-<span class="number">53</span>),         </span><br><span class="line"><span class="built_in">rep</span>(<span class="number">1.861</span>,<span class="number">62</span>-<span class="number">61</span>),         </span><br><span class="line"><span class="built_in">rep</span>(<span class="number">1.884</span>,<span class="number">60</span>-<span class="number">60</span>))</span><br><span class="line">dose_dead&lt;-<span class="built_in">c</span>(</span><br><span class="line"><span class="built_in">rep</span>(<span class="number">1.691</span>, <span class="number">6</span>),</span><br><span class="line"><span class="built_in">rep</span>(<span class="number">1.724</span>,<span class="number">13</span>),</span><br><span class="line"><span class="built_in">rep</span>(<span class="number">1.755</span>,<span class="number">18</span>),</span><br><span class="line"><span class="built_in">rep</span>(<span class="number">1.784</span>,<span class="number">28</span>),</span><br><span class="line"><span class="built_in">rep</span>(<span class="number">1.811</span>,<span class="number">52</span>),</span><br><span class="line"><span class="built_in">rep</span>(<span class="number">1.837</span>,<span class="number">53</span>),</span><br><span class="line"><span class="built_in">rep</span>(<span class="number">1.861</span>,<span class="number">61</span>),</span><br><span class="line"><span class="built_in">rep</span>(<span class="number">1.884</span>,<span class="number">60</span>))</span><br><span class="line">status_survived&lt;-<span class="built_in">c</span>(</span><br><span class="line"><span class="built_in">rep</span>(<span class="string">&quot;alive&quot;</span>,<span class="number">59</span>-<span class="number">6</span>),</span><br><span class="line"><span class="built_in">rep</span>(<span class="string">&quot;alive&quot;</span>,<span class="number">60</span>-<span class="number">13</span>),</span><br><span class="line"><span class="built_in">rep</span>(<span class="string">&quot;alive&quot;</span>,<span class="number">62</span>-<span class="number">18</span>),</span><br><span class="line"><span class="built_in">rep</span>(<span class="string">&quot;alive&quot;</span>,<span class="number">56</span>-<span class="number">28</span>),</span><br><span class="line"><span class="built_in">rep</span>(<span class="string">&quot;alive&quot;</span>,<span class="number">63</span>-<span class="number">52</span>),</span><br><span class="line"><span class="built_in">rep</span>(<span class="string">&quot;alive&quot;</span>,<span class="number">59</span>-<span class="number">53</span>),</span><br><span class="line"><span class="built_in">rep</span>(<span class="string">&quot;alive&quot;</span>,<span class="number">62</span>-<span class="number">61</span>),</span><br><span class="line"><span class="built_in">rep</span>(<span class="string">&quot;alive&quot;</span>,<span class="number">60</span>-<span class="number">60</span>))</span><br><span class="line">status_dead&lt;-<span class="built_in">c</span>(</span><br><span class="line"><span class="built_in">rep</span>(<span class="string">&quot;dead&quot;</span>, <span class="number">6</span>),</span><br><span class="line"><span class="built_in">rep</span>(<span class="string">&quot;dead&quot;</span>,<span class="number">13</span>),</span><br><span class="line"><span class="built_in">rep</span>(<span class="string">&quot;dead&quot;</span>,<span class="number">18</span>),</span><br><span class="line"><span class="built_in">rep</span>(<span class="string">&quot;dead&quot;</span>,<span class="number">28</span>),</span><br><span class="line"><span class="built_in">rep</span>(<span class="string">&quot;dead&quot;</span>,<span class="number">52</span>),</span><br><span class="line"><span class="built_in">rep</span>(<span class="string">&quot;dead&quot;</span>,<span class="number">53</span>),</span><br><span class="line"><span class="built_in">rep</span>(<span class="string">&quot;dead&quot;</span>,<span class="number">61</span>),</span><br><span class="line"><span class="built_in">rep</span>(<span class="string">&quot;dead&quot;</span>,<span class="number">60</span>)</span><br><span class="line">)</span><br><span class="line"></span><br><span class="line">dose_long&lt;-<span class="built_in">c</span>(dose_dead,dose_surived)</span><br><span class="line">status   &lt;-as.factor(<span class="built_in">c</span>(status_dead,status_survived))</span><br><span class="line"></span><br><span class="line">incidence_mod_logit  &lt;-glm(status~dose_long,family=binomial(link=<span class="string">&quot;logit&quot;</span>))</span><br><span class="line">incidence_mod_probit &lt;-glm(status~dose_long,family=binomial(link=<span class="string">&quot;probit&quot;</span>))</span><br><span class="line">incidence_mod_cloglog&lt;-glm(status~dose_long,family=binomial(link=<span class="string">&quot;cloglog&quot;</span>))</span><br><span class="line"></span><br><span class="line">\<span class="comment">####Showing significance of the variables</span></span><br><span class="line">tidy(incidence_mod_logit )</span><br><span class="line"></span><br><span class="line">\<span class="comment">##         term  estimate std.error statistic      p.value</span></span><br><span class="line">\<span class="comment">##1 (Intercept) -60.74013  5.181834 -11.72174 9.861985e-32</span></span><br><span class="line">\<span class="comment">##2   dose_long  34.28593  2.913192  11.76920 5.625873e-32</span></span><br><span class="line"></span><br><span class="line">tidy(incidence_mod_probit)</span><br><span class="line"></span><br><span class="line">\<span class="comment">##         term  estimate std.error statistic      p.value</span></span><br><span class="line">\<span class="comment">##1 (Intercept) -34.95613  2.649010 -13.19592 9.261453e-40</span></span><br><span class="line">\<span class="comment">##2   dose_long  19.74102  1.488061  13.26627 3.631968e-40</span></span><br><span class="line"></span><br><span class="line">tidy(incidence_mod_cloglog)</span><br><span class="line"></span><br><span class="line">\<span class="comment">##         term  estimate std.error statistic      p.value</span></span><br><span class="line">\<span class="comment">##1 (Intercept) -39.52224  3.235632 -12.21469 2.594819e-34</span></span><br><span class="line">\<span class="comment">##2   dose_long  22.01474  1.796986  12.25092 1.660982e-34</span></span><br><span class="line"></span><br><span class="line">\<span class="comment">####Showing AIC et al.</span></span><br><span class="line">glance(incidence_mod_logit )</span><br><span class="line"></span><br><span class="line">\<span class="comment">##  null.deviance df.null    logLik      AIC      BIC deviancedf.residual</span></span><br><span class="line">\<span class="comment">##1       645.441     480 -186.1771 376.3542 384.7059 372.3542        479</span></span><br><span class="line"></span><br><span class="line">glance(incidence_mod_probit)</span><br><span class="line"></span><br><span class="line">\<span class="comment">##  null.deviance df.null    logLik      AIC      BIC deviancedf.residual</span></span><br><span class="line">\<span class="comment">##1       645.441     480 -185.6128 375.2255 383.5773 371.2255        479</span></span><br><span class="line"></span><br><span class="line">glance(incidence_mod_cloglog)</span><br><span class="line"></span><br><span class="line">\<span class="comment">##  null.deviance df.null    logLik      AIC      BIC deviancedf.residual</span></span><br><span class="line">\<span class="comment">##1       645.441     480 -182.3765 368.7529 377.1046 364.7529        479</span></span><br></pre></td></tr></table></figure>
      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2017/10/19/2017-10-19-glm/" data-id="ckrsvkzau008190w059wqclxy" data-title="广义线性模型中的逻辑斯蒂回归" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E7%BB%9F%E8%AE%A1/" rel="tag">统计</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2017-10-18-plotsinfdp" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2017/10/18/2017-10-18-plotsinfdp/" class="article-date">
  <time class="dt-published" datetime="2017-10-18T00:19:54.000Z" itemprop="datePublished">2017-10-18</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2017/10/18/2017-10-18-plotsinfdp/">大型森林样地样方的命名与坐标转换</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>大型森林监测样地一般是指面积在16公顷以上的样地，在样地内部， 所有胸径大于1cm的木本植物及其分枝都要测量平面直角坐标，胸径等。在调查过程中，一般是先确定样地范围，然后用全站仪打点，确定样地的四角，以及每个20m样方顶点位置，为了方便计算，一般以西南点为原点， x轴指向正东，向东为正，单位为m；y轴指向正北，向北为正，单位为m。</p>
<p>为了调查方便，一般先将样地划分成20m的样方，每个样方按照在xy轴出现的顺序命名。名字一般为4位，例如0518，前两位表示从原点开始算，向东计第5个20m样方，后两位表示从原点开始计，向北计第18个样方。因此，大样地西南角的样方名总是0101。对于东西方向宽400m，南北方向长500m的大样地，东北角的样方名为2025，西北角的样方名为0125，东南角的样方名为2001。</p>
<p>要在20m样方内部测量每棵树的xy坐标仍然困难。所以实际调查中一般是将20m的样方再进一步划分成16个5m*5m的小样方，每个5m的小样方均以自己的西南点为原点，xy轴的方向与20m样方相同。这样在调查中就可以较方便的测量每棵树的坐标。每个20m样地内的16个小样方一般按照蛇形顺序调查，以减少调查人员在样地内的走动。5m小样方习惯上一般命名为1到16。</p>
<p>按照这种调查方法，从野外调查获得的每棵树坐标位置实际上为该个体在5m小样方内的坐标。在生态学数据分析中，要进行空间分析、点格局分析、变换空间尺度时，需要每棵树在样地尺度的平面直角xy坐标。这就需要将5m的坐标按照样方名称等，转换为样地尺度的xy坐标。这也是大样地数据标准化的重要一步。</p>
<p>本文以一个东西宽400m*南北长500m的样地为例，简要介绍如何将调查数据转换为样地尺度。先生成模拟数据，再基于模拟数据演示如何处理。</p>
<p>调查后输入Excel的数据，20m样方的名称通常为0105, 0203, 0505，不过excel常常将名称自动转换为105, 203, 505等只包含三位数字；而有些样方号并不以0开头，例如1820, 2022，这些样方在Excel自动转换后包含四位数字。由于这会带来处理上的不便，所以最好统一为4位，在R中可以用stringr程序包的str_pad 函数，在3位数字的字符串前面加0，转换为0105，0203，0505等。</p>
<p>下面我们生成一套模拟数据，假设我们的样地设定在河南省内乡县宝天曼自然保护区，缩写为BTM。所有胸径大于1cm的，都已经挂牌，并已经测量了在5m样方内的横纵坐标位置。</p>
<p>假设样地内共调查到88325个个体，没有考虑分枝。每个个体有如下的信息：</p>
<ol>
<li>tag：树的号牌，原则上讲，从号牌上需要能看出每棵树所在的20m样方号，然后在每个样方号内顺序增加。因为这里是模拟数据，就暂时不考虑了。这里按照以下方式命名 BTM000001 - BTM088325；</li>
<li>plot20_names：该个体对应的20m样方名；</li>
<li>g5x：该个体在5m样方内的x坐标，单位为m；</li>
<li>g5y：该个体在5m样方内的y坐标，单位为m；</li>
<li>subplot5：该个体所在的小样方名称，按照蛇形顺序命名，从1-16；</li>
<li>dbh：该个体的胸径，单位为cm</li>
</ol>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#### 以下代码先生成模拟的调查数据</span></span><br><span class="line">library(stringr)</span><br><span class="line"><span class="comment">#### 生成模拟数据，假设整个样地有88325个个体</span></span><br><span class="line">nindividual &lt;-<span class="number">88325</span></span><br><span class="line"><span class="comment">#### 生成树牌号</span></span><br><span class="line">tag &lt;-paste(<span class="string">&quot;BTM&quot;</span>, str_pad(<span class="number">1</span>:nindividual, <span class="number">6</span>, pad =<span class="string">&quot;0&quot;</span>), sep =<span class="string">&quot;&quot;</span>)</span><br><span class="line"><span class="comment">#### 生成20m样地的名称</span></span><br><span class="line">plot20_x &lt;- str_pad(<span class="number">1</span>:<span class="number">20</span>, <span class="number">2</span>, pad =<span class="string">&quot;0&quot;</span>)</span><br><span class="line">plot20_y &lt;- str_pad(<span class="number">1</span>:<span class="number">25</span>, <span class="number">2</span>, pad =<span class="string">&quot;0&quot;</span>)</span><br><span class="line">names_temp &lt;- expand.grid(plot20_x, plot20_y)</span><br><span class="line">plot20_names &lt;-paste(names_temp[,<span class="number">1</span>],names_temp[,<span class="number">2</span>], sep =<span class="string">&quot;&quot;</span>)</span><br><span class="line">survey_plot20_names &lt;-sample(plot20_names,nindividual, replace=<span class="literal">TRUE</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">#### 随机生成5m样地内，调查个体的坐标</span></span><br><span class="line">g5x &lt;-<span class="built_in">round</span>(runif(nindividual)*<span class="number">5</span>, <span class="number">2</span>)</span><br><span class="line">g5y &lt;-<span class="built_in">round</span>(runif(nindividual)*<span class="number">5</span>, <span class="number">2</span>)</span><br><span class="line"><span class="comment">#### 生成5m小样方的名称</span></span><br><span class="line">subplot5 &lt;-sample(<span class="number">1</span>:<span class="number">16</span>, nindividual, replace=<span class="literal">TRUE</span>)</span><br><span class="line"><span class="comment">#### 模拟生成每棵树的胸径dbh, 假设服从指数分布</span></span><br><span class="line">dbh &lt;-<span class="built_in">round</span>(rexp(nindividual,rate =<span class="number">2</span>)*<span class="number">50</span>+<span class="number">1</span>, <span class="number">1</span>)</span><br><span class="line"><span class="comment">#### 所有生成的数据合并到一个dataframe中</span></span><br><span class="line">btmdat &lt;- data.frame(tag, survey_plot20_names, subplot5, g5x, g5y, dbh)</span><br><span class="line"><span class="comment">#### 做简单排序</span></span><br><span class="line">btmdat &lt;- btmdat[order(btmdat$survey_plot20_names, btmdat$subplot5),]</span><br><span class="line"></span><br><span class="line"><span class="comment">############## 以上生成模拟数据结束 ####################</span></span><br><span class="line"><span class="comment">#### 以下代码用于转换5m坐标到大样地尺度的平面直角坐标</span></span><br><span class="line"><span class="comment">#######################################################</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment">#### 生成定义5m样方在20m样方x位置的中间变量</span></span><br><span class="line">btmdat$location5x &lt;-<span class="built_in">rep</span>(<span class="literal">NA</span>, nrow(btmdat))</span><br><span class="line">btmdat$location5y &lt;-<span class="built_in">rep</span>(<span class="literal">NA</span>, nrow(btmdat))</span><br><span class="line"><span class="comment">#### 因为是蛇形调查顺序，所以5m的样方名要分别找出x轴和y轴的对应编号</span></span><br><span class="line">btmdat$location5x[btmdat$subplot5 ==<span class="number">1</span>]&lt;-<span class="number">1</span></span><br><span class="line">btmdat$location5x[btmdat$subplot5 ==<span class="number">2</span>]&lt;-<span class="number">2</span></span><br><span class="line">btmdat$location5x[btmdat$subplot5 ==<span class="number">3</span>]&lt;-<span class="number">3</span></span><br><span class="line">btmdat$location5x[btmdat$subplot5 ==<span class="number">4</span>]&lt;-<span class="number">4</span></span><br><span class="line">btmdat$location5x[btmdat$subplot5 ==<span class="number">5</span>]&lt;-<span class="number">4</span></span><br><span class="line">btmdat$location5x[btmdat$subplot5 ==<span class="number">6</span>]&lt;-<span class="number">3</span></span><br><span class="line">btmdat$location5x[btmdat$subplot5 ==<span class="number">7</span>]&lt;-<span class="number">2</span></span><br><span class="line">btmdat$location5x[btmdat$subplot5 ==<span class="number">8</span>]&lt;-<span class="number">1</span></span><br><span class="line">btmdat$location5x[btmdat$subplot5 ==<span class="number">9</span>]&lt;-<span class="number">1</span></span><br><span class="line">btmdat$location5x[btmdat$subplot5 ==<span class="number">10</span>]&lt;-<span class="number">2</span></span><br><span class="line">btmdat$location5x[btmdat$subplot5 ==<span class="number">11</span>]&lt;-<span class="number">3</span></span><br><span class="line">btmdat$location5x[btmdat$subplot5 ==<span class="number">12</span>]&lt;-<span class="number">4</span></span><br><span class="line">btmdat$location5x[btmdat$subplot5 ==<span class="number">13</span>]&lt;-<span class="number">4</span></span><br><span class="line">btmdat$location5x[btmdat$subplot5 ==<span class="number">14</span>]&lt;-<span class="number">3</span></span><br><span class="line">btmdat$location5x[btmdat$subplot5 ==<span class="number">15</span>]&lt;-<span class="number">2</span></span><br><span class="line">btmdat$location5x[btmdat$subplot5 ==<span class="number">16</span>]&lt;-<span class="number">1</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#### 生成定义5m样方在20m样方y位置的中间变量</span></span><br><span class="line">btmdat$location5y[btmdat$subplot5 ==<span class="number">1</span>]&lt;-<span class="number">1</span></span><br><span class="line">btmdat$location5y[btmdat$subplot5 ==<span class="number">2</span>]&lt;-<span class="number">1</span></span><br><span class="line">btmdat$location5y[btmdat$subplot5 ==<span class="number">3</span>]&lt;-<span class="number">1</span></span><br><span class="line">btmdat$location5y[btmdat$subplot5 ==<span class="number">4</span>]&lt;-<span class="number">1</span></span><br><span class="line">btmdat$location5y[btmdat$subplot5 ==<span class="number">5</span>]&lt;-<span class="number">2</span></span><br><span class="line">btmdat$location5y[btmdat$subplot5 ==<span class="number">6</span>]&lt;-<span class="number">2</span></span><br><span class="line">btmdat$location5y[btmdat$subplot5 ==<span class="number">7</span>]&lt;-<span class="number">2</span></span><br><span class="line">btmdat$location5y[btmdat$subplot5 ==<span class="number">8</span>]&lt;-<span class="number">2</span></span><br><span class="line">btmdat$location5y[btmdat$subplot5 ==<span class="number">9</span>]&lt;-<span class="number">3</span></span><br><span class="line">btmdat$location5y[btmdat$subplot5 ==<span class="number">10</span>]&lt;-<span class="number">3</span></span><br><span class="line">btmdat$location5y[btmdat$subplot5 ==<span class="number">11</span>]&lt;-<span class="number">3</span></span><br><span class="line">btmdat$location5y[btmdat$subplot5 ==<span class="number">12</span>]&lt;-<span class="number">3</span></span><br><span class="line">btmdat$location5y[btmdat$subplot5 ==<span class="number">13</span>]&lt;-<span class="number">4</span></span><br><span class="line">btmdat$location5y[btmdat$subplot5 ==<span class="number">14</span>]&lt;-<span class="number">4</span></span><br><span class="line">btmdat$location5y[btmdat$subplot5 ==<span class="number">15</span>]&lt;-<span class="number">4</span></span><br><span class="line">btmdat$location5y[btmdat$subplot5 ==<span class="number">16</span>]&lt;-<span class="number">4</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#### 将20m样方的名称补充完整为4位</span></span><br><span class="line">btmdat$survey_plot20_names &lt;- <span class="built_in">as.character</span>(str_pad(btmdat$survey_plot20_names, <span class="number">4</span>, pad =<span class="string">&quot;0&quot;</span>))</span><br><span class="line"><span class="comment">#### 提取20m样方名的前两位，作为20m样方的x坐标</span></span><br><span class="line">x20 &lt;-substr(btmdat$survey_plot20_names,start=<span class="number">1</span>, stop=<span class="number">2</span>)</span><br><span class="line"><span class="comment">#### 提取20m样方名的后两位，作为20m样方的y坐标</span></span><br><span class="line">y20 &lt;-substr(btmdat$survey_plot20_names,start=<span class="number">3</span>, stop=<span class="number">4</span>)</span><br><span class="line"><span class="comment">#### 计算每个个体的以大样地原点为原点的坐标</span></span><br><span class="line"><span class="comment">#### 无论是x轴，还是y轴，注意这里要-1</span></span><br><span class="line">btmdat$gx &lt;- g5x +(btmdat$location5x -<span class="number">1</span>)*<span class="number">5</span>+(<span class="built_in">as.numeric</span>(x20)-<span class="number">1</span>)*<span class="number">20</span></span><br><span class="line">btmdat$gy &lt;- g5y +(btmdat$location5y -<span class="number">1</span>)*<span class="number">5</span>+(<span class="built_in">as.numeric</span>(y20)-<span class="number">1</span>)*<span class="number">20</span></span><br><span class="line"><span class="comment">#### 去掉不需要的中间变量</span></span><br><span class="line">final_btmdat &lt;-subset(btmdat, select =!colnames(btmdat)%in%<span class="built_in">c</span>(<span class="string">&quot;location5x&quot;</span>, <span class="string">&quot;location5y&quot;</span>))</span><br><span class="line">head(final_btmdat)</span><br><span class="line">summary(final_btmdat)</span><br></pre></td></tr></table></figure>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2017/10/18/2017-10-18-plotsinfdp/" data-id="ckrsvkzat007z90w05oyjegt1" data-title="大型森林样地样方的命名与坐标转换" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E7%94%9F%E6%80%81/" rel="tag">生态</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2017-10-7-english" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2017/10/07/2017-10-7-english/" class="article-date">
  <time class="dt-published" datetime="2017-10-07T05:53:54.000Z" itemprop="datePublished">2017-10-07</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2017/10/07/2017-10-7-english/">学英语记</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>自从初中一年级，接触Good Morning， Mr. Green 以来， 这么多年一直在学英语， 虽然从未学到多好。那时候课本是人民教育出版社的， Li Lei， Han Meimei，Lucy， Lily， Susan， Sue还有鹦鹉Poly：大家去摘苹果、过生日、放风筝、郊游、购物….第一次听到磁带中的中国人读英语， 虽然感觉发音都很正确， 总感觉缺点儿老外读的那种感觉。 特别是有一课， 有一个中国小孩儿问外国老太太的年龄， 老太太说， That’s a secret！大家都在笑老太太的发音。老师说， 在国外， 问人家年龄、是否结婚、一个月挣多少钱都是非常不礼貌的。这是文化的差异， 事情虽然很小， 我第一次知道还有这种忌讳。</p>
<p>英语难么？ 其实并没有觉得， 初中时要学的单词那么少， 对话那么简单，很快就全部掌握了。 而且小测和考试， 我还常常能得全班最高分， 有时候甚至是满分。</p>
<p>不得不提的，是中央电视台当时播了一个叫《走向未来》的英语教学节目。这个教程从Nice to meet you 开始，每期一个主题， 穿插情景对话， 语法讲解，访谈，动画以及情景小品， 从开始最简单，到最后到相当高深的内容， 如议院、探险、石油钻井平台的工作、海关人员的访谈等。 我第一次感觉到课堂上的英语与生活中的英语有那么大的差别。 于是不惜将英语听力练习的磁带洗过， 守在黑白电视机前录制《走向未来》，一遍又一遍地听， 模仿BBC的口音。这是我后来英语发音的基础。我真切感受到， 那才是接近真实的英语， 就应该那么自然、流畅。差不多二十年后，2014年我去新加坡参加森林样地数据分析培训， 当时一位菲律宾的学员问我是在哪儿拿的博士， 我说是在北京， 她很惊讶，继续问，那为什么你英语发音很英式？ 我说这主要是因为我跟着BBC的一个节目学的英语。</p>
<p>其实中学课本上的英语是相对有意思的， 但是相关的练习就非常枯燥。我甚至怀疑英语是否应该通过那样做题来学。 听力、选择、完形填空、阅读、作文， 难道真的能考察一个人的英语水平？ 难道只有那样一遍一遍地做题， 人才能进步？</p>
<p>英语听力有时候也要绕一个圈子： 一个人买了去伦敦的车票， 然后迟到，没有赶上车。听力选项要问你， 这个人是不是在伦敦。说真的，到现在也没有去过伦敦，而且相信去过伦敦的初中同学也没几个。 我们真的不在乎那个没赶上车的人在哪儿。 但是这是听力题， 你听不懂， 就答不对，然后就可以跟班级前三名Bye bye了。</p>
<p>最难的是完形填空， 要想出合适的单词， 把一个完全不熟悉的故事编出来， 然后自己还要相信。 完形填空相信是很多人学英语的噩梦。 我也没有并没有什么高招。</p>
<p>作文还好，写点儿想写的话，无奈没什么书法天分， 总是觉得自己写的英文太难看。</p>
<p>听力、 选择、完形填空这样学英语的模式一直延续到中考、高考、博士入学考试。 一直到考博士时， 我还在学习那60课的《走向未来》。倒不是学得太慢： 这次是找到了视频， 将所有的采访内容， 一句一句写下来， 一遍听不懂， 再继续听， 最后将听写下来的内容， 背到滚瓜烂熟。当然， 这都是后来到北京的事情了。</p>
<p>高考后，到大学报到， 刚到大学门口， 有多失望就别提了： 我从来没有见过那种脏乱差。 虽然从小生活在农村， 但那时乡下各处的生活垃圾基本上是可以天然降解的， 所以并不是特别脏。 学校位于城中村，公滨路木材街上，下雨后门口泥泞不堪。 我看到学校的大门甚至有回家复读的冲动， 不过无论是家庭经济情况， 还是心理承受能力， 都不允许。</p>
<p>大学的英语课相当轻松， 考试也轻松。 平实成绩占很大一部分，所以但凡不缺课， 成绩差不多就通过期末考试了，至于有没有提高英语水平， 就见仁见智。 我们用的是浙江大学的教材，非常难。难的意思就是，我从来没有读懂过任何一篇课文，当然也对这套书没有任何印象。 不记得英语课是怎么上的， 试是怎么考的。大学时有一位美国还是加拿大的老太太做外教， 每个学生都要起一个英文名， 我不记得为什么给自己起了一个叫Tony的英文名，大概是与常见的随身听Sony有关吧， 要不就是受到莎士比亚威尼斯商人里面的安东尼奥启发， 觉得Tony这个名字不错， 因为Tony是Anthoy的昵称。 起了个英文名， 并不代表自己的英文就有多大长进了，相反， 还真没觉得有什么进步，Tony这个英文名自从上完那学期英语课后就没再用过。教辅书充斥着每个学生的生活，不知道今天是不是还是如此， 什么《英语周报》，王长喜英语练习，星火四六级， 以及后来的新东方，疯狂英语等等等等。有些教材让学生不停做练习， 认为可以巩固语法、熟悉词汇，通过阅读拓展词汇；有些教材让学生不断重复、大声朗读演讲名篇， 什么马丁路德金《我有一个梦想》， 丘吉尔什么什么演说，以便找出学英语的感觉。现在看来，都很可笑。</p>
<p>踏踏实实能让人进步的， 就是《新概念英语1-4册》， 而且， 每一篇都要背诵，烂熟于心， 这样词汇量和对英文的语感自然就出来了。各种练习题其实都不必。</p>
<p>前几天同Lisa和金刚一起聊学英语的事。他们说， 自己都是只考了一次大学英语六级，就通过了。我相当佩服， 因为我没有那么厉害。所以这里说说我通过英语四六级的事。那时候， 英语四级与大学的学位证挂钩， 也不知道是谁制定的这个政策。</p>
<p>我考入的是国家生物学理科人才培养基地，也就是所谓的生物学基地班，只有一个班，28个人。那时候很多人相信， 二十一世纪是生物学和计算机的世纪，在那几年，生物学和计算机异常火爆，我也不小心上了生物学的船，从此慢慢走向了植物生态。 由于我们专业的学生基本上是全学校最高分，普遍比一本线高出50-80分左右，学校和老师们对我们也很照顾：前几届学生也都非常争气， 学习氛围很好。 学院当然对我们也寄予厚望。允许我们大一下学期报考英语四级。幸运的是，在没有复习的情况下，我的英语四级居然考过了。四级通过，就可以考六级。我那时对学校基础课的老师失望到极点， 情绪低落， 成绩也排在中下游。由于上课时每堂课都要点名， 所以我并不缺课， 但是自己在后排翻《中国高等植物图鉴》或者《东北植物检索表》等书， 也就是觉得这些专业书还有点儿意思。</p>
<p>大学的基础课和专业课的教材不好， 也是一个问题。 那时候考试， 基本上要从练习册上划定范围， 但是我对做练习非常抵触。 主要是因为大学的正式课本主要是农科教材， 而这些教材对各种公式的原理、来源等大都介绍的不明不白，装帧和印刷也很差。也就是说，课本本身都讲不明白， 做相应的练习只是为了记忆这些公式和原理，大学仍然用这种灌输式的教学方式， 枯燥是必然的。 我当时想， 知识肯定不是这么来的。 于是到大学的图书馆，去一探究竟。</p>
<p>万幸的是， 学校拥有一个藏书十分丰富的图书馆， 生物学和农学的藏书， 对于农业院校是十分重要的。不过话又说回来， 虽然是农业大学， 但是其实图书馆的藏书还是比较全面的， 文学、历史、艺术、音乐等各种书籍， 应有尽有。 去年去海南拜访大学时候曾资助我的李庆章校长， 李校长说， 他当时特别重视图书馆的建设， 如果学生对学校的师资力量不满意， 还可以通过图书馆去弥补， 所以他在任时极为重视图书馆图书的更新和保存。 校长的高瞻远瞩， 救了我这个不听话的学生， 谢天谢地。由于我常去图书馆， 和图书管理员都混熟了， 甚至有几次还一起出来吃饭。 我借了《有机化学》、《动物学》、《普通生物学》、《植物学系统分类学》的英文教科书，努力去阅读。 虽然是九十年代或者八十年代的教材，有些是影印版， 但是内容实在是太丰富了。 我本身对英文不够敏感，阅读这些教材， 其实并没有让我的英语有多少长进， 不过确实拓宽视野，当时最重要的印象就是英文的教材真好， 讲得真清楚。</p>
<p>后来读博士后， 我才切身意识到， 世界上最新和最重要的科学研究成果，绝大部分是用英文写成的。 不能熟练阅读英文的论文和论著， 研究不可能做好， 而且同国际同行交流都成问题， 所以学好英文是有志青年们要留意的， 否则虽然身处国家队， 却有变成民科的风险。</p>
<p>继续回到四六级考试。可能是由于大学英语的教材太难， 我提不起什么兴趣， 也许是过了英语四级已经满足了毕业的要求， 我有报名参加英语六级考试， 但是又没有怎么努力。 终于大二考了两次也没有考过。第三次考六级是在大三上学期，好像是下午两点开始考，我一点半还在宿舍睡觉，对考试也有一种心灰意冷的状态。 宿舍的老大万里兄说：快起来，去试一试， 你试一下， 还有过的可能，不试， 连过的可能也没有。报名费已经交了，为什么不试试？ 我想这说的很有道理。 于是抄起耳麦， 直奔考场。从宿舍到考场要20分钟。进入考场后找到自己的座位， 听力已经开始了。打开自己的电源， 发现耳麦没电，只能临时借了考场的备用耳机，开始答题。 也不知道是不是上天眷顾， 同样在考前没有做真题， 没有做模拟题， 直接裸考的我， 这次居然过了， 64.5分， 虽然刚刚过及格线。不过四六级这种考试， 通过了就好， 虽然自己英语水平其实并不怎么样。</p>
<p>由于我的硕士生导师李晶教授要出国， 帮我联系了在中科院植物所生物多样性与生物安全研究组联合培养。我还记得第一次见到自己将来的导师马克平研究员时的情景。马老师问：你英语六级过了吗？ 我说：大三的时候通过了。马老师点点头。 之所以这么问， 是因为上面已经提到， 英语在中科院系统太重要了。 每个要读博士的学生， 都要通过中科院自己的博士入学考试。单词量要求为9000单词， 相比之下， 六级是6000单词， 四级是4000单词。</p>
<p>临近博士入学考试， 我还是要努力扩展自己的词汇量， 也加入了背单词大军， 虽然长期以来对背单词很不屑。 其实背单词并不太枯燥。 我先是把不认识的单词抄写到一个小本子上， 然后回忆单词的意思，回忆不起来的， 又抄到另外一张纸， 几轮下来， 单词量基本上上去了。 做足准备后， 我仍然只看了看之前的真题的题型就去考试了。成绩还可以。 这样我就成了中科院的博士生。</p>
<p>读博的过程是辛苦的。首先是选题并研读英文的论文， 然后学习理论和相关分析技术， 自己的研究成果要有创新性， 并且要用英文写出来并且发表。 这是很大的挑战， 特别是周围没有人跟你很好的讨论时更是如此。 博士生们往往很迷茫， 需要导师更多的指导与爱护。</p>
<p>学习生态学， 数据分析技术当然非常重要， 而新的分析技术和方法， 新的理论，文献全都是英文的， 不得不一头扎进去。这样， 又是三年。 直到有一天， 我需要用英文汇报自己的成果， 忽然之间傻了眼。 研究组邀请了美国密歇根州立大学的Nate Swenson博士报告， 并且帮博士生修改论文。 我当时想表达一句话， “本研究的结论与之前的研究结论是一致的”， 居然不知道怎样用英语表达，急的冒汗。幸好当时华南植物园裴男才博士在场帮忙翻译， Nate才明白我想说什么。</p>
<p>当时还有一件事触动了我。</p>
<p>由于博士毕业面临找工作， 我有两个选择， 一个是中南大学湘雅校区做师资博后， 另一个是去香港嘉道理农场暨植物园应聘 Plant Identification and Records Officer。那时候我博士期间工作的SCI论文还没有发表， 当时达不到中南大学的要求， 但是面试成绩还不错， 学校比较满意， 给我保留了一年的期限， 这一年内发表一篇SCI论文， 即可入职。这种情况，我只能先想想在嘉道理农场面试应该注意什么，怎么准备。面试还有两天的时间， 我忽然意识到面试可能是英文。怎么办？ 总不能到时候卡在那里，眼睁睁看着工作机会从自己眼前溜走。于是我找了一些英文面试技巧， 并且认真准备了自我介绍以及常见问题的答案。大约50条。 谢天谢地， 面试时所有的问题， 都在我所准备的答案里面，所以那次面试应该还是给上司留下了不错的印象， 所以后来才能录用。那是我第一次面对两个金发碧眼的洋人， 面试官在一个小时里不断问你问题， 能否把握住机会， 就看自己怎么回答。 面试结束后， 部门主管很高兴， 还专门开车把我送回住处。最终我也被录取了。</p>
<p>虽然面试通过了， 不过我非常清楚自己的英语能力， 其实还停留很初步的阶段， 而且很多表达方式还很中式，口语更是一塌糊涂。</p>
<p>嘉道理的工作语言是英语。也许是对我特别关照， 全植物部二三十个坐办公室的同事，领导每个星期要和我面对面谈话超过8个小时，也就是每星期一天或者一天半的时间在跟我谈话。虽然说不上喜欢， 但是权当是英语听力练习吧。</p>
<p>经过不断的练习英语听力，去年的某一天， 我发现终于能不看字幕听懂英文纪录片和新闻，那种开心是很难形容的。后来还被公司找去做了两天中译英， 英译中的口语翻译。</p>
<p>二十年英语学习路，对于一个农村长大的人来说，艰辛而漫长。 学习英语的时间， 如果用来做别的事情， 能不能做得更好， 人生会不会更精彩？ 我不知道。 对我来说， 英语打开了看世界的窗户， 让我有机会认识外国的文化、外国的新技术，认识国外的朋友。</p>
<p>如果再问我一次， 你愿意用这么多时间学英语吗？ 我的答案是肯定的。</p>
<p>2017年10月7日于香港大埔</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2017/10/07/2017-10-7-english/" data-id="ckrsvkzaw008790w09mxqe4ws" data-title="学英语记" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2017-9-24-england" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2017/09/23/2017-9-24-england/" class="article-date">
  <time class="dt-published" datetime="2017-09-23T11:17:54.000Z" itemprop="datePublished">2017-09-23</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2017/09/23/2017-9-24-england/">雪花飘落在英格兰</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>雪花飘落在英格兰，</p>
<p>一位年轻人拿起了相机…</p>
<p>一个人的二十四小时，</p>
<p>想怎样支配就怎样支配</p>
<p>&nbsp;</p>
<p>三十天里，</p>
<p>总有阴晴圆缺…</p>
<p>从六年前的某一天起，</p>
<p>这里便是你的全部？</p>
<p>&nbsp;</p>
<p>生命只在一呼一吸，</p>
<p>你真的希望就那样老去？</p>
<p>我不信，</p>
<p>这世界上的人都一样孤独；</p>
<p>我不信，</p>
<p>秋天的风酿不成烈酒</p>
<p>&nbsp;</p>
<p>雪花也飘落在中国的大地上，</p>
<p>覆盖了最让人心疼的地方</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2017/09/23/2017-9-24-england/" data-id="ckrsvkzb8008z90w0eq4zhv5x" data-title="雪花飘落在英格兰" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E8%AF%97%E6%AD%8C/" rel="tag">诗歌</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2017-8-31-authoring" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2017/08/31/2017-8-31-authoring/" class="article-date">
  <time class="dt-published" datetime="2017-08-30T17:32:54.000Z" itemprop="datePublished">2017-08-31</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2017/08/31/2017-8-31-authoring/">对研究论文署名的一些观点</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>最近在论文署名上遇到一些困惑， 于是整理了一下个人观点。 这些原则或者观点当然毫无约束力， 但是， 对于不愿意遵循这些原则的同行， 合作是有难度的。 当然， 人微言轻， 这可能对学科没有任何影响。</p>
<p>以下是相应的原则或者观点：</p>
<ol>
<li><p>先有研究，后有论文，不能本末倒置。</p>
</li>
<li><p>研究论文不但是展示研究成果，更是展示作者的观点与思想。署名意味着自己对数据的准确性，方法的正确性，结果的准确性与文章的结论、观点都是认同，且能够负责的。</p>
</li>
<li><p>如果不同意论文的观点，不要署名。</p>
</li>
<li><p>文章第一稿的作者自动成为第一作者。无论稿件被别人做了多少修改，甚至部分内容重写，修改稿件的人都无权成为第一作者。修改论文功劳再大，都不能成为该稿件第一作者。</p>
</li>
<li><p>数据提供人和进行数据分析处理的人员必须列为作者。如果对方不同意列为作者，应获得书面确认，此时应在致谢中表示感谢。</p>
</li>
<li><p>通讯作者主要负责对论文的认真修改以及投稿，统筹回复审稿人意见等，对稿件的质量有最后把关的责任。审稿人需要负起这些责任。</p>
</li>
<li><p>论文作者只能是对论文撰写有直接贡献的人，最少要参与论文正文的修改。将对论文毫无直接贡献的人列入作者并不合适。</p>
</li>
<li><p>只对论文提出修改意见而未对论文做出修改的，一般只应列入致谢，不可列为作者。</p>
</li>
<li><p>论文只有一个第一作者，世界上并不存在共同第一作者。</p>
</li>
</ol>
<p>10.论文只有一个通讯作者，不允许出现两个通讯作者。</p>
<p>11.第一作者完成初稿后，不要急于按顺序列出作者，而是最好由各作者协商，在投稿前确定署名顺序。一般来说，贡献越小，作者的排名就越靠后。文章作者数不宜超过五个。</p>
<ol start="12">
<li> 遇到有意见分歧的情况，各贡献者之间协商解决，以免发生不快。</li>
</ol>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2017/08/31/2017-8-31-authoring/" data-id="ckrsvkzb6008v90w0dbvyautf" data-title="对研究论文署名的一些观点" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E5%86%99%E4%BD%9C/" rel="tag">写作</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2017-8-28-drum" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2017/08/27/2017-8-28-drum/" class="article-date">
  <time class="dt-published" datetime="2017-08-27T09:52:54.000Z" itemprop="datePublished">2017-08-27</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2017/08/27/2017-8-28-drum/">鼓--写给村里的侠客</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>广袤的华北平原的上，</p>
<p>处暑时节，</p>
<p>一盏灯熄灭了，</p>
<p>他的秋天也在这一天停止；</p>
<p>&nbsp;</p>
<p>川流不息的潮白河边，</p>
<p>在那个名不见经传的小村里，</p>
<p>扭秧歌的鼓声停了，</p>
<p>侠客的大鼓陷入了永久的沉默；</p>
<p>&nbsp;</p>
<p>他没有了秋天，也就没有了冬天和春天，</p>
<p>明年的大年初一，</p>
<p>人们就不用去吸他家里的烟气，</p>
<p>也不用再求他家墙上那被熏黄的菩萨保佑；</p>
<p>&nbsp;</p>
<p>大兴安岭、张家口、二连浩特，</p>
<p>在他的故事里总是年轻三十岁。</p>
<p>村里人都说他会武术，能飞檐走壁。</p>
<p>他说，“这辈子不知道应该坐多少次牢，从未失手……”；</p>
<p>&nbsp;</p>
<p>从明天开始，</p>
<p>世界上便没有人再记得他的“丰功伟绩”，</p>
<p>连村里人也不再认识他，</p>
<p>只低头看庄稼或者手机；</p>
<p>&nbsp;</p>
<p>侠客的鼓声停了，</p>
<p>侠客消失在河畔的夕阳里，</p>
<p>侠客再也不会回来，</p>
<p>就跟从未存在过一样。</p>
<p>&nbsp;</p>
<p>2017年8月25日于香港大埔</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2017/08/27/2017-8-28-drum/" data-id="ckrsvkzb6008t90w0fmntgj7n" data-title="鼓--写给村里的侠客" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E8%AF%97%E6%AD%8C/" rel="tag">诗歌</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2017-8-24-resize-image" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2017/08/24/2017-8-24-resize-image/" class="article-date">
  <time class="dt-published" datetime="2017-08-23T18:50:54.000Z" itemprop="datePublished">2017-08-24</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2017/08/24/2017-8-24-resize-image/">ImageMagick批量调整多个文件夹下照片的尺寸</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <ol>
<li> Install ImageMagick <a target="_blank" rel="noopener" href="https://www.imagemagick.org/script/index.php">https://www.imagemagick.org/script/index.php</a></li>
</ol>
<p>安装 ImageMagick </p>
<ol start="2">
<li> Make sure the command mogrify could be called from CMD</li>
</ol>
<p>安装时选择添加路径，以便mogrify组建能够从CMD调用。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#### Run the following R code</span></span><br><span class="line"></span><br><span class="line">运行以下R代码。location为照片所保存的文件夹。 </span><br><span class="line"></span><br><span class="line">location =<span class="string">&quot;D:/photos&quot;</span></span><br><span class="line">setwd(location)</span><br><span class="line"></span><br><span class="line">dirs &lt;- list.dirs(path = location, full.names =<span class="literal">TRUE</span>, recursive =<span class="literal">TRUE</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span>(i <span class="keyword">in</span> <span class="number">1</span>:<span class="built_in">length</span>(dirs))&#123;</span><br><span class="line">   temp &lt;- dirs[i]</span><br><span class="line">   setwd(temp)</span><br><span class="line">   system(<span class="string">&quot;mogrify -resize 2909x4000 -density 72 *.jpg&quot;</span>)</span><br><span class="line">   print(paste(<span class="string">&quot;Resizing&quot;</span>,temp))</span><br><span class="line">&#125;</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>更多选项参见 mogrify的帮助。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2017/08/24/2017-8-24-resize-image/" data-id="ckrsvkzb5008r90w0er0pal0f" data-title="ImageMagick批量调整多个文件夹下照片的尺寸" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E5%9B%BE%E5%83%8F%E5%A4%84%E7%90%86/" rel="tag">图像处理</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2017-8-5-partitionfinder" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2017/08/05/2017-8-5-partitionfinder/" class="article-date">
  <time class="dt-published" datetime="2017-08-04T19:14:54.000Z" itemprop="datePublished">2017-08-05</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2017/08/05/2017-8-5-partitionfinder/">用PartitionFinder2 筛选系统发育分析分隔模型</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>PartitionFinder2是用来在建立进化树时筛选最优进化模型以及分区（分隔）方案的软件，是用Python2语言编写的。</p>
<p>核苷酸序列在不同位点有不同的突变速率。核苷酸序列又分为编码基因和非编码基因。编码基因中，密码子第一、二位往往较为稳定，第三位往往变异速率较高。非编码基因因为受到的选择压力一般较小，所以往往可保留更多突变。不同基因以及不同位点的突变速率不同，可能对所推断进化树的稳定性有很大影响。所以在多基因建立进化树过程中，设置分隔模型就显得很重要。</p>
<p>但是分隔模型怎样设置才算合理呢？在PartitionFinder软件开发出来之前，研究人员一般通过将不同基因分开，例如gene1、gene2、gene3，再将每个基因的不同位点分开如gene1_1、gene1_2、gene1_3作为分隔模型，以提高进化树的稳健性。</p>
<p>RAxML、MrBayes、BEAST等常用系统发育软件都支持分隔模型，但并不能用于确定最优化的分隔模型设定方案。在建树过程中，如果设置的分隔模型过多，则拟合的参数会过多，造成结果不准确；如果设置的分隔模型过少，或者不设置分隔模型、设置的不合理，也会造成进化树不准确。很多学者已经意识到这个问题，但是一直苦于没有很好的应对方法。</p>
<p>赤池信息量AIC整合了模型的似然值（Likelihood）以及所要估计的参数数量。模型选择的简约理论认为，AIC或者BIC最小的模型是最优的模型。如果似然值已经足够高，进一步增加模型的参数已经不能再显著提高似然值的情况下，可以认为已经找到了最优模型。但是，对于分隔模型的筛选来说，碱基替换模型与分区方式的组合数量非常多，如果要对每一种分区方案对应的碱基模型筛选出最优组合，在很多情况下几乎是不可能的。这是因为：首先，从碱基比对矩阵计算Likelihood是十分耗费时间的。其次，分隔模型的各种组合的数量呈几何级数增长，如果通过几个基因建树，则可能的分隔模型的数量已经超出了大部分计算机的计算能力。此时就需要引入分隔模型的启发式搜索（Heuristic Search）。</p>
<p>PartitionFinder的作者从原理上解决了以上的问题，并通过Python语言实现了相应的算法（Lanfear et al., 2012）。作者证明，PartitionFinder所得的分隔模型比之前的简单处理更加合理。不仅如此，PartitionFinder在获得最优化分隔方案后，同时会给出每个分隔模块所对应的最优进化模型。因此，Modeltest、jModeltest以及ProtTest等软件都已经被PartitionFinder所超越。事实上，2012年Lanfear介绍PartitionFinder的论文发表后，在google scholar上已经被引用了3223次（截至2019年1月19日）。</p>
<p>本文简述PartitionFinder的安装和使用，供同行参考。</p>
<h1 id="1-下载和安装"><a href="#1-下载和安装" class="headerlink" title="1. 下载和安装"></a>1. 下载和安装</h1><h2 id="1-1-下载"><a href="#1-1-下载" class="headerlink" title="1.1 下载"></a>1.1 下载</h2><p>PartitionFinder的下载网址<a target="_blank" rel="noopener" href="http://www.robertlanfear.com/partitionfinder/%EF%BC%8C%E4%B8%8B%E8%BD%BD%E5%90%8E%E7%94%A8WinRAR%E6%88%96%E8%80%857zip%E8%A7%A3%E5%8E%8B%E7%BC%A9%E3%80%82">http://www.robertlanfear.com/partitionfinder/，下载后用WinRAR或者7zip解压缩。</a></p>
<p>图1 PartitionFinder的结构</p>
<p>其中:</p>
<ul>
<li> <code>docs</code>为说明文档</li>
<li> <code>Examples</code> 文件夹下为示例文件，包含DNA序列、形态以及氨基酸序列三个例子</li>
<li> <code>partfinder</code> 文件夹下为实现算法的 python脚本</li>
<li> <code>Programs</code> 文件夹下为PHYML可执行文件，用来计算likelihood</li>
<li> <code>Submodules</code> 文件夹下包含了raxml文件夹，但是该文件夹为空。作者并未交代这个文件夹的内容。</li>
<li> <code>tests</code> 文件夹下为开发时的测试文件</li>
<li> <code>PartitionFinder.py</code> 为筛选DNA序列最优分隔模型的Python脚本</li>
<li> <code>PartitionFinderMorphology.py</code> 为筛选形态数据最优分隔模型的Python脚本</li>
<li> <code>PartitionFinderProtein.py</code> 为筛选氨基酸序列最优分隔模型的Python脚本</li>
</ul>
<p>运行PartitionFinder时，要根据数据类型分别调用三个Python脚本，不能混淆。</p>
<h2 id="1-2-安装Anaconda"><a href="#1-2-安装Anaconda" class="headerlink" title="1.2. 安装Anaconda"></a>1.2. 安装Anaconda</h2><p>PartitionFinder是用Python2开发的，依赖于numpy、pandas、pytables、pyparsing、scipy、sklearn 等若干程序包。在Windows下，安装这些程序包往往较为麻烦。作者Lanfear推荐已经预装了这些程序包的Anaconda Python。</p>
<p>Anaconda Python的下载地址为：</p>
<p><a target="_blank" rel="noopener" href="https://www.anaconda.com/distribution/">https://www.anaconda.com/distribution/</a></p>
<p>PartitionFinder只能在Python2.7上运行，不能使用Python3。因此，这里只能下载Anaconda Python2.7。</p>
<p>双击开始安装。</p>
<p>选择将以下路径添加到系统路径:</p>
<p><code>C:\ProgramData\Anaconda2;C:\ProgramData\Anaconda2\Scripts; C:\ProgramData\Anaconda2\Library\bin;</code> 以便程序能从命令行CMD运行。</p>
<p>图2 下载Anaconda Python2.7</p>
<h1 id="2-所需文件"><a href="#2-所需文件" class="headerlink" title="2. 所需文件"></a>2. 所需文件</h1><p>运行PartitionFinder，用户需要提供：</p>
<ol>
<li> phylip文件：包含所要读取的序列</li>
<li> partition_finder.cfg文件：包含模型筛选时设定的参数</li>
</ol>
<h2 id="2-1-PHYLIP文件"><a href="#2-1-PHYLIP文件" class="headerlink" title="2.1 PHYLIP文件"></a>2.1 PHYLIP文件</h2><p>phylip文件可以通过MUSCLE、CLUSTAL、MAFFT等比对序列的软件生成。如果是多基因序列，一般对每个基因分别比对，手工调整之后，再通过Geneious或phylotools（<a target="_blank" rel="noopener" href="https://github.com/helixcn/phylotools%EF%BC%89%E7%AD%89%E8%BD%AF%E4%BB%B6%E7%94%9F%E6%88%90supermatrix%EF%BC%8C%E4%BB%A5phylip%E6%A0%BC%E5%BC%8F%E5%AF%BC%E5%87%BA%E5%8D%B3%E5%8F%AF">https://github.com/helixcn/phylotools</a><a target="_blank" rel="noopener" href="https://github.com/helixcn/phylotools%EF%BC%89%E7%AD%89%E8%BD%AF%E4%BB%B6%E7%94%9F%E6%88%90supermatrix%EF%BC%8C%E4%BB%A5phylip%E6%A0%BC%E5%BC%8F%E5%AF%BC%E5%87%BA%E5%8D%B3%E5%8F%AF">）等软件生成supermatrix，以phylip格式导出即可</a>。</p>
<p>PHYLIP格式的详细定义，参考</p>
<p><a target="_blank" rel="noopener" href="http://www.atgc-montpellier.fr/phyml/usersguide.php?type=phylip">http://www.atgc-montpellier.fr/phyml/usersguide.php?type=phylip</a>。</p>
<p>PartitionFinder可以接受relaxed phylip，文件中的物种名最长可以达100个字符。</p>
<p>图3 Phylip文件</p>
<h2 id="2-2-partition-finder-cfg-文件"><a href="#2-2-partition-finder-cfg-文件" class="headerlink" title="2.2 partition_finder.cfg 文件"></a>2.2 partition_finder.cfg 文件</h2><p>PartitionFinder可以处理三种类型的数据，分别为：</p>
<ol>
<li> 编码区DNA序列</li>
<li> 氨基酸序列或非编码区DNA序列</li>
<li> 形态数据</li>
</ol>
<p>根据数据类型的差别，partition_finder.cfg文件会各有不同。</p>
<ol>
<li> 编码区DNA序列的配置文件要提供每个片段的</li>
<li></li>
<li> 起始和终止位置;</li>
<li> 编码位置（1、2、3）</li>
<li> 氨基酸序列和非编码区DNA序列，只需提供每个基因片段的起始和终止位置。</li>
<li>形态数据，无需直接指定模块，在计算时采用k-means聚类等算法以划分成不同的组别。</li>
</ol>
<p>图4 partition_finder.cfg文件的内容</p>
<h1 id="3-运行"><a href="#3-运行" class="headerlink" title="3. 运行"></a>3. 运行</h1><p>对于DNA序列：</p>
<ol>
<li><p> 在partitionFinder文件夹下创建一个新文件夹，命名为test20190219</p>
</li>
<li><p> 将examples/nucleotide文件夹下的test.phy文件以及partition_finder.cfg配置文件拷贝到test20190219文件夹下</p>
</li>
<li><p> 在CMD中，将工作目录转换到partitionFinder下</p>
</li>
<li><p> 调用PartitionFinder.py，格式如下：</p>
</li>
</ol>
<p><code>python PartitionFinder.py test20190219</code></p>
<ol start="5">
<li><p> 生成的结果会自己保存在test20190219文件夹下，并生成若干新文件（如log.txt）以及文件夹analysis。</p>
</li>
<li><p> 模型筛选的结果就保存在analysis/best_scheme.txt文件中。</p>
</li>
</ol>
<p>图5 运行partitionFinder</p>
<p>图6 partitionFinder的运行结果</p>
<h1 id="4-cfg配置文件"><a href="#4-cfg配置文件" class="headerlink" title="4. cfg配置文件"></a>4. cfg配置文件</h1><p>数据类型不同的PHYLIP文件，需要在partition_finder.cfg文件中设置不同的参数。</p>
<h2 id="4-1-cfg文件的参数"><a href="#4-1-cfg文件的参数" class="headerlink" title="4.1 cfg文件的参数"></a>4.1 cfg文件的参数</h2><h3 id="4-1-1-alignment"><a href="#4-1-1-alignment" class="headerlink" title="4.1.1 alignment"></a>4.1.1 alignment</h3><p>指出phylip的文件名</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">alignment &#x3D; test.phy;</span><br></pre></td></tr></table></figure>

<h3 id="4-1-2-branchlengths"><a href="#4-1-2-branchlengths" class="headerlink" title="4.1.2 branchlengths"></a>4.1.2 branchlengths</h3><p>在优化进化树枝长用于Likelihood计算时，是否各枝长一起优化。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">branchlengths &#x3D; linked;</span><br></pre></td></tr></table></figure>

<h3 id="4-1-3-models"><a href="#4-1-3-models" class="headerlink" title="4.1.3 models"></a>4.1.3 models</h3><p>models指定模型筛选的范围</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">models &#x3D; all;</span><br></pre></td></tr></table></figure>

<h3 id="4-1-4-model-selection"><a href="#4-1-4-model-selection" class="headerlink" title="4.1.4 model_selection"></a>4.1.4 model_selection</h3><p>model_selection模型筛选的标准，一般用aicc</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">model_selection &#x3D; aicc;</span><br></pre></td></tr></table></figure>

<p>4.1.5 data_blocks</p>
<p>datablocks 设定数据分区，例如：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">[data_blocks]</span><br><span class="line">Gene1_pos1 &#x3D; 1-789\3;</span><br><span class="line">Gene1_pos2 &#x3D; 2-789\3;</span><br><span class="line">Gene1_pos3 &#x3D; 3-789\3;</span><br></pre></td></tr></table></figure>

<h3 id="4-1-6-search"><a href="#4-1-6-search" class="headerlink" title="4.1.6 search"></a>4.1.6 search</h3><p>模型筛选的算法 <code>search = greedy;</code></p>
<h2 id="4-2-部分参数详解"><a href="#4-2-部分参数详解" class="headerlink" title="4.2 部分参数详解"></a>4.2 部分参数详解</h2><h3 id="4-2-1-branchlengths"><a href="#4-2-1-branchlengths" class="headerlink" title="4.2.1 branchlengths"></a>4.2.1 branchlengths</h3><p>枝长是否相互关联，即是否对每个分隔模式独立估计进化树的枝长（目前只有BEAST、MrBayes、RAxML支持独立估算枝长）</p>
<h4 id="4-2-2-不同数据类型模型筛选的范围"><a href="#4-2-2-不同数据类型模型筛选的范围" class="headerlink" title="4.2.2 不同数据类型模型筛选的范围"></a>4.2.2 不同数据类型模型筛选的范围</h4><p>默认情况下<code>models=all</code>；如果用户需要使用更多模型，可设定<code>models=allx</code>，但应尽量避免，这是因为对少数几个模型进行优化要耗费很多时间。</p>
<ol>
<li> 对于<strong>离散性状</strong>，例如01性状（BINARY）以及无序质量性状，模型为(MULTISTATE)BINARY+G、BINARY+G+A、MULTISTATE+G、MULTISTATE+G+A，相关详情可参考RAxML的说明书。注意：性状数据模型不能计算AIC。</li>
<li> 对于<strong>DNA****序列</strong>，<code>models=all</code>将比较内置的56种模型。即由14种基本模型为基础：JC、K80、TrNef、K81、TVMef、TIMef、SYM、F81、HKY、TrN、K81uf、TVM、TIM、GTR，同时参考是否用Gamma或者I来表征不同位点进化速率的差异。注意：在分析DNA数据时，如果添加了 <code>--raxml</code>选项，则只计算GTR、GTR+G、GTR+I+G三种模型的AIC的值。</li>
<li> 对于<strong>氨基酸序列</strong>，默认情况下（使用PHYML计算），要比较的基本模型为：LG、WAG、MTREV、DAYHOFF、DCMUT、JTT、VT、BLOSUM62、CPREV、RTREV、MTMAM、MTART、HIVB、HIVW，每种又分成8种情况：LG、LG+F、LG+G、LG+G+F、LG+I、LG+I+F、LG+I+G、LG+I+G+F ，其中F为是否使用经验氨基酸频率。如果开启了RAxML选项，则会比较LG、WAG、MTREV、DAYHOFF、DCMUT、JTT、VT、BLOSUM62、CPREV、RTREV、MTMAM、MTART、HIVB、HIVW、MTZOA、PMB、JTTDCMUT、FLU、STMTREV、DUMMY、DUMMY2等21种模型，每个模型考虑6种情况。</li>
</ol>
<p>所筛选模型范围，如果改为<code>models = allx</code>，将进一步扩大：</p>
<ol>
<li> 对于DNA序列，使用PHYML时，会比较84种模型；使用RAxML时，会比较6种模型；</li>
<li> 对于氨基酸序列，使用PHYML不能获得结果，所以只能使用RAxML，比较的模型数量为195个。此时，所有的碱基频率或者氨基酸频率都是用极大似然估计获得的。</li>
</ol>
<h4 id="若从指定的模型中筛选："><a href="#若从指定的模型中筛选：" class="headerlink" title="若从指定的模型中筛选："></a>若从指定的模型中筛选：</h4><ol>
<li> 如果后续用mrbayes或者beast做分析，则需要设定models为 <code>models = mrbayes;</code>或 <code>models = beast;</code>，以便只从这两个软件中有的模型中筛选。</li>
<li> 用户也可设定PartitionFinder只考虑有Gamma或Gamma+I的相关的模型</li>
</ol>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">models &#x3D; gamma;</span><br><span class="line">models &#x3D; gammai;</span><br></pre></td></tr></table></figure>

<ol>
<li> 或直接指定DNA模型的筛选范围:</li>
</ol>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">models &#x3D; JC, JC+G, HKY, HKY+G, GTR, GTR+G;</span><br></pre></td></tr></table></figure>

<ol>
<li> 指定氨基酸模型的筛选范围</li>
</ol>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">models &#x3D; LG, LG+G, LG+G+F, WAG, WAG+G, WAG+G+F;</span><br></pre></td></tr></table></figure>

<ol>
<li> 指定模型选择的标准，作者推荐使用corrected AIC, 即AICc</li>
</ol>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"># model_selection: AIC | AICc | BIC</span><br><span class="line">model_selection &#x3D; aic;</span><br></pre></td></tr></table></figure>

<h3 id="4-2-3-设定分区"><a href="#4-2-3-设定分区" class="headerlink" title="4.2.3 设定分区"></a>4.2.3 设定分区</h3><p>分区模块需要用<code>[data_blocks]</code>标记。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">[data_blocks]</span><br><span class="line">Gene1_codon1 &#x3D; 1-1000\3; </span><br><span class="line">Gene1_codon2 &#x3D; 2-1000\3;</span><br><span class="line">Gene1_codon3 &#x3D; 3-1000\3;</span><br><span class="line">intron &#x3D; 1001-2000;</span><br></pre></td></tr></table></figure>

<p><code>[data_blocks]</code> 后紧跟基因以及位点的名称。如上面代码所示： 数据包括Gene1以及intron两部分。对于编码基因Gene1，由于密码子的1、2、3位的进化速率可能不同，因此可将1-1000位点的1、2、3碱基分开，分别作为单独的分区。</p>
<p>上述命令中：</p>
<p><code>Gene1_codon1 = 1-1000\3;</code> 表示：在第1-1000位点，从第1个碱基开始，步长为3的所有碱基</p>
<p><code>Gene1_codon2 = 2-1000\3;</code> 表示：在第2-1000位点，从第2个碱基开始，步长为3的所有碱基</p>
<p><code>Gene1_codon3 = 3-1000\3;</code> 表示：在第3-1000位点，从第3个碱基开始，步长为3的所有碱基</p>
<p><code>intron = 1001-2000;</code> 表示：从第1001-2000位点。由于是内含子，没有编码功能，所以不用将1、2、3位碱基分开。</p>
<p>如果是为MrBayes筛选分区模型，可在每行前增加charset关键词。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">charset Gene1_codon1 &#x3D; 1-1000\3;</span><br><span class="line">charset Gene1_codon2 &#x3D; 2-1000\3;</span><br><span class="line">charset Gene1_codon3 &#x3D; 3-1000\3;</span><br><span class="line">charset intron &#x3D; 1001-2000;</span><br></pre></td></tr></table></figure>

<h3 id="4-2-4-搜寻策略"><a href="#4-2-4-搜寻策略" class="headerlink" title="4.2.4 搜寻策略"></a>4.2.4 搜寻策略</h3><p>根据数据量以及要筛选的分区组合的数目，PartitionFinder的搜索策略，要通过修改<code>search</code>参数做相应调整。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">search: all | greedy | rcluster | rclusterf | hcluster | kmeans | user</span><br></pre></td></tr></table></figure>

<p>其中:</p>
<ul>
<li> <code>search = all</code> 一般用于12个分区以下；</li>
<li> <code>search = greedy</code> 10-100个分区；</li>
<li> <code>search = rcluster</code> 100个以上的分区；</li>
<li> <code>search = rclusterf</code> 只适用于RAxML选项开启时，尤其适用于有若干分析需要数据很大，同时ML需要优化很长时间时，可减少等待的时间；</li>
<li> <code>search = hcluster</code> 在研究中不推荐使用，仅为了开发而保留；</li>
<li> <code>search = user</code> 如果用户要对上述定义的分区组合，按照指定的方式筛选， 则可在<code>search = user</code>命令之后，指定搜索的策略。同一对括号中的基因表示将作为一个整体进行分析（参见英文版说明书第21页）</li>
</ul>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">together&#x3D; (Gene1_codon1, Gene1_codon2, Gene1_codon3, intron); </span><br><span class="line">intron_123  &#x3D; (Gene1_codon1, Gene1_codon2, Gene1_codon3) (intron); </span><br><span class="line">intron_12_3 &#x3D; (Gene1_codon1, Gene1_codon2) (Gene1_codon3) (intron); </span><br><span class="line">separate&#x3D; (Gene1_codon1) (Gene1_codon2) (Gene1_codon3) (intron);</span><br></pre></td></tr></table></figure>

<h3 id="4-2-5-搜寻策略调整举例"><a href="#4-2-5-搜寻策略调整举例" class="headerlink" title="4.2.5 搜寻策略调整举例"></a>4.2.5 搜寻策略调整举例</h3><h4 id="4-2-5-1-少于12个分区"><a href="#4-2-5-1-少于12个分区" class="headerlink" title="4.2.5.1 少于12个分区"></a>4.2.5.1 少于12个分区</h4><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">branchlengths &#x3D; linked;</span><br><span class="line">models &#x3D; all;</span><br><span class="line">model_selection &#x3D; aicc;</span><br><span class="line">search &#x3D; all;</span><br></pre></td></tr></table></figure>

<h4 id="4-2-5-2-对10-100个分区进行筛选"><a href="#4-2-5-2-对10-100个分区进行筛选" class="headerlink" title="4.2.5.2 对10-100个分区进行筛选"></a>4.2.5.2 对10-100个分区进行筛选</h4><p>超过100个等位基因的序列，需要用贪婪算法，基本步骤如下：</p>
<ol>
<li> 设定每个基因的起始位点以及结束位点，编码位置（1，2，3）等</li>
<li> 设定.cfg文件</li>
<li> 用raxml计算Likelihood</li>
</ol>
<p>对于DNA为：</p>
<p><code>python PartitionFinder.py InputFoldername --raxml</code></p>
<p>对于氨基酸为：</p>
<p><code>python PartitionFinderProtein.py InputFoldername --raxml</code></p>
<h4 id="4-2-5-3-超大数据（1000个）的分区筛选"><a href="#4-2-5-3-超大数据（1000个）的分区筛选" class="headerlink" title="4.2.5.3 超大数据（1000个）的分区筛选"></a>4.2.5.3 超大数据（1000个）的分区筛选</h4><p>超大型数据的位点（1000个以上），贪婪算法仍然太慢，此时推荐使用其他算法，以筛选最优分隔模型设置方案（Lanfear et al 2014），基本步骤如下:</p>
<ol>
<li> 设定每个基因的起始位点以及结束位点，编码位置（1、2、3）等</li>
<li> 更改.cfg文件，设定 <code>search = rcluster</code></li>
<li> 在cmd中运行PartitionFinder:</li>
</ol>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">python PartitionFinder.py InputFoldername --raxml</span><br></pre></td></tr></table></figure>

<p>如仍然过慢，则更改recluster-max 参数:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">python PartitionFinder.py InputFoldername --raxml --rcluster-max 100</span><br></pre></td></tr></table></figure>

<h3 id="4-2-6-用户自定义进化树"><a href="#4-2-6-用户自定义进化树" class="headerlink" title="4.2.6 用户自定义进化树"></a>4.2.6 用户自定义进化树</h3><p>在模拟研究中，可能会假设已知进化树的结构，探讨能否筛选到最优分区模型。此时可指定一棵进化树，在后续的分析中，该进化树的拓扑结构将被保留，但是枝长会根据GTR+I+GAMMA模型重新估计。指定进化树的命令如下：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">ALIGNMENT FILE # </span><br><span class="line">alignment &#x3D; test.phy; </span><br><span class="line">user_tree_topology &#x3D; tree.phy</span><br></pre></td></tr></table></figure>

<h1 id="5-输出结果"><a href="#5-输出结果" class="headerlink" title="5.输出结果"></a>5.输出结果</h1><p>生成的结果位于同一文件夹下的analysis文件夹中，该文件夹包含以下文件：</p>
<ol>
<li> <code>best_schemes.txt</code>纯文本文件，包括所筛选到的最优化分隔模型划分方案以及每个区块对应的最优模型以及RAxML和Nexus的分隔模型。</li>
<li> <code>subsets</code>文件夹:每个分区所对应的AICc值以及相应的筛选过程</li>
<li> <code>schemes</code>文件夹:所分析的分隔模型划分方案，详细的分析过程</li>
</ol>
<h1 id="6-命令行选项"><a href="#6-命令行选项" class="headerlink" title="6.命令行选项"></a>6.命令行选项</h1><p>在运行PartitionFinder时，可在命令行设定一些参数，具体如下：</p>
<ul>
<li> <code>--all-states</code> 仅用于k-means选项开启的情况，用于限定状态的数量</li>
<li> <code>--force-restart</code> 删除之前的分析，完全重新开始</li>
<li> <code>--min-subset-size</code> 设定子集的大小。只用于k-means算法（这个选项在说明书中交代得不是很清楚）</li>
<li> <code>--no-ml-tree</code> 设定此选项后，PartitionFinder将从Neighbour Joining（PHYML）或者Parsimony树（RAxML）开始。</li>
<li> <code>--processors N, -p N</code> 多线程计算，使用多个CPU内核。默认情况下，PartitionFinder会使用多个核，如果要控制所用到的核数，则应做相应设定。</li>
<li> <code>--quick, -q</code> 让PartitionFinder停止写一些记录文件，如果处理的数据特别大，有可能会提升速度，常用于使用贪婪算法的情况下。</li>
<li> <code>--raxml</code> 使用RAxML帮忙寻找最优分隔模型。</li>
<li> <code>--rcluster-max N</code> 设定cluster的数量，默认值N为1000或者分区数乘以10中的较大者，一般已经可以保证搜寻到正确的分隔模型。</li>
<li> <code>--rcluster-percent N</code> 设定要搜寻的所有可能的分区组成的百分数，默认为10%。该参数配合<code>--rcluster-max N</code>，可减少重复计算。</li>
<li> <code>--save-phylofiles</code> 保存数据分析过程中生成的进化树，主要用来纠错。</li>
<li> <code>--weights &quot;W_rate , W_base , W_model , W_alpha&quot;</code> 只用于<code>--raxml</code>或<code>hcluster or rcluster</code>选项开启时，分别表示几个参数的权重（参考说明书第26页）:</li>
</ul>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">1. the overall rate for a subset</span><br><span class="line">2. the base&#x2F;amino acid frequencies</span><br><span class="line">3. the model parameters</span><br><span class="line">4. the alpha parameter (which describes  gamma distributed rates across sites)</span><br></pre></td></tr></table></figure>

<h1 id="进一步阅读"><a href="#进一步阅读" class="headerlink" title="进一步阅读"></a>进一步阅读</h1><ul>
<li> Lanfear, R., Calcott, B., Ho, S. Y., &amp; Guindon, S. (2012). PartitionFinder: combined selection of partitioning schemes and substitution models for phylogenetic analyses. Molecular biology and evolution, 29(6), 1695-1701.</li>
</ul>
<h1 id="附录"><a href="#附录" class="headerlink" title="附录"></a>附录</h1><h2 id="DNA序列碱基替换模型的三大要素"><a href="#DNA序列碱基替换模型的三大要素" class="headerlink" title="DNA序列碱基替换模型的三大要素"></a>DNA序列碱基替换模型的三大要素</h2><h3 id="1-碱基频率的估算"><a href="#1-碱基频率的估算" class="headerlink" title="1. 碱基频率的估算"></a>1. 碱基频率的估算</h3><p>模型需要提供ATCG频率的估计方法，目前有四种方法:</p>
<ol>
<li> equal 直接假定ATCG的频率相同</li>
<li> model 是从其他数据读取不同氨基酸的频率（仅用于氨基酸序列）</li>
<li> empirical 是直接从数据计算ATCG的频率</li>
<li> ML是用极大似然法估计ATCG的频率</li>
</ol>
<p>注意：在模型后添加<code>+X</code>，即可开启ML方法估计ATCG频率，例如<code>F81+X</code>，不过一般用户都是用empirical的方法估计频率，而且结果与ML的结果相差无几。</p>
<h3 id="2-碱基替换速率矩阵"><a href="#2-碱基替换速率矩阵" class="headerlink" title="2. 碱基替换速率矩阵"></a>2. 碱基替换速率矩阵</h3><p>用来表征碱基之间的转换速率（Relative rates of substitution），其中GTR模型最为复杂，需要估计所有6个参数。</p>
<h3 id="3-不同位点进化速率的差异"><a href="#3-不同位点进化速率的差异" class="headerlink" title="3. 不同位点进化速率的差异"></a>3. 不同位点进化速率的差异</h3><p>分成以下几种情形：</p>
<ol>
<li> 各位点速率恒定；</li>
<li> 部分位点进化树进化树速率恒定不变 （I用来表示进化速率恒定位点的比例）：GTR + I；</li>
<li> 假设各位点的进化速率符合Gamma分布（因为Gamma分布要估计的参数少，同时曲线有足够的变化，能够较好地拟合多数情况，所以用Gamma分布）：GTR + GAMMA；</li>
<li> 同时考虑各位点的进化树速率的差异以及恒定位点的比例：GTR + GAMMA + I。</li>
</ol>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2017/08/05/2017-8-5-partitionfinder/" data-id="ckrsvkzb7008x90w099dibojl" data-title="用PartitionFinder2 筛选系统发育分析分隔模型" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6/" rel="tag">软件</a></li></ul>

    </footer>
  </div>
  
</article>



  


  <nav id="page-nav">
    
    <a class="extend prev" rel="prev" href="/page/9/">&laquo; Prev</a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/8/">8</a><a class="page-number" href="/page/9/">9</a><span class="page-number current">10</span><a class="page-number" href="/page/11/">11</a><a class="page-number" href="/page/12/">12</a><span class="space">&hellip;</span><a class="page-number" href="/page/26/">26</a><a class="extend next" rel="next" href="/page/11/">Next &raquo;</a>
  </nav>

</section>
        
          <aside id="sidebar">
  
    

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">Tags</h3>
    <div class="widget">
      <ul class="tag-list" itemprop="keywords"><li class="tag-list-item"><a class="tag-list-link" href="/tags/Python/" rel="tag">Python</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/R%E7%BC%96%E7%A8%8B/" rel="tag">R编程</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E4%BA%BA%E7%89%A9/" rel="tag">人物</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%86%99%E4%BD%9C/" rel="tag">写作</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" rel="tag">可视化</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9B%BE%E4%B9%A6/" rel="tag">图书</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9B%BE%E5%83%8F%E5%A4%84%E7%90%86/" rel="tag">图像处理</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9C%B0%E7%90%86/" rel="tag">地理</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%A3%B0%E6%98%8E/" rel="tag">声明</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%A4%A9%E6%96%87/" rel="tag">天文</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%AD%A6%E4%B9%A0%E8%B5%84%E6%BA%90/" rel="tag">学习资源</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%AF%B9%E8%AF%9D%E5%BD%95/" rel="tag">对话录</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%BF%83%E5%BE%97/" rel="tag">心得</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%8A%A5%E5%91%8A/" rel="tag">报告</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%8E%92%E7%89%88/" rel="tag">排版</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%85%E4%B9%A1/" rel="tag">故乡</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E5%BA%93/" rel="tag">数据库</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%A0%87%E6%9C%AC%E9%A6%86/" rel="tag">标本馆</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%A4%8D%E7%89%A9%E5%88%86%E7%B1%BB/" rel="tag">植物分类</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%A4%8D%E7%89%A9%E5%AD%A6%E5%90%8D/" rel="tag">植物学名</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B1%82%E8%81%8C/" rel="tag">求职</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%89%88%E6%9C%AC%E6%8E%A7%E5%88%B6/" rel="tag">版本控制</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%94%9F%E6%80%81/" rel="tag">生态</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%94%9F%E6%80%81%E8%B5%84%E8%AE%AF/" rel="tag">生态资讯</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%94%9F%E6%B4%BB%E7%82%B9%E6%BB%B4/" rel="tag">生活点滴</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%A0%94%E7%A9%B6%E8%B5%84%E8%AE%AF/" rel="tag">研究资讯</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" rel="tag">系统发育</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%9F%E8%AE%A1/" rel="tag">统计</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BC%96%E7%A8%8B/" rel="tag">编程</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%A1%8C%E6%91%84%E5%8C%86%E5%8C%86/" rel="tag">行摄匆匆</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%A7%81%E9%97%BB%E5%BD%95/" rel="tag">见闻录</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%AF%97%E6%AD%8C/" rel="tag">诗歌</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6/" rel="tag">软件</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BF%9B%E5%8C%96/" rel="tag">进化</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%9A%8F%E6%83%B3/" rel="tag">随想</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%A9%B3%E8%AE%BA/" rel="tag">驳论</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">Tag Cloud</h3>
    <div class="widget tagcloud">
      <a href="/tags/Python/" style="font-size: 10px;">Python</a> <a href="/tags/R%E7%BC%96%E7%A8%8B/" style="font-size: 18.75px;">R编程</a> <a href="/tags/%E4%BA%BA%E7%89%A9/" style="font-size: 11.88px;">人物</a> <a href="/tags/%E5%86%99%E4%BD%9C/" style="font-size: 13.13px;">写作</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" style="font-size: 12.5px;">可视化</a> <a href="/tags/%E5%9B%BE%E4%B9%A6/" style="font-size: 10px;">图书</a> <a href="/tags/%E5%9B%BE%E5%83%8F%E5%A4%84%E7%90%86/" style="font-size: 11.88px;">图像处理</a> <a href="/tags/%E5%9C%B0%E7%90%86/" style="font-size: 17.5px;">地理</a> <a href="/tags/%E5%A3%B0%E6%98%8E/" style="font-size: 10.63px;">声明</a> <a href="/tags/%E5%A4%A9%E6%96%87/" style="font-size: 16.25px;">天文</a> <a href="/tags/%E5%AD%A6%E4%B9%A0%E8%B5%84%E6%BA%90/" style="font-size: 10px;">学习资源</a> <a href="/tags/%E5%AF%B9%E8%AF%9D%E5%BD%95/" style="font-size: 10px;">对话录</a> <a href="/tags/%E5%BF%83%E5%BE%97/" style="font-size: 10px;">心得</a> <a href="/tags/%E6%8A%A5%E5%91%8A/" style="font-size: 12.5px;">报告</a> <a href="/tags/%E6%8E%92%E7%89%88/" style="font-size: 13.75px;">排版</a> <a href="/tags/%E6%95%85%E4%B9%A1/" style="font-size: 10.63px;">故乡</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E5%BA%93/" style="font-size: 10.63px;">数据库</a> <a href="/tags/%E6%A0%87%E6%9C%AC%E9%A6%86/" style="font-size: 14.38px;">标本馆</a> <a href="/tags/%E6%A4%8D%E7%89%A9%E5%88%86%E7%B1%BB/" style="font-size: 10px;">植物分类</a> <a href="/tags/%E6%A4%8D%E7%89%A9%E5%AD%A6%E5%90%8D/" style="font-size: 10px;">植物学名</a> <a href="/tags/%E6%B1%82%E8%81%8C/" style="font-size: 10px;">求职</a> <a href="/tags/%E7%89%88%E6%9C%AC%E6%8E%A7%E5%88%B6/" style="font-size: 11.25px;">版本控制</a> <a href="/tags/%E7%94%9F%E6%80%81/" style="font-size: 16.88px;">生态</a> <a href="/tags/%E7%94%9F%E6%80%81%E8%B5%84%E8%AE%AF/" style="font-size: 10px;">生态资讯</a> <a href="/tags/%E7%94%9F%E6%B4%BB%E7%82%B9%E6%BB%B4/" style="font-size: 10px;">生活点滴</a> <a href="/tags/%E7%A0%94%E7%A9%B6%E8%B5%84%E8%AE%AF/" style="font-size: 10px;">研究资讯</a> <a href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" style="font-size: 10px;">系统发育</a> <a href="/tags/%E7%BB%9F%E8%AE%A1/" style="font-size: 18.13px;">统计</a> <a href="/tags/%E7%BC%96%E7%A8%8B/" style="font-size: 10.63px;">编程</a> <a href="/tags/%E8%A1%8C%E6%91%84%E5%8C%86%E5%8C%86/" style="font-size: 15px;">行摄匆匆</a> <a href="/tags/%E8%A7%81%E9%97%BB%E5%BD%95/" style="font-size: 10px;">见闻录</a> <a href="/tags/%E8%AF%97%E6%AD%8C/" style="font-size: 15.63px;">诗歌</a> <a href="/tags/%E8%BD%AF%E4%BB%B6/" style="font-size: 19.38px;">软件</a> <a href="/tags/%E8%BF%9B%E5%8C%96/" style="font-size: 10.63px;">进化</a> <a href="/tags/%E9%9A%8F%E6%83%B3/" style="font-size: 11.25px;">随想</a> <a href="/tags/%E9%9A%8F%E7%AC%94/" style="font-size: 20px;">随笔</a> <a href="/tags/%E9%A9%B3%E8%AE%BA/" style="font-size: 10px;">驳论</a>
    </div>
  </div>

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">Archives</h3>
    <div class="widget">
      <ul class="archive-list"><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/09/">September 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/07/">July 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/05/">May 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/03/">March 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/02/">February 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/12/">December 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/11/">November 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/06/">June 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/05/">May 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/02/">February 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/01/">January 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/10/">October 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/09/">September 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/07/">July 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/06/">June 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/05/">May 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/04/">April 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/03/">March 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/02/">February 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/01/">January 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/12/">December 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/11/">November 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/10/">October 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/09/">September 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/08/">August 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/07/">July 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/06/">June 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/05/">May 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/04/">April 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/03/">March 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/02/">February 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/01/">January 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/11/">November 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/10/">October 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/09/">September 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/08/">August 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/06/">June 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/05/">May 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/04/">April 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/02/">February 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/12/">December 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/11/">November 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/09/">September 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/06/">June 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/05/">May 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/04/">April 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/03/">March 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/01/">January 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/11/">November 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/10/">October 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/09/">September 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/07/">July 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/05/">May 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/04/">April 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/03/">March 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/02/">February 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/01/">January 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/12/">December 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/11/">November 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/10/">October 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/09/">September 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/07/">July 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/05/">May 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/03/">March 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/02/">February 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/01/">January 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/12/">December 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/10/">October 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/05/">May 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/04/">April 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/03/">March 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/02/">February 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/01/">January 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/12/">December 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/10/">October 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/08/">August 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/07/">July 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/06/">June 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/05/">May 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/04/">April 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/02/">February 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/01/">January 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/12/">December 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/11/">November 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/10/">October 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/08/">August 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/07/">July 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/05/">May 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/04/">April 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/03/">March 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/02/">February 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/12/">December 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/11/">November 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/09/">September 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/08/">August 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/07/">July 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/06/">June 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/05/">May 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/04/">April 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/03/">March 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/01/">January 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2009/08/">August 2009</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2009/07/">July 2009</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2009/06/">June 2009</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2009/05/">May 2009</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">Recent Posts</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/2021/09/13/2021-10-19school/">最初的小学时光</a>
          </li>
        
          <li>
            <a href="/2021/07/25/2021-07-25beans/">种豆记</a>
          </li>
        
          <li>
            <a href="/2021/07/19/2021-07-19-rresources/">植物学和生态学的一些R语言学习资源</a>
          </li>
        
          <li>
            <a href="/2021/05/15/2021-05-15-ecopapers100/">生态学研究生要清楚的三件事：从《生态学家必读的100篇论文》说起</a>
          </li>
        
          <li>
            <a href="/2021/05/13/2021-05-13-life/">生活的答案</a>
          </li>
        
      </ul>
    </div>
  </div>

  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      
      &copy; 2021 Jinlong Zhang<br>
      Powered by <a href="https://hexo.io/" target="_blank">Hexo</a>
    </div>
  </div>
</footer>

    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">Home</a>
  
    <a href="/archives" class="mobile-nav-link">Archives</a>
  
    <a href="/about" class="mobile-nav-link">About</a>
  
</nav>
    


<script src="/js/jquery-3.4.1.min.js"></script>



  
<script src="/fancybox/jquery.fancybox.min.js"></script>




<script src="/js/script.js"></script>





  </div>
</body>
</html>