<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  
  <title>Jinlong Zhang&#39;s blog</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta property="og:type" content="website">
<meta property="og:title" content="Jinlong Zhang&#39;s blog">
<meta property="og:url" content="https://helixcn.github.io/page/24/index.html">
<meta property="og:site_name" content="Jinlong Zhang&#39;s blog">
<meta property="og:locale" content="en_US">
<meta property="article:author" content="Jinlong Zhang">
<meta name="twitter:card" content="summary">
  
    <link rel="alternate" href="/atom.xml" title="Jinlong Zhang's blog" type="application/atom+xml">
  
  
    <link rel="shortcut icon" href="/favicon.png">
  
  
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-source-code-pro@0.0.71/index.min.css">

  
  
<link rel="stylesheet" href="/css/style.css">

  
    
<link rel="stylesheet" href="/fancybox/jquery.fancybox.min.css">

  
</head>

<body>
  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">Jinlong Zhang&#39;s blog</a>
      </h1>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"></a>
        
          <a class="main-nav-link" href="/">Home</a>
        
          <a class="main-nav-link" href="/archives">Archives</a>
        
          <a class="main-nav-link" href="/about">About</a>
        
      </nav>
      <nav id="sub-nav">
        
          <a id="nav-rss-link" class="nav-icon" href="/atom.xml" title="RSS Feed"></a>
        
        <a id="nav-search-btn" class="nav-icon" title="Search"></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="Search"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="https://helixcn.github.io"></form>
      </div>
    </div>
  </div>
</header>

      <div class="outer">
        <section id="main">
  
    <article id="post-2010-9-6-skyatlas" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2010/09/05/2010-9-6-skyatlas/" class="article-date">
  <time class="dt-published" datetime="2010-09-05T08:05:54.000Z" itemprop="datePublished">2010-09-05</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2010/09/05/2010-9-6-skyatlas/">skyatlas:用R绘制星图和行星的周年视运动</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>在一些天文期刊上经常看到赤道坐标星图，不但给出了某一个时刻太阳、月亮的位置，还给出了大行星在一年中相对于星空背景的运行轨迹。今天我们就用R软件绘制这样的星图。</p>
<p>当然，为了绘制星图，我们需要知道恒星的位置和亮度。这就需要引入天文坐标系。天文学上天体的位置有多种坐标系统，包括赤道坐标系，黄道坐标系和地平坐标系等。</p>
<p>首先介绍一下今天用到的赤道坐标系。</p>
<p>赤道坐标系是以春分点作为起点，春分点和秋分点是赤道和黄道两个大圈相交的点。由于地球的自转，我们看到太阳相对于恒星背景从西向东运行，太阳所经过的路线就称为黄道。不难想象，春分点过后，太阳相对于星空背景一天天逐渐向北移动（当然也同时向东移动，差不多每天移动1度多一点儿）。我们就说春分点为升交点。太阳到达夏至点的时候，是最北的，之后逐渐向南，到秋分点又与赤道相交。因此秋分点又称为降交点。太阳继续向南，此时太阳位于天赤道以南。到达冬至点之后又逐渐向北。如此周而复始。</p>
<p>我们定义赤道坐标系的原点就是升交点，向左度量，一圈用24h表示。其经度就称为“赤经”，其纬度就称为“赤纬”，用度表示， 北半球为正，南半球为负。这就是我们将要看到的图形中的星图。天文学上的星图左东右西，与地图正好相反。</p>
<p>其次，我们需要了解一下恒星的位置和亮度。<br>这里我们选取了耶鲁亮星星表。这个星表包含了肉眼可见的9110颗较亮恒星的位置亮度等信息。按照肉眼可见的亮度分成不同的星等，星的亮度是用照相测量法测量的。</p>
<p>第三，我们需要知道行星在任何一个时刻的位置。即，我们输入时刻，就有程序可以计算出该天体的位置等信息。这是球面天文学和天体力学的内容了。我们这里使用R的moonsun程序包计算某一时刻天体的位置和亮度。 我们需要输入的时间为儒略日（Julian day）。儒略日是什么？儒略日是指从公元前4713年1月1日中午12时开始，开始距离现在的天数。我们输入的日期都必须先转换成儒略日，再用儒略日作为时间点计算。儒略日计算两个时间点之间相隔的天数非常方便，因此在天文学上得到了广泛的应用。</p>
<p>第四，我们需要了解一下行星运行的基本知识。<br>水星，金星位于地球内侧，更靠近太阳，称为内行星。火星，木星，土星，天王星，海王星位于地球外侧，称为外行星。<br>八大行星的轨道与黄道相差不大，而且均自西向东运行。靠近太阳最近的水星运行速度最快。从地球上看，行星相对于星空背景有时候向东，有时候向西。向东称为顺行，各行星大部分时间在顺行中，向西称为逆行，在顺行和逆行的转换过程中，有较短的时间，行星相对星空背景处于相对静止的状态，称为留。</p>
<p>这里给出了skyatlasR程序能够绘制行星周年视运动的轨迹，当然，由于没有考虑到星体之间的摄动等更精确的天体力学因素，行星的位置计算不太高，精度在几角分之内，但是仍然能够满足大部分天文科普的需求。更精确的历表就得参考NASA,也就是美国国家航空航天局喷气推进实验室的HORIZON历书系统了。用户可以在线查询一些信息。</p>
<p>下面简要介绍一下 skyatlas 程序的用法</p>
<p>使用方法</p>
<p>当然，首先要下载我们编好的程序了。 下载 <a target="_blank" rel="noopener" href="http://blog.sciencenet.cn/upload/blog/file/2010/9/20109603912834771.rar.rar">skyatlas.rar</a></p>
<p>1 下载R软件 <a target="_blank" rel="noopener" href="http://ftp.ctex.org/mirrors/CRAN/bin/windows/base/R-2.11.1-win32.exe">http://ftp.ctex.org/mirrors/CRAN/bin/windows/base/R-2.11.1-win32.exe</a> 下载<br>2 将 “plot.r”, “resstar.csv”, “moonsun_0.1.1.zip” 拷贝到 C:下,”moonsun_0.1.1.zip” 为R程序包，安装时不需要解压缩。<br>3 安装moonsun.zip程序包，双击R的图标，路径为 开始&gt;所有程序&gt; R, 2.11.1, 然后再菜单中选择，菜单“程序包”，“从本地zip安装程序包”，选择 moonsun_0.1.1.zip 即可。<br>4 从R的菜单 “文件” “source R code”. 打开 “C:/plot.r”,或者输入命令source(“C:/plot.r”)即可。</p>
<p>运行实例<br>R图形的组件都是一步一步添加上去的，故绘制一个图形步骤较多，这里举出了绘制水星1980年和2011年的例子，可供有兴趣的读者参考。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">######################################</span></span><br><span class="line"><span class="comment">######################################</span></span><br><span class="line"></span><br><span class="line"><span class="comment">### 例一 绘制1980年水星的周年视运动  </span></span><br><span class="line"><span class="comment">##  该结果与《中国大百科全书 天文学卷》参见 485页 进行比较，水星的运行曲线结果是一致的。</span></span><br><span class="line">x11(<span class="number">14</span>, <span class="number">5</span>)</span><br><span class="line"><span class="comment">##绘制底图</span></span><br><span class="line">background(type = <span class="string">&quot;n&quot;</span>, xlim =<span class="built_in">c</span>(<span class="number">0</span>, <span class="number">24</span>), ylim = <span class="built_in">c</span>(-<span class="number">40</span>, <span class="number">40</span>), xlab = <span class="string">&quot;赤经&quot;</span>, ylab = <span class="string">&quot;赤纬&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">##添加黄道</span></span><br><span class="line">add.ecliptic()</span><br><span class="line"></span><br><span class="line"><span class="comment">##添加4等以上的恒星</span></span><br><span class="line">add.stars(resstar, mag = <span class="number">4</span>, pch = <span class="number">19</span>, time = <span class="number">1</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">## 添加水星的运行轨迹</span></span><br><span class="line">tracks(mercury(jd(year = <span class="number">1980</span>,month = <span class="number">1</span>,day = <span class="number">1</span>,<span class="built_in">length</span> = <span class="number">365</span>)))</span><br><span class="line"><span class="comment">## 添加亮星的名称</span></span><br><span class="line">add.bright.lab ()</span><br><span class="line"></span><br><span class="line"><span class="comment">###生成每隔30天的点数据，</span></span><br><span class="line">position &lt;- mercury(jd(year = <span class="number">1980</span>,month = <span class="number">1</span>, day = <span class="number">1</span>,<span class="built_in">length</span> = <span class="number">12</span>, by = <span class="number">30</span>))</span><br><span class="line">addlab(position)</span><br><span class="line">title(<span class="string">&quot;1980年水星的周年视运动&quot;</span>)</span><br></pre></td></tr></table></figure>
<p>结果如下</p>
<p><img src="http://blog.sciencenet.cn/upload/blog/images/2010/9/20109601735522.JPG" alt="img"></p>
<h1 id="例二"><a href="#例二" class="headerlink" title="例二"></a>例二</h1><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">### 2011水星的周年视运动</span></span><br><span class="line"></span><br><span class="line"><span class="comment">### 第一步 打开一个绘图窗口</span></span><br><span class="line">x11(<span class="number">14</span>, <span class="number">5</span>)</span><br><span class="line"><span class="comment">##  第二步 绘制底图</span></span><br><span class="line">background(type = <span class="string">&quot;n&quot;</span>, xlim =<span class="built_in">c</span>(<span class="number">0</span>, <span class="number">24</span>), ylim = <span class="built_in">c</span>(-<span class="number">40</span>, <span class="number">40</span>), xlab = <span class="string">&quot;赤经&quot;</span>, ylab = <span class="string">&quot;赤纬&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">##添加黄道</span></span><br><span class="line">add.ecliptic()</span><br><span class="line"></span><br><span class="line"><span class="comment">##第四步 添加4等以上的恒星</span></span><br><span class="line">add.stars(resstar, mag = <span class="number">4</span>, pch = <span class="number">19</span>, time = <span class="number">1</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">## 第五步 添加水星的运行轨迹</span></span><br><span class="line">tracks(mercury(jd(year = <span class="number">2010</span>,month = <span class="number">1</span>,day = <span class="number">1</span>,<span class="built_in">length</span> = <span class="number">365</span>)))</span><br><span class="line"></span><br><span class="line"><span class="comment">## 第六步 添加亮星的名称</span></span><br><span class="line">add.bright.lab()</span><br><span class="line"></span><br><span class="line"><span class="comment">###第七步 添加生成每隔30天的点数据，并标在图上</span></span><br><span class="line">position &lt;- mercury(jd(year = <span class="number">2010</span>,month = <span class="number">1</span>, day = <span class="number">1</span>,<span class="built_in">length</span> = <span class="number">12</span>, by = <span class="number">30</span>))</span><br><span class="line">addlab(position)</span><br><span class="line"></span><br><span class="line"><span class="comment">### 第八步 添加标题</span></span><br><span class="line">title(<span class="string">&quot;2011年水星的周年视运动&quot;</span>)</span><br></pre></td></tr></table></figure>
<p> <img src="http://blog.sciencenet.cn/upload/blog/images/2010/9/20109601748756.JPG" alt="img"></p>
<h1 id="例三-当前时刻日月及大行星的位置图"><a href="#例三-当前时刻日月及大行星的位置图" class="headerlink" title="例三 当前时刻日月及大行星的位置图"></a>例三 当前时刻日月及大行星的位置图</h1><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">### 打开一个绘图窗口</span></span><br><span class="line">x11(<span class="number">14</span>, <span class="number">5</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">## 绘制赤道背景坐标</span></span><br><span class="line">background(type = <span class="string">&quot;n&quot;</span>, xlim =<span class="built_in">c</span>(<span class="number">0</span>, <span class="number">24</span>), ylim = <span class="built_in">c</span>(-<span class="number">40</span>, <span class="number">40</span>), xlab = <span class="string">&quot;赤经&quot;</span>, ylab = <span class="string">&quot;赤纬&quot;</span>, bg = <span class="string">&quot;darkblue&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">##添加黄道</span></span><br><span class="line">add.ecliptic()</span><br><span class="line"></span><br><span class="line"><span class="comment">## 添加4等以上的恒星</span></span><br><span class="line">add.stars(resstar, mag = <span class="number">4</span>, pch = <span class="number">19</span>, time = <span class="number">1</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">## 添加亮星的名称</span></span><br><span class="line">add.bright.lab()</span><br><span class="line"></span><br><span class="line"><span class="comment">#### 添加大行星</span></span><br><span class="line">add.planet(col = <span class="string">&quot;black&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">## 添加标题</span></span><br><span class="line">title(paste(<span class="string">&quot;&quot;</span>, <span class="string">&quot;日月及大行星的位置(&quot;</span>,<span class="built_in">as.character</span>(Sys.Date()),<span class="string">&quot;)&quot;</span> ))</span><br></pre></td></tr></table></figure>

<p> <img src="http://blog.sciencenet.cn/upload/blog/images/2010/9/20109601719225.JPG" alt="img"></p>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2010/09/05/2010-9-6-skyatlas/" data-id="cl1n7qj26001zeo8473r80yxe" data-title="skyatlas:用R绘制星图和行星的周年视运动" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E5%A4%A9%E6%96%87/" rel="tag">天文</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2010-8-9-niche" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2010/08/08/2010-8-9-niche/" class="article-date">
  <time class="dt-published" datetime="2010-08-08T09:18:54.000Z" itemprop="datePublished">2010-08-08</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2010/08/08/2010-8-9-niche/">计算生态位宽度与生态位重叠的R程序</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>生态位理论是解释生物多样性格局和物种共存的重要理论之一。每个物种对生境的偏好，就确定了其生态位。从1910年至1980年以前，国外生态学家对物种生态位的计算等生态位的宽度和计算进行了大量研究，主要的计算方法都是那时候提出的。从上世纪80年代到现在，我国的生态学界仍然能见到一些计算物种生态位宽度和生态位重叠的论文。</p>
<p>张金屯老师的《数量生态学》更是对生态位和生态位重叠的计量方法进行了详细介绍。</p>
<p>为了方便的完成这些指数的计算，本人粗略编写了一些R程序，完成相应指数的计算.还增加了Bootstrap检验，从而获得生态位重叠的置信区间估计。</p>
<p>输入数据的格式是标准的群落物种矩阵格式。即每一行代表每一个样方，每一列代表每一个物种。</p>
<p>下面就是相应的R程序，可供下载。</p>
<p>由于当前代买还存在很多冗余，欢迎各位同行提出宝贵意见。以便修订之后整合到进行种间关联度分析的spaa 程序包中。</p>
<p>有任何意见或建议，均可发邮件至 <a href="mailto:jinlongzhang01@gmail.com">jinlongzhang01@gmail.com</a> 来信必复。</p>
<p>种间关联度及生态位重叠的R程序包</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2010/08/08/2010-8-9-niche/" data-id="cl1n7qj21001seo8437i6htyo" data-title="计算生态位宽度与生态位重叠的R程序" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6/" rel="tag">软件</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2010-8-3-lagrange" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2010/08/03/2010-8-3-lagrange/" class="article-date">
  <time class="dt-published" datetime="2010-08-02T22:32:54.000Z" itemprop="datePublished">2010-08-03</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2010/08/03/2010-8-3-lagrange/">怎样运行Lagrange软件</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>注意！ 运行Lagrange软件的指南随后贴出，现在的版本仅供参考。</p>
<p><strong>“Lagrange</strong> is a <a target="_blank" rel="noopener" href="http://www.python.org/">Python</a> package implementing likelihood models for geographic range evolution on phylogenetic trees, with methods for inferring rates of dispersal and local extinction and ancestral ranges. “     <a target="_blank" rel="noopener" href="http://code.google.com/p/lagrange/">http://code.google.com/p/lagrange/</a> </p>
<h1 id="1-Download-and-install-Python-NumPy-and-SciPy"><a href="#1-Download-and-install-Python-NumPy-and-SciPy" class="headerlink" title="1 Download and install Python NumPy and SciPy"></a>1 Download and install Python NumPy and SciPy</h1><p>Lagrange requires Python, NumPy and SciPy, the first step is to download and install these software.</p>
<p>Download and install Python2.5, because SciPy requires this version, it can be downloaded from<a target="_blank" rel="noopener" href="http://www.python.org/download/">http://www.python.org/download/</a></p>
<p>Download and install NumPy and SciPy ,you can down load the exe files form</p>
<p><a target="_blank" rel="noopener" href="http://sourceforge.net/project/showfiles.php?group_id=27747">http://sourceforge.net/project/showfiles.php?group_id=27747</a>  for SciPy</p>
<p><a target="_blank" rel="noopener" href="http://sourceforge.net/project/showfiles.php?group_id=1369&package_id=175103">http://sourceforge.net/project/showfiles.php?group_id=1369&amp;package_id=175103</a> for NumPi</p>
<p>It is strongly recommended that you download the  numpy-1.2.1-win32-superpack-python2.5.exe, since the  imcompatible problems may exist if you installed a incorrect version.</p>
<p>You can visit scipy’s website for more detail.</p>
<p><a target="_blank" rel="noopener" href="http://www.scipy.org/">http://www.scipy.org/</a></p>
<h1 id="2-Copy-the-Lagrange-modules-to-working-directory"><a href="#2-Copy-the-Lagrange-modules-to-working-directory" class="headerlink" title="2 Copy the Lagrange modules to working directory"></a>2 Copy the Lagrange modules to working directory</h1><p>Assume you have had all the software that Lagrange required installed on your PC. You may seed a directory under C:Python25</p>
<p>now unzip your lagrange2[1].0.1.tar.gz</p>
<p>copy all the file in the directory which just unzipped to C:python25</p>
<h1 id="3-Data-preparation"><a href="#3-Data-preparation" class="headerlink" title="3 Data preparation"></a>3 Data preparation</h1><p>The python scripts for exercise can be downloaded at</p>
<p><a target="_blank" rel="noopener" href="http://bodegaphylo.wikispot.org/Biogeography_(Donoghue_%26_Smith)">http://bodegaphylo.wikispot.org/Biogeography_%28Donoghue_%26_Smith%29</a></p>
<p>copy the test python script named psychotria.py and psychotria.matrix to python directory, C:\python25  .</p>
<h1 id="4-Run-python-scripts-in-cmd"><a href="#4-Run-python-scripts-in-cmd" class="headerlink" title="4 Run python scripts in cmd"></a>4 Run python scripts in cmd</h1><p>The python scripts can be implement in commander just which just like the DOS , to run the script, begin here:</p>
<p>Begin&gt;run&gt; type <code>cmd</code></p>
<p>a commander window will appeared</p>
<p>just type cd C:python25  to change your working directory to test</p>
<p>and type  <code>psychotria.py&gt;results.txt</code>  “&gt;results.txt” allowed you to save the results of the scripts in “results.txt”</p>
<p>When all are got ready , the python script will be run in your PC</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2010/08/03/2010-8-3-lagrange/" data-id="cl1n7qj1z001peo843k2r59v0" data-title="怎样运行Lagrange软件" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6/" rel="tag">软件</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2010-7-3-seqRFLP" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2010/07/03/2010-7-3-seqRFLP/" class="article-date">
  <time class="dt-published" datetime="2010-07-02T23:24:54.000Z" itemprop="datePublished">2010-07-03</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2010/07/03/2010-7-3-seqRFLP/">从DNA序列预测RFLP—seqRFLP软件包简介</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <h1 id="1-seqRFLP简介"><a href="#1-seqRFLP简介" class="headerlink" title="1. seqRFLP简介"></a>1. seqRFLP简介</h1><p>RFLP和TRFLP等技术在生物多样性研究中十分广泛。随着测序技术的日趋普及，在进行RFLP或TRFLP之前，用户很可能已经获得了DNA序列。为了对RFLP和TRFLP的结果进行更好的分析，特别是选择合适的内切酶，对模糊的电泳条带进行确认，基于RFLP/TRFLP及序列分别对物种进行鉴定等显得日趋重要，但是目前还缺乏这样一种已知序列就能获得RFLP/TRFLP的分析平台。为此我们编写了seqRFLP软件。</p>
<p>seqRFLP是用来对DNA序列进行模拟酶切、模拟PCR引物筛选的软件包，目前版本是1.0.0。用开源的R语言 (<a target="_blank" rel="noopener" href="http://www.r-project.org/">http://www.r-project.org/</a> )写成，可以在Linux, Windows， MacOS等多种计算机平台上运行。seqRFLP已经被RCRAN接收，用户可以在R的任何一个镜像下载。本说明假设用户使用Windows，其他平台用户的使用方法与之类似，均需先安装R软件。</p>
<h1 id="2-seqRFLP的下载和安装"><a href="#2-seqRFLP的下载和安装" class="headerlink" title="2. seqRFLP的下载和安装"></a>2. seqRFLP的下载和安装</h1><p>首先下载和安装R软件。建议选择较新的版本，如Windows: R-2.11.1，其下载地址之一为:</p>
<p><a target="_blank" rel="noopener" href="http://ftp.ctex.org/mirrors/CRAN/bin/windows/base/R-2.11.1-win32.exe">http://ftp.ctex.org/mirrors/CRAN/bin/windows/base/R-2.11.1-win32.exe</a></p>
<p>下载到本地后，双击R-2.11.1-win32.exe开始安装，各选项默认即可。安装完成后双击</p>
<p>开始&gt;所有程序&gt;R&gt;R 2.11.1</p>
<p>输入命令:</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">install.packages(<span class="string">&quot;seqRFLP&quot;</span>)</span><br></pre></td></tr></table></figure>
<p>R将提示选择CRAN,选择距离较近的CRAN镜像，将软件包自动下载和安装好。</p>
<p>导入seqRFLP程序包：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">library(seqRFLP)</span><br></pre></td></tr></table></figure>
<h1 id="3-查询seqRFLP的帮助"><a href="#3-查询seqRFLP的帮助" class="headerlink" title="3.查询seqRFLP的帮助"></a>3.查询seqRFLP的帮助</h1><p>seqRFLP 的每个函数都有详细的帮助文件，要查询帮助可输入:</p>
<p><code>?seqRFLP</code> 查看详尽的帮助手册:用户也可以在CRAN上下载pdf版本的帮助手册。</p>
<p>seqRFLP符合GPL-2协议，用户可以免费使用并拷贝该软件，也可以更改其源代码。</p>
<p>运行seqRFLP实例:</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">example(seqRFLP)</span><br></pre></td></tr></table></figure>

<h1 id="致-谢"><a href="#致-谢" class="headerlink" title="致 谢"></a>致 谢</h1><p>感谢马克平研究员、梁宇博士的指导，感谢博士研究生孙秀峰及龙恩熙对软件的测试，感谢软件包的编写有过帮助的研究组各同学和老师。</p>
<h1 id="参考文献"><a href="#参考文献" class="headerlink" title="参考文献"></a>参考文献</h1><p>Saiki RK, Scharf S, Faloona F, Mullis KB, Erlich HA, Arnheim N (1985). Enzymatic amplification of beta-globin genomic sequences and restriction site analysis for diagnosis of sickle cell anemia. Science 230 (4723):1350-4</p>
<p>Roberts, R.J., Vincze, T., Posfai, J., Macelis, D. (2010) REBASE–a database for DNA restriction and modification: enzymes, genes and genomes. Nucl. Acids Res. 38: D234-D236. <a target="_blank" rel="noopener" href="http://rebase.neb.com/">http://rebase.neb.com</a></p>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2010/07/03/2010-7-3-seqRFLP/" data-id="cl1n7qj1y001neo84a2nmbxes" data-title="从DNA序列预测RFLP—seqRFLP软件包简介" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6/" rel="tag">软件</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2010-6-14-phylocom" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2010/06/14/2010-6-14-phylocom/" class="article-date">
  <time class="dt-published" datetime="2010-06-14T07:05:54.000Z" itemprop="datePublished">2010-06-14</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2010/06/14/2010-6-14-phylocom/">用Phylomatic和PhyloCom进行群落系统进化分析(20180820)</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>Phylomatic (<a target="_blank" rel="noopener" href="http://phylodiversity.net/phylomatic/">http://phylodiversity.net/phylomatic/</a>) 和PhyloCom (<a target="_blank" rel="noopener" href="http://phylodiversity.net/phylocom/">http://phylodiversity.net/phylocom/</a>) 是Cam Webb、Steve Kembel和David Ackerly编写的软件，用于群落系统发育分析。</p>
<p>Phylomatic可基于植物名录以及相应进化树骨架生成进化树。不过，通过Phylomatic建立的进化树，一般只能准确反映科或者属以上分类等级的进化关系（取决于该属是否在骨架进化树上是否有信息）。由于不能准确反映科或属以下的系统发育关系，Phylomatic生成的进化树，一般只用于群落系统发育，或者空间尺度较大，类群较多，且系统发育关系较远的系统发育多样性分析中。如果分类单元数比较少，在Genbank中又有足够的序列， 则应该优先考虑自行建立分子系统树，再进行相应分析，具体请参考 （Roquet et al. 2013) 。</p>
<p>通过植物名录建立进化树， 目前也有两个程序包可以选择， 一个是浙江大学金毅老师编写的 S.PhyloMaker R 脚本 (<a target="_blank" rel="noopener" href="https://github.com/jinyizju/S.PhyloMaker">https://github.com/jinyizju/S.PhyloMaker</a>), 另一个是 Scott Chamberlain 编写的 brranching R 程序包 (<a target="_blank" rel="noopener" href="https://github.com/ropensci/brranching)%E3%80%82">https://github.com/ropensci/brranching)。</a> S.PhyloMaker 的数据是基于Zanne 2014进化树生成的，可建立包含几万个分类单元的进化树，但速度较慢。 brranching 主要是基于Phylomatic的，可以直接输入植物名录获得进化树，速度较快，但是在分类单元数量上有一定限制。以上两个程序包的用法，用户可自行参考指南，本文只介绍通过Phylomatic建立进化树。</p>
<p>Phylocom是用来进行群落系统发育分析的软件。该软件可以为Phylomatic软件得到的无枝长进化树，用BLADJ方法拟合出枝长（这是因为Phylomatic软件建立的进化树， 有科属种的标签，而另外一个提供的ages文件， 有科属种的分化时间），不过BLADJ算法在2014年以后已经很少使用。此外，Phylocom还可以：计算系统发育多样性（PD），系统发育结构(community structure)，群落的系统发育距离(community phylogenetic distance)，分析群落的性状进化(AOT)等。下面就介绍如何使用Phylomatic和Phylocom进行相应的分析。</p>
<p>假设有如下植物名录要建立进化树：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line">Castanopsis eyrei</span><br><span class="line">Castanopsis carlesii</span><br><span class="line">Castanopsis fargesii</span><br><span class="line">Castanopsis tibetana</span><br><span class="line">Cyclobalanopsis gracilis</span><br><span class="line">Cyclobalanopsis glauca</span><br><span class="line">Lithocarpus glaber</span><br><span class="line">Myrica rubra</span><br><span class="line">Elaeocarpus decipiens</span><br><span class="line">Syzygium buxifolium</span><br><span class="line">Daphniphyllum oldhamii</span><br><span class="line">Loropetalum chinense</span><br><span class="line">Rhododendron latoucheae</span><br><span class="line">Rhododendron ovatum</span><br><span class="line">Vaccinium carlesii</span><br><span class="line">Schima superba</span><br><span class="line">Adinandra millettii</span><br><span class="line">Eurya muricata</span><br><span class="line">Camellia fraterna</span><br><span class="line">Ternstroemia gymnanthera</span><br><span class="line">Machilus thunbergii</span><br><span class="line">Neolitsea aurata var. chekiangensis</span><br><span class="line">Chimonanthus salicifolius</span><br></pre></td></tr></table></figure>

<p>一般流程如下</p>
<h1 id="1-准备-科-属-种名录"><a href="#1-准备-科-属-种名录" class="headerlink" title="1. 准备 科/属/种名录"></a>1. 准备 科/属/种名录</h1><p>请注意: 《中国植物志》 以及Flora of China等植物志, 由于成书较早，所以采用的分类系统都不是APG。而APG分类系统是目前为止能够较为客观体现科水平系统发育关系的分类系统（最新版本APG IV），因而是推荐采用的。不过目前能获得的准确科属列表为APGIII的版本，与APG IV差别不大。为了批量查询植物所在APG的科名，可使用本人编写的R程序包 plantlist (<a target="_blank" rel="noopener" href="https://github.com/helixcn/plantlist)%E3%80%82%E8%AF%A5%E7%A8%8B%E5%BA%8F%E5%8C%85%E7%9A%84%E7%AE%80%E8%A6%81%E4%BB%8B%E7%BB%8D%E5%8F%82%E8%A7%81">https://github.com/helixcn/plantlist)。该程序包的简要介绍参见</a> <a target="_blank" rel="noopener" href="http://bbs.sciencenet.cn/blog-255662-846673.html">http://blog.sciencenet.cn/blog-255662-846673.html</a> 。</p>
<p>注意： 由于plantlist给出的是APGIII的科属列表，如果用户要使用APG IV，请自行修订。</p>
<h2 id="1-1-plantlist程序包的安装"><a href="#1-1-plantlist程序包的安装" class="headerlink" title="1.1 plantlist程序包的安装"></a>1.1 plantlist程序包的安装</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">library(devtools)</span><br><span class="line">install_github(<span class="string">&quot;helixcn/plantlist&quot;</span>)</span><br></pre></td></tr></table></figure>

<h2 id="1-2-生成科属种列表"><a href="#1-2-生成科属种列表" class="headerlink" title="1.2 生成科属种列表"></a>1.2 生成科属种列表</h2><p>R 命令为</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">sp &lt;- <span class="built_in">c</span>( <span class="string">&quot;Castanopsis eyrei&quot;</span>,<span class="string">&quot;Castanopsis carlesii&quot;</span>,</span><br><span class="line"><span class="string">&quot;Castanopsis fargesii&quot;</span>,<span class="string">&quot;Castanopsis tibetana&quot;</span>,</span><br><span class="line"><span class="string">&quot;Cyclobalanopsis gracilis&quot;</span>,<span class="string">&quot;Cyclobalanopsis glauca&quot;</span>,</span><br><span class="line"><span class="string">&quot;Lithocarpus glaber&quot;</span>,<span class="string">&quot;Myrica rubra&quot;</span>,</span><br><span class="line"><span class="string">&quot;Elaeocarpus decipiens&quot;</span>,<span class="string">&quot;Syzygium buxifolium&quot;</span>,</span><br><span class="line"><span class="string">&quot;Daphniphyllum oldhamii&quot;</span>,<span class="string">&quot;Loropetalum chinense&quot;</span>,</span><br><span class="line"><span class="string">&quot;Rhododendron latoucheae&quot;</span>,<span class="string">&quot;Rhododendron ovatum&quot;</span>,</span><br><span class="line"><span class="string">&quot;Vaccinium carlesii&quot;</span>,<span class="string">&quot;Schima superba&quot;</span>,</span><br><span class="line"><span class="string">&quot;Adinandra millettii&quot;</span>,<span class="string">&quot;Eurya muricata&quot;</span>,</span><br><span class="line"><span class="string">&quot;Camellia fraterna&quot;</span>,<span class="string">&quot;Ternstroemia gymnanthera&quot;</span>,</span><br><span class="line"><span class="string">&quot;Machilus thunbergii&quot;</span>,<span class="string">&quot;Neolitsea aurata var. chekiangensis&quot;</span>,</span><br><span class="line"><span class="string">&quot;Chimonanthus salicifolius&quot;</span>)</span><br><span class="line"></span><br><span class="line">library(plantlist)</span><br><span class="line">sp2 &lt;- TPL(sp)</span><br><span class="line">res &lt;- taxa.table(sp2)</span><br><span class="line">writeLines(res, <span class="string">&quot;taxa.table.txt&quot;</span>)</span><br><span class="line">getwd()</span><br></pre></td></tr></table></figure>

<p>运行以上R代码，科属种列表就保存在taxa.table.txt 文件中，保存的路径在R中有所显示。</p>
<p>所生成文件taxa.table.txt用Notepad++ (<a target="_blank" rel="noopener" href="https://notepad-plus-plus.org/download/">https://notepad-plus-plus.org/download/</a> ) 记事本打开， 内容如下：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line">Fagaceae&#x2F;Castanopsis&#x2F;Castanopsis_eyrei</span><br><span class="line">Fagaceae&#x2F;Castanopsis&#x2F;Castanopsis_carlesii</span><br><span class="line">Fagaceae&#x2F;Castanopsis&#x2F;Castanopsis_fargesii</span><br><span class="line">Fagaceae&#x2F;Castanopsis&#x2F;Castanopsis_tibetana</span><br><span class="line">Fagaceae&#x2F;Cyclobalanopsis&#x2F;Cyclobalanopsis_gracilis</span><br><span class="line">Fagaceae&#x2F;Cyclobalanopsis&#x2F;Cyclobalanopsis_glauca</span><br><span class="line">Fagaceae&#x2F;Lithocarpus&#x2F;Lithocarpus_glaber</span><br><span class="line">Myricaceae&#x2F;Myrica&#x2F;Myrica_rubra</span><br><span class="line">Elaeocarpaceae&#x2F;Elaeocarpus&#x2F;Elaeocarpus_decipiens</span><br><span class="line">Myrtaceae&#x2F;Syzygium&#x2F;Syzygium_buxifolium</span><br><span class="line">Daphniphyllaceae&#x2F;Daphniphyllum&#x2F;Daphniphyllum_oldhamii</span><br><span class="line">Hamamelidaceae&#x2F;Loropetalum&#x2F;Loropetalum_chinense</span><br><span class="line">Ericaceae&#x2F;Rhododendron&#x2F;Rhododendron_latoucheae</span><br><span class="line">Ericaceae&#x2F;Rhododendron&#x2F;Rhododendron_ovatum</span><br><span class="line">Ericaceae&#x2F;Vaccinium&#x2F;Vaccinium_carlesii</span><br><span class="line">Theaceae&#x2F;Schima&#x2F;Schima_superba</span><br><span class="line">Theaceae&#x2F;Camellia&#x2F;Camellia_fraterna</span><br><span class="line">Pentaphylacaceae&#x2F;Adinandra&#x2F;Adinandra_millettii</span><br><span class="line">Pentaphylacaceae&#x2F;Eurya&#x2F;Eurya_muricata</span><br><span class="line">Pentaphylacaceae&#x2F;Ternstroemia&#x2F;Ternstroemia_gymnanthera</span><br><span class="line">Lauraceae&#x2F;Machilus&#x2F;Machilus_thunbergii</span><br><span class="line">Lauraceae&#x2F;Neolitsea&#x2F;Neolitsea_aurata_var._chekiangensis</span><br><span class="line">Calycanthaceae&#x2F;Chimonanthus&#x2F;Chimonanthus_salicifolius</span><br></pre></td></tr></table></figure>

<p>注： 查询植物科名，也可以用 taxize程序包（<a target="_blank" rel="noopener" href="https://ropensci.org/tutorials/taxize_tutorial/%EF%BC%89">https://ropensci.org/tutorials/taxize_tutorial/）</a></p>
<h1 id="2-Phylomatic建树"><a href="#2-Phylomatic建树" class="headerlink" title="2 Phylomatic建树"></a>2 Phylomatic建树</h1><p>在Zanne 2014进化树发表之前，用Phylomatic建立进化树一般只是按照APGIII的骨架，生成无枝长的newick进化树。然后再在Phylocom软件中，用BLADJ模块，将Wikstroem等人推断的各类群分化时间拟合到无枝长的进化树上。Wikstroem的分化时间由于发表年代久远， 被子植物各类群的分化时间有一些变动，通过Phylocom BLADJ的方法校正分化时间已经显得过时。从2015年开始, Phylomatic网站整合了 Zanne 2014的进化树骨架, 可以直接基于该进化树直接获得有枝长的进化树。</p>
<p>方法如下:</p>
<ol>
<li>打开 <a target="_blank" rel="noopener" href="http://phylodiversity.net/phylomatic/">http://phylodiversity.net/phylomatic/</a></li>
<li>将 科/属/种 列表 拷贝到 taxa = 框中</li>
<li>storedtree = 选择 Zanne 2014</li>
<li>勾选 clean = TRUE， 其余选择默认.</li>
<li>点击Send按钮获得结果</li>
</ol>
<p>将网页中的输出结果拷贝到文本文件中， 并另存为 phylo文件， 请注意，如果后续用Phylocom进行分析，则该phylo文件不能有任何扩展名。该进化树为Newick格式， 不含Singletons，因此可以用R的ape读取并进行数据处理，或者Figtree软件绘图。</p>
<p>获得的进化树如下：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">((Chimonanthus_salicifolius:108.8,(Neolitsea_aurata_var._chekiangensis:88.1284,</span><br><span class="line">Machilus_thunbergii:88.1284)Lauraceae:20.672)Laurales:71.8279,</span><br><span class="line">(((((((Cyclobalanopsis_gracilis:20.205,Cyclobalanopsis_glauca:20.205)</span><br><span class="line">Cyclobalanopsis:20.205,(Lithocarpus_glaber:3.30372,</span><br><span class="line">(Castanopsis_eyrei:0.94544,(Castanopsis_carlesii:0.320126,</span><br><span class="line">(Castanopsis_tibetana:0.28527,Castanopsis_fargesii:0.28527):0.034856):</span><br><span class="line">0.625313):2.35828):37.1063)Fagaceae:21.8954,Myrica_rubra:62.3054):</span><br><span class="line">50.9532,Elaeocarpus_decipiens:113.259)Fabidae:4.27613,Syzygium_buxifolium:</span><br><span class="line">117.535)Rosidae:0.138911,(Daphniphyllum_oldhamii:107.877,Loropetalum_chinense:</span><br><span class="line">107.877):9.79685)Superrosidae:1.49186,((Rhododendron_latoucheae:62.0517,</span><br><span class="line">Vaccinium_carlesii:62.0517,Rhododendron_ovatum:62.0517)Ericaceae:20.7165,</span><br><span class="line">((Ternstroemia_gymnanthera:43.8733,(Adinandra_millettii:33.7061,</span><br><span class="line">Eurya_muricata:33.7061):10.1672):27.9519,(Schima_superba:57.076,</span><br><span class="line">Camellia_fraterna:57.076):14.7492):10.943):36.3974)</span><br><span class="line">Gunneridae:61.4628):259.665;</span><br></pre></td></tr></table></figure>

<p>注：用brranching R程序包可以直接输入物种名，获得进化树。</p>
<h1 id="3-用Phylocom计算PD-NRI-和NTI-（以Example文件为例）"><a href="#3-用Phylocom计算PD-NRI-和NTI-（以Example文件为例）" class="headerlink" title="3. 用Phylocom计算PD, NRI 和NTI （以Example文件为例）"></a>3. 用Phylocom计算PD, NRI 和NTI （以Example文件为例）</h1><p>注意：这部分计算使用的是Example文件，与上文的植物名录已经没有关系。</p>
<h2 id="3-1-下载phylocom软件"><a href="#3-1-下载phylocom软件" class="headerlink" title="3.1 下载phylocom软件"></a>3.1 下载phylocom软件</h2><p>打开 <a target="_blank" rel="noopener" href="https://github.com/phylocom/phylocom/releases">https://github.com/phylocom/phylocom/releases</a> 下载zip文件，解压缩, 可见以下几个文件夹：</p>
<ol>
<li>Example_data 中是练习数据</li>
<li>Mac 中是苹果机的运行软件</li>
<li>R，是驱动Phylocom的R脚本</li>
<li>Src 是Phylocom的源程序，phylocom是C语言写的</li>
<li>W32 是Windows平台下可以运行的exe程序</li>
<li>Phylocom_manual.pdf 是软件说明书</li>
<li>README 注意事项</li>
</ol>
<h2 id="3-2-创建工作路径"><a href="#3-2-创建工作路径" class="headerlink" title="3.2 创建工作路径"></a>3.2 创建工作路径</h2><p>Phylocom必须在纯ASCII码的路径下运行，运行路径不能有中文。在C盘根目录下，创建一个名为phylocom的文件夹。将w32中的phylocom.exe文件，拷贝到文件夹 C:/phylocom下。 将example_data文件夹中的phylo和sample文件，拷贝到该文件夹C:/phylocom中。</p>
<h2 id="3-3-利用phylocom软件计算PD、NRI、NTI等指数"><a href="#3-3-利用phylocom软件计算PD、NRI、NTI等指数" class="headerlink" title="3.3 利用phylocom软件计算PD、NRI、NTI等指数"></a>3.3 利用phylocom软件计算PD、NRI、NTI等指数</h2><p>sample文件第一列是样方的名称，第二列是个体数，第三列是物种名。物种名一定要与phylo文件中的物种名完全对应（包括大小写）。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">clump1    1    sp1</span><br><span class="line">clump1    1    sp2</span><br><span class="line">clump1    1    sp3</span><br><span class="line">clump1    1    sp4</span><br><span class="line">clump1    1    sp5</span><br><span class="line">clump1    1    sp6</span><br><span class="line">clump1    1    sp7</span><br><span class="line">clump1    1    sp8</span><br><span class="line">clump2a    1    sp1</span><br><span class="line">clump2a    1    sp2</span><br><span class="line">.... (中间省略)</span><br><span class="line">random    2    sp15</span><br><span class="line">random    3    sp17</span><br><span class="line">random    1    sp22</span><br><span class="line">random    2    sp24</span><br></pre></td></tr></table></figure>

<p>将phylo和sample文件夹拷贝到 C:\phylocom 之后，</p>
<p>点击 开始&gt; cmd &gt; 输入 <code>cd C:\phylocom&gt;</code></p>
<p>输入 <code>phylocom comstruct &gt; comstruct.output</code> 计算群落的系统发育结构的NRI和NTI指数</p>
<p>输入 <code>phylocom pd &gt; pd.output.txt</code> 计算系统发育多样性 Faith’s PD</p>
<p>输入 <code>phylocom comdist &gt; comdist.output.txt</code> 计算群落两两之间系统发育距离 MPD的MNDT等</p>
<p>注意： 进行系统发育多样性相关的计算， 可以参考picante 程序包 (<a target="_blank" rel="noopener" href="http://bbs.sciencenet.cn/home.php?mod=space&uid=255662&do=blog&id=1116212">http://blog.sciencenet.cn/home.php?mod=space&amp;uid=255662&amp;do=blog&amp;id=1116212</a>)</p>
<h1 id="参考文献"><a href="#参考文献" class="headerlink" title="参考文献"></a>参考文献</h1><ol>
<li>Chamberlain, S. (2018). brranching: Fetch ‘Phylogenies’ from Many Sources. R package version 0.3.0. <a target="_blank" rel="noopener" href="https://cran.r-project.org/package=brranching">https://CRAN.R-project.org/package=brranching</a></li>
<li>Qian, H. and Y. Jin. (2016) An updated megaphylogeny of plants, a tool for generating plant phylogenies and an analysis of phylogenetic community structure. Journal of Plant Ecology 9(2): 233–239.</li>
<li>Roquet, C., Thuiller, W., &amp; Lavergne, S. (2013). Building megaphylogenies for macroecology: taking up the challenge. Ecography, 36(1), 13-26.</li>
<li>Webb, C. O., &amp; Donoghue, M. J. (2005). Phylomatic: tree assembly for applied phylogenetics. Molecular Ecology Notes, 5(1), 181-183.</li>
<li>Zhang, J.L. (2018). plantlist: Looking Up the Status of Plant Scientific Names based on The Plant List Database. R package version 0.3.7/r31. <a target="_blank" rel="noopener" href="https://r-forge.r-project.org/projects/plantlist/">https://R-Forge.R-project.org/projects/plantlist/</a></li>
</ol>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2010/06/14/2010-6-14-phylocom/" data-id="cl1n7qj1y001keo84391vfe5y" data-title="用Phylomatic和PhyloCom进行群落系统进化分析(20180820)" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6/" rel="tag">软件</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2010-05-18-spaa" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2010/05/18/2010-05-18-spaa/" class="article-date">
  <time class="dt-published" datetime="2010-05-18T03:37:54.000Z" itemprop="datePublished">2010-05-18</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2010/05/18/2010-05-18-spaa/">计算种间关联度(种间连接)的spaa程序包</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>编写一个计算种间关联的小程序是出于以下原因：</p>
<p>查阅了很多进行种间关联分析的论文，一直也没有提及用什么软件。 为此我觉得有必要编写一个程序包，用来进行群落数据的处理及种间关联度计算和绘图。</p>
<p>R当然成为首选，因为R是开源的，所有代码可见；同时又是免费的，不用注册，就可以使用R。</p>
<p>花费几天的时间，终于完成了。定名为spaa, 意为 <code>species association analysis</code>。可以在R的CRAN上下载</p>
<p><a target="_blank" rel="noopener" href="http://cran.r-project.org/web/packages/spaa/index.html">http://cran.r-project.org/web/packages/spaa/index.html</a></p>
<p>spaa中有如下几个函数：</p>
<ul>
<li><code>data2mat</code> 将物种列表转换为矩阵 </li>
<li><code>freq.calc</code> 计算物种频度 </li>
<li><code>plot.lowertri</code> 绘制半矩阵图 </li>
<li><code>plot.network</code> 绘制种间关联图 （有待进一步完善） </li>
<li><code>sp.assoc</code> 计算种间总体关联性 </li>
<li><code>sp.pair</code> 计算物种两两之间的关联性 </li>
<li><code>sub.sp.matrix</code> 在总矩阵中，按照物种出现的频度筛选物种。</li>
</ul>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2010/05/18/2010-05-18-spaa/" data-id="cl1n7qj1r0012eo8482ide9fl" data-title="计算种间关联度(种间连接)的spaa程序包" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6/" rel="tag">软件</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2010-4-8-clustering" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2010/04/08/2010-4-8-clustering/" class="article-date">
  <time class="dt-published" datetime="2010-04-07T21:55:54.000Z" itemprop="datePublished">2010-04-08</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2010/04/08/2010-4-8-clustering/">R软件中如何进行群落聚类分析？</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>群落按照物种相似形组成进行聚类分析，可以用树状图较好的表现物种的组成关系。受到很多植被学家的重视。这里以R软件实现聚类分析为例。</p>
<p>如果按照物种组成的相似性做聚类分析，那么可以用Jaccard指数（经过转换的）。Jaccard指数只考虑物种在两个样方间是否重复出现，盖度在分析的过程中并不起什么作用。但是如果对乔木和灌木进行分析，就可以考虑个体的数量，计算样方物种组成的相似性的时候用Bray-Curtis指数。Jaccard指数和Bray-Curtis指数在众多生态学相关的程序包中都是可以计算的。下面说一下在R软件中，结合vegan程序包，对草本样方的物种组成进行聚类分析。 </p>
<p>下面是在R中的具体操作过程：</p>
<h2 id="第一步-矩阵的整理，建议先整理一下各样地的名录，成如下格式，再用R整理成物种矩阵。"><a href="#第一步-矩阵的整理，建议先整理一下各样地的名录，成如下格式，再用R整理成物种矩阵。" class="headerlink" title="第一步  矩阵的整理，建议先整理一下各样地的名录，成如下格式，再用R整理成物种矩阵。"></a>第一步  矩阵的整理，建议先整理一下各样地的名录，成如下格式，再用R整理成物种矩阵。</h2><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">plotname species</span><br><span class="line">plot1 sp1</span><br><span class="line">plot1 sp2</span><br><span class="line">plot1 sp3</span><br><span class="line">plot1 sp4</span><br><span class="line">plot1 sp5</span><br><span class="line">plot2 sp1</span><br><span class="line">plot2 sp3</span><br><span class="line">plot3 sp4</span><br><span class="line">plot3 sp2</span><br><span class="line">plot3 sp6</span><br><span class="line">plot3 sp7</span><br><span class="line">.....</span><br></pre></td></tr></table></figure>
<p>在Excel中，另存为csv格式，如存名称为 herbplots.csv。</p>
<h2 id="第二步-读取文件"><a href="#第二步-读取文件" class="headerlink" title="第二步 读取文件"></a>第二步 读取文件</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">herb.data&lt;- read.csv(<span class="string">&quot;D:/herb/herbplots.csv&quot;</span>, header=<span class="built_in">T</span>)</span><br></pre></td></tr></table></figure>
<h2 id="第三步-转换为-矩阵"><a href="#第三步-转换为-矩阵" class="headerlink" title="第三步 转换为 矩阵"></a>第三步 转换为 矩阵</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">pre.matrix &lt;- table(herb.data)</span><br></pre></td></tr></table></figure>
<p>此时生成的矩阵，形式如下：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">plotname sp1 sp2 sp3 sp4 sp5 sp6 sp7</span><br><span class="line">plot1 1 1 1 1 1 0 0</span><br><span class="line">plot2 1 0 1 0 0 0 0</span><br><span class="line">plot3 0 1 0 1 0 1 1</span><br></pre></td></tr></table></figure>
<h2 id="第四步-计算各样方的Jaccard距离"><a href="#第四步-计算各样方的Jaccard距离" class="headerlink" title="第四步 计算各样方的Jaccard距离"></a>第四步 计算各样方的Jaccard距离</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plot.dist &lt;- vegdist(pre.matrix, method=<span class="string">&quot;jaccard&quot;</span>)</span><br></pre></td></tr></table></figure>
<p>距离矩阵为一个半矩阵，形式如下</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">plot1 plot2</span><br><span class="line">plot2 0.6000000</span><br><span class="line">plot3 0.7142857 1.0000000</span><br></pre></td></tr></table></figure>

<h2 id="第五步-使用UPGMA方法或者Ward法，对群落进行聚类"><a href="#第五步-使用UPGMA方法或者Ward法，对群落进行聚类" class="headerlink" title="第五步 使用UPGMA方法或者Ward法，对群落进行聚类"></a>第五步 使用UPGMA方法或者Ward法，对群落进行聚类</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plot.hc &lt;- hclust(plot.dist , <span class="string">&quot;ave&quot;</span>)</span><br></pre></td></tr></table></figure>
<p>绘图</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">plot(plot.hc)</span><br><span class="line">plot(plot.hc, hang=-<span class="number">1</span>)</span><br></pre></td></tr></table></figure>
<p>转换为dendrogram之后绘图</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">dendro.plot&lt;-as.dendrogram(plot.hc)</span><br><span class="line">plot(dendro.plot, horiz=<span class="built_in">T</span>)</span><br></pre></td></tr></table></figure>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2010/04/08/2010-4-8-clustering/" data-id="cl1n7qj1x001ieo848ag6600w" data-title="R软件中如何进行群落聚类分析？" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E7%94%9F%E6%80%81/" rel="tag">生态</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2010-04-8-biodiversity" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2010/04/08/2010-04-8-biodiversity/" class="article-date">
  <time class="dt-published" datetime="2010-04-07T21:52:54.000Z" itemprop="datePublished">2010-04-08</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2010/04/08/2010-04-8-biodiversity/">R软件计算生物多样性指数</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>生物多样性指数的计算在R中计算十分简单，这里介绍一下用vegan包计算生物多样性指数的方法。</p>
<p>首先要将R软件和用到的程序包安装好。在数据处理和格式转换过程中，需要用到openxlsx、reshape2和vegan包。如果没有安装这几个程序包，就在R console中输入以下命令安装。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">install.packages(<span class="string">&#x27;reshape2&#x27;</span>)</span><br><span class="line">install.packages(<span class="string">&#x27;openxlsx&#x27;</span>)</span><br><span class="line">install.packages(<span class="string">&#x27;vegan&#x27;</span>)</span><br></pre></td></tr></table></figure>

<h1 id="第一步-在Excel中输入数据"><a href="#第一步-在Excel中输入数据" class="headerlink" title="第一步 在Excel中输入数据"></a>第一步 在Excel中输入数据</h1><p>建议将数据整理成样方、物种、个体数格式，这也是样方调查时最常见的一种格式，将数据输入到Excel中。注意:Excel文件列的名字要与下面的格式完全一致，因为R对大小写敏感，所以大小写也要一致。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">plotname species abundance</span><br><span class="line">plot1     sp1         3</span><br><span class="line">plot1     sp2         6</span><br><span class="line">plot1     sp3         1</span><br><span class="line">plot1     sp4         2</span><br><span class="line">plot1     sp5         1</span><br><span class="line">plot2     sp1         8</span><br><span class="line">plot2     sp3        30</span><br><span class="line">plot3     sp4         2</span><br><span class="line">plot3     sp2         1</span><br><span class="line">plot3     sp6         1</span><br><span class="line">plot3     sp7         3</span><br><span class="line">.....</span><br></pre></td></tr></table></figure>

<p>保存为 herbplots.xlsx，假设保存在“D:/herb/”。</p>
<h1 id="第二步-读取Excel文件"><a href="#第二步-读取Excel文件" class="headerlink" title="第二步 读取Excel文件"></a>第二步 读取Excel文件</h1><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">library(openxlsx)</span><br><span class="line">library(reshape2)</span><br><span class="line">library(vegan)</span><br><span class="line">herb.data &lt;- read.xlsx(<span class="string">&quot;D:/herb/herbplots.xlsx&quot;</span>)</span><br></pre></td></tr></table></figure>

<h1 id="第三步-转换为样方-物种矩阵"><a href="#第三步-转换为样方-物种矩阵" class="headerlink" title="第三步 转换为样方-物种矩阵"></a>第三步 转换为样方-物种矩阵</h1><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">library(reshape2)</span><br><span class="line">herb.mat &lt;- acast(herb.data, </span><br><span class="line">             formula = plotname ~ species , </span><br><span class="line">             value.var = <span class="string">&quot;abundance&quot;</span>, </span><br><span class="line">             fill = <span class="number">0</span>)</span><br></pre></td></tr></table></figure>

<p>此时生成的矩阵，格式如下：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">       sp1 sp2 sp3 sp4 sp5 sp6 sp7</span><br><span class="line">plot1   3   6   1   2   1   0   0</span><br><span class="line">plot2   8   0  30   0   0   0   0</span><br><span class="line">plot3   0   1   0   2   0   1   3</span><br></pre></td></tr></table></figure>

<h1 id="第四步-计算生物多样性指数"><a href="#第四步-计算生物多样性指数" class="headerlink" title="第四步 计算生物多样性指数"></a>第四步 计算生物多样性指数</h1><h2 id="Shannon-Wiener指数"><a href="#Shannon-Wiener指数" class="headerlink" title="Shannon-Wiener指数"></a>Shannon-Wiener指数</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">Shannon.Wiener &lt;- diversity(herb.mat, index = <span class="string">&quot;shannon&quot;</span>)</span><br></pre></td></tr></table></figure>

<h2 id="Simpson指数"><a href="#Simpson指数" class="headerlink" title="Simpson指数"></a>Simpson指数</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">Simpson &lt;- diversity(herb.mat, index = <span class="string">&quot;simpson&quot;</span>)</span><br></pre></td></tr></table></figure>

<h2 id="Inverse-Simpson指数"><a href="#Inverse-Simpson指数" class="headerlink" title="Inverse Simpson指数"></a>Inverse Simpson指数</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">Inverse.Simpson &lt;- diversity(herb.mat, index = <span class="string">&quot;inv&quot;</span>)</span><br></pre></td></tr></table></figure>

<h2 id="物种累计数"><a href="#物种累计数" class="headerlink" title="物种累计数"></a>物种累计数</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">S &lt;- specnumber(herb.mat)</span><br><span class="line">plot(S)</span><br></pre></td></tr></table></figure>

<h2 id="Pielou均匀度指数"><a href="#Pielou均匀度指数" class="headerlink" title="#Pielou均匀度指数"></a>#Pielou均匀度指数</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">J &lt;- Shannon.Wiener/<span class="built_in">log</span>(S)</span><br></pre></td></tr></table></figure>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2010/04/08/2010-04-8-biodiversity/" data-id="cl1n7qj1q0010eo84bwr4dg3b" data-title="R软件计算生物多样性指数" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E7%94%9F%E6%80%81/" rel="tag">生态</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2010-03-27-hamma" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2010/03/27/2010-03-27-hamma/" class="article-date">
  <time class="dt-published" datetime="2010-03-27T04:59:54.000Z" itemprop="datePublished">2010-03-27</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2010/03/27/2010-03-27-hamma/">早春的金缕梅</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>金缕梅原产于我国亚热带地区。在湖北神农架山区，干旱的山顶上，生长有大量的金缕梅。北京的金缕梅均为栽培。</p>
<p>金缕梅是春季最早开花的植物之一。金缕梅先花后叶，花瓣金色，狭长，在风中微微飘动，十分惹人喜爱。</p>
<p>现将今天拍摄的几张照片粘贴上来，与读者共享。</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2010/03/27/2010-03-27-hamma/" data-id="cl1n7qj1p000xeo84g8zh8eek" data-title="早春的金缕梅" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E8%A1%8C%E6%91%84%E5%8C%86%E5%8C%86/" rel="tag">行摄匆匆</a></li></ul>

    </footer>
  </div>
  
</article>



  
    <article id="post-2010-03-27-biomod" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2010/03/27/2010-03-27-biomod/" class="article-date">
  <time class="dt-published" datetime="2010-03-27T00:39:54.000Z" itemprop="datePublished">2010-03-27</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="p-name article-title" href="/2010/03/27/2010-03-27-biomod/">物种分布随气候的变化:BIOMOD使用指南</a>
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>BIOMOD是R软件的一个程序包，是BIOdiversity MODelling的缩写，主要用来模型物种分布区及其随气候的变化。该软件是Wilfried Thuiller等三位学者在法国蒙彼利埃CNRS功能与进化生态学中心开发的。W. Thuiller是活跃在学术一线的年轻科学家，其发表的学术论文具有较大影响。</p>
<p>BIOMOD程序包可以运行以下模型，对物种潜在分布区进行估算，同时也能够对相应的气候变化的情形下，物种分布区的变化给出预测。预测过程中可以设置物种迁移速率，从而更接近真实结果。</p>
<ul>
<li>Generalised Linear Models (GLM)</li>
<li>Generalised Additive Models (GAM)</li>
<li>Classication Tree Analysis (CTA)</li>
<li>Articial Neural Networks (ANN)</li>
<li>Surface Range Envelope (SRE)</li>
<li>Generalised Boosting Model (GBM)</li>
<li>Breiman and Cutler’s random forest for classication and regression (randomForest)</li>
<li>Mixture Discriminant Analysis (MDA)</li>
<li>Multiple Adaptive Regression Splines (MARS)</li>
</ul>
<p>希望本文档对刚刚开始使用BIOMOD的程序包的人员有所帮助。</p>
<p>由于译者水平有限，在翻译过程中难免有不妥的地方，还请读者及时批评指正, 若有疑问可发邮件至<a href="mailto:jinlongzhang01@gmail.com">jinlongzhang01@gmail.com</a> 。</p>
<p>译 者</p>
<p>2010年1月</p>
<p>于北京 香山</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2010/03/27/2010-03-27-biomod/" data-id="cl1n7qj1o000veo843jaze6x6" data-title="物种分布随气候的变化:BIOMOD使用指南" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6/" rel="tag">软件</a></li></ul>

    </footer>
  </div>
  
</article>



  


  <nav id="page-nav">
    
    <a class="extend prev" rel="prev" href="/page/23/">&laquo; Prev</a><a class="page-number" href="/">1</a><span class="space">&hellip;</span><a class="page-number" href="/page/22/">22</a><a class="page-number" href="/page/23/">23</a><span class="page-number current">24</span><a class="page-number" href="/page/25/">25</a><a class="page-number" href="/page/26/">26</a><a class="extend next" rel="next" href="/page/25/">Next &raquo;</a>
  </nav>

</section>
        
          <aside id="sidebar">
  
    

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">Tags</h3>
    <div class="widget">
      <ul class="tag-list" itemprop="keywords"><li class="tag-list-item"><a class="tag-list-link" href="/tags/Python/" rel="tag">Python</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/R%E7%BC%96%E7%A8%8B/" rel="tag">R编程</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E4%BA%BA%E7%89%A9/" rel="tag">人物</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%86%99%E4%BD%9C/" rel="tag">写作</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" rel="tag">可视化</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9B%BE%E4%B9%A6/" rel="tag">图书</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9B%BE%E5%83%8F%E5%A4%84%E7%90%86/" rel="tag">图像处理</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9C%B0%E7%90%86/" rel="tag">地理</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%A3%B0%E6%98%8E/" rel="tag">声明</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%A4%A9%E6%96%87/" rel="tag">天文</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%AD%A6%E4%B9%A0%E8%B5%84%E6%BA%90/" rel="tag">学习资源</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%AF%B9%E8%AF%9D%E5%BD%95/" rel="tag">对话录</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%BF%83%E5%BE%97/" rel="tag">心得</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%8A%A5%E5%91%8A/" rel="tag">报告</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%8E%92%E7%89%88/" rel="tag">排版</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%85%E4%B9%A1/" rel="tag">故乡</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E5%BA%93/" rel="tag">数据库</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%A0%87%E6%9C%AC%E9%A6%86/" rel="tag">标本馆</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%A4%8D%E7%89%A9%E5%88%86%E7%B1%BB/" rel="tag">植物分类</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%A4%8D%E7%89%A9%E5%AD%A6%E5%90%8D/" rel="tag">植物学名</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B1%82%E8%81%8C/" rel="tag">求职</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%89%88%E6%9C%AC%E6%8E%A7%E5%88%B6/" rel="tag">版本控制</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%94%9F%E6%80%81/" rel="tag">生态</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%94%9F%E6%80%81%E8%B5%84%E8%AE%AF/" rel="tag">生态资讯</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%94%9F%E6%B4%BB%E7%82%B9%E6%BB%B4/" rel="tag">生活点滴</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%A0%94%E7%A9%B6%E8%B5%84%E8%AE%AF/" rel="tag">研究资讯</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" rel="tag">系统发育</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%9F%E8%AE%A1/" rel="tag">统计</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BC%96%E7%A8%8B/" rel="tag">编程</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%A1%8C%E6%91%84%E5%8C%86%E5%8C%86/" rel="tag">行摄匆匆</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%A7%81%E9%97%BB%E5%BD%95/" rel="tag">见闻录</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%AF%97%E6%AD%8C/" rel="tag">诗歌</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6/" rel="tag">软件</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BF%9B%E5%8C%96/" rel="tag">进化</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%9A%8F%E6%83%B3/" rel="tag">随想</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%A9%B3%E8%AE%BA/" rel="tag">驳论</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">Tag Cloud</h3>
    <div class="widget tagcloud">
      <a href="/tags/Python/" style="font-size: 10px;">Python</a> <a href="/tags/R%E7%BC%96%E7%A8%8B/" style="font-size: 18.75px;">R编程</a> <a href="/tags/%E4%BA%BA%E7%89%A9/" style="font-size: 11.88px;">人物</a> <a href="/tags/%E5%86%99%E4%BD%9C/" style="font-size: 13.13px;">写作</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" style="font-size: 12.5px;">可视化</a> <a href="/tags/%E5%9B%BE%E4%B9%A6/" style="font-size: 10px;">图书</a> <a href="/tags/%E5%9B%BE%E5%83%8F%E5%A4%84%E7%90%86/" style="font-size: 11.88px;">图像处理</a> <a href="/tags/%E5%9C%B0%E7%90%86/" style="font-size: 17.5px;">地理</a> <a href="/tags/%E5%A3%B0%E6%98%8E/" style="font-size: 10.63px;">声明</a> <a href="/tags/%E5%A4%A9%E6%96%87/" style="font-size: 16.25px;">天文</a> <a href="/tags/%E5%AD%A6%E4%B9%A0%E8%B5%84%E6%BA%90/" style="font-size: 10px;">学习资源</a> <a href="/tags/%E5%AF%B9%E8%AF%9D%E5%BD%95/" style="font-size: 10px;">对话录</a> <a href="/tags/%E5%BF%83%E5%BE%97/" style="font-size: 10px;">心得</a> <a href="/tags/%E6%8A%A5%E5%91%8A/" style="font-size: 12.5px;">报告</a> <a href="/tags/%E6%8E%92%E7%89%88/" style="font-size: 13.75px;">排版</a> <a href="/tags/%E6%95%85%E4%B9%A1/" style="font-size: 10.63px;">故乡</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E5%BA%93/" style="font-size: 10.63px;">数据库</a> <a href="/tags/%E6%A0%87%E6%9C%AC%E9%A6%86/" style="font-size: 14.38px;">标本馆</a> <a href="/tags/%E6%A4%8D%E7%89%A9%E5%88%86%E7%B1%BB/" style="font-size: 10px;">植物分类</a> <a href="/tags/%E6%A4%8D%E7%89%A9%E5%AD%A6%E5%90%8D/" style="font-size: 10px;">植物学名</a> <a href="/tags/%E6%B1%82%E8%81%8C/" style="font-size: 10px;">求职</a> <a href="/tags/%E7%89%88%E6%9C%AC%E6%8E%A7%E5%88%B6/" style="font-size: 11.25px;">版本控制</a> <a href="/tags/%E7%94%9F%E6%80%81/" style="font-size: 16.88px;">生态</a> <a href="/tags/%E7%94%9F%E6%80%81%E8%B5%84%E8%AE%AF/" style="font-size: 10px;">生态资讯</a> <a href="/tags/%E7%94%9F%E6%B4%BB%E7%82%B9%E6%BB%B4/" style="font-size: 10px;">生活点滴</a> <a href="/tags/%E7%A0%94%E7%A9%B6%E8%B5%84%E8%AE%AF/" style="font-size: 10px;">研究资讯</a> <a href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" style="font-size: 10px;">系统发育</a> <a href="/tags/%E7%BB%9F%E8%AE%A1/" style="font-size: 18.13px;">统计</a> <a href="/tags/%E7%BC%96%E7%A8%8B/" style="font-size: 10.63px;">编程</a> <a href="/tags/%E8%A1%8C%E6%91%84%E5%8C%86%E5%8C%86/" style="font-size: 15px;">行摄匆匆</a> <a href="/tags/%E8%A7%81%E9%97%BB%E5%BD%95/" style="font-size: 10px;">见闻录</a> <a href="/tags/%E8%AF%97%E6%AD%8C/" style="font-size: 15.63px;">诗歌</a> <a href="/tags/%E8%BD%AF%E4%BB%B6/" style="font-size: 19.38px;">软件</a> <a href="/tags/%E8%BF%9B%E5%8C%96/" style="font-size: 10.63px;">进化</a> <a href="/tags/%E9%9A%8F%E6%83%B3/" style="font-size: 11.25px;">随想</a> <a href="/tags/%E9%9A%8F%E7%AC%94/" style="font-size: 20px;">随笔</a> <a href="/tags/%E9%A9%B3%E8%AE%BA/" style="font-size: 10px;">驳论</a>
    </div>
  </div>

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">Archives</h3>
    <div class="widget">
      <ul class="archive-list"><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/04/">April 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/03/">March 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2022/01/">January 2022</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/11/">November 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/09/">September 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/07/">July 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/05/">May 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/03/">March 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/02/">February 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/12/">December 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/11/">November 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/06/">June 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/05/">May 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/02/">February 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/01/">January 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/10/">October 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/09/">September 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/07/">July 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/06/">June 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/05/">May 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/04/">April 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/03/">March 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/02/">February 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/01/">January 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/12/">December 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/11/">November 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/10/">October 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/09/">September 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/08/">August 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/07/">July 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/06/">June 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/05/">May 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/04/">April 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/03/">March 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/02/">February 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/01/">January 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/11/">November 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/10/">October 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/09/">September 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/08/">August 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/06/">June 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/05/">May 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/04/">April 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/02/">February 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/12/">December 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/11/">November 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/09/">September 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/06/">June 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/05/">May 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/04/">April 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/03/">March 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/01/">January 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/11/">November 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/10/">October 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/09/">September 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/07/">July 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/05/">May 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/04/">April 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/03/">March 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/02/">February 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/01/">January 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/12/">December 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/11/">November 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/10/">October 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/09/">September 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/07/">July 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/05/">May 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/03/">March 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/02/">February 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/01/">January 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/12/">December 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/10/">October 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/05/">May 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/04/">April 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/03/">March 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/02/">February 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/01/">January 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/12/">December 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/10/">October 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/08/">August 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/07/">July 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/06/">June 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/05/">May 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/04/">April 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/02/">February 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/01/">January 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/12/">December 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/11/">November 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/10/">October 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/08/">August 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/07/">July 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/05/">May 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/04/">April 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/03/">March 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/02/">February 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/12/">December 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/11/">November 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/09/">September 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/08/">August 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/07/">July 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/06/">June 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/05/">May 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/04/">April 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/03/">March 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/01/">January 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2009/08/">August 2009</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2009/07/">July 2009</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2009/06/">June 2009</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2009/05/">May 2009</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">Recent Posts</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/2022/04/01/2022-03-31run_phylomatic/">(no title)</a>
          </li>
        
          <li>
            <a href="/2022/03/31/2022-03-31run_phylomatic/">在本地运行awk版本的phylomatic-awk利用植物名录生成进化树</a>
          </li>
        
          <li>
            <a href="/2022/01/15/2022-01-15dangxia/">当下</a>
          </li>
        
          <li>
            <a href="/2021/11/05/2021-11-05lamtsuenriver/">林村河</a>
          </li>
        
          <li>
            <a href="/2021/09/13/2021-10-19school/">最初的小学时光</a>
          </li>
        
      </ul>
    </div>
  </div>

  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      
      &copy; 2022 Jinlong Zhang<br>
      Powered by <a href="https://hexo.io/" target="_blank">Hexo</a>
    </div>
  </div>
</footer>

    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">Home</a>
  
    <a href="/archives" class="mobile-nav-link">Archives</a>
  
    <a href="/about" class="mobile-nav-link">About</a>
  
</nav>
    


<script src="/js/jquery-3.4.1.min.js"></script>



  
<script src="/fancybox/jquery.fancybox.min.js"></script>




<script src="/js/script.js"></script>





  </div>
</body>
</html>